{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "7SnUxsLO5ldc"
   },
   "source": [
    "# <p style=\"text-align: center;\">RNA Knowledge Graph Analysis</p>\n",
    "    \n",
    "***\n",
    "***\n",
    "\n",
    "**Authors:** [ECavalleri](https://mail.google.com/mail/u/0/?view=cm&fs=1&tf=1&to=emanuele.cavalleri@unimi.it), [ACabri](https://mail.google.com/mail/u/0/?view=cm&fs=1&tf=1&to=alberto.cabri@unimi.it)\n",
    "\n",
    "**GitHub Repositories:** [testRNA-KG](https://github.com/emanuelecavalleri/testRNA-KG), [PheKnowLator](https://github.com/callahantiff/PheKnowLator/)\n",
    "  \n",
    "<br>  \n",
    "  \n",
    "**Purpose:** In the following notebook, we will be loading the simplified RNA-centered KG represented as the N-Triples (.nt) file generated by PheKnowLator. We will then proceed to analyze and visualize its network topology.\n",
    "\n",
    "<br>\n",
    "\n",
    "**Assumptions:**   \n",
    "- Knowledge graphs ➞ `./resources/knowledge_graphs`\n",
    "<br>\n",
    "\n",
    "**Dependencies:**   \n",
    "- **Scripts**: This notebook makes use of the [`GRAPE`](https://github.com/AnacletoLAB/grape/) tool.  \n",
    "- **Data**: All downloaded and generated data sources are provided through [this](https://drive.google.com/drive/folders/1sev5zczMviX7UVqMhTpkFXG43K3nQa9f) dedicated Google Drive repository. \n",
    "_____\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table of Contents\n",
    "***\n",
    "\n",
    "### [Preprocessing](#pre-processing)\n",
    "\n",
    "\n",
    "### [Neo4j](#neo4j)   \n",
    "\n",
    "\n",
    "### [GRAPE](#grape)  \n",
    "\n",
    "____\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set-Up Environment\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run this to install/update grape\n",
    "#!pip install --upgrade grape ensmallen embiggen graphviz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "UA3Xq9f58Dd9"
   },
   "outputs": [],
   "source": [
    "# import needed libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from typing import Union\n",
    "import re\n",
    "import requests\n",
    "from tqdm import tqdm\n",
    "\n",
    "from grape import Graph, GraphVisualizer\n",
    "from grape.embedders import Node2VecCBOWEnsmallen\n",
    "\n",
    "tqdm.pandas()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# Preprocessing  <a class=\"anchor\" id=\"pre-processing\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The aim of this section is defining the required file path to access the graph, the nodes and edges type lists."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fpath = \"./testRNA-KG/resources/knowledge_graphs/\"\n",
    "graphname = \"PheKnowLator_v3.1.1_full_instance_inverseRelations_OWLNETS\"\n",
    "graphext = \".nt\"\n",
    "\n",
    "graph_fname = fpath+graphname+graphext\n",
    "\n",
    "colnames = [\"subject\",\"predicate\",\"object\",\"unused\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fulldata = pd.read_csv(graph_fname,sep=' ',header=None, names=colnames)\n",
    "fulldata.drop([colnames[3]],axis=1,inplace=True) # remove the last column containing the dot symbol\n",
    "fulldata.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build the nodes dataframe\n",
    "This is extracted from the list of nodes in the graph file and is then integrated with the relevant object types as a new column named \"type\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_df = pd.DataFrame(set(fulldata[colnames[0]])|set(fulldata[colnames[2]]),columns=[\"name\"])\n",
    "nodes_df.dropna(inplace=True)\n",
    "nodes_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Number of nodes in ' + graph_fname + ': ' + str(len(nodes_df)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Full mapping for all node types in RNA-KG\n",
    "RNAonly = False # when false all nodes are considered otherwise only RNA nodes are selected\n",
    "\n",
    "def uri2ntype(uri: str)->Union[str,None]:\n",
    "    \n",
    "    retval = None\n",
    "    \n",
    "    # match regular expression for all RNA genes in ncbi format\n",
    "    retlist=re.split(r\"gene/[\\w\\-]+[?]\", uri)\n",
    "    if len(retlist) == 2:  # pattern matched therefore list item 1 contains the RNA type\n",
    "        value = retlist[1][:-1]\n",
    "        if value == \"others\":\n",
    "            retval = \"otherRNA\"\n",
    "        elif value == \"pseudo\":\n",
    "            retval = \"Pseudogene\"\n",
    "        elif value == \"unknown\":\n",
    "            retval = \"unknown RNA\"\n",
    "        else:\n",
    "            retval = value\n",
    "    # regular expressions didn't match -> continue with direct string matching\n",
    "    elif (\"https://www.mirbase.org/\" in uri): \n",
    "        retval = \"miRNA\"\n",
    "    elif (\"https://www.addgene.org/\" in uri):\n",
    "        retval = \"gRNA\"\n",
    "    elif (\"https://www.ncbi.nlm.nih.gov/nuccore/\" in uri):\n",
    "        retval = \"Viral RNA\"\n",
    "    elif (\"http://web.mit.edu/sirna/\" in uri):\n",
    "        retval = \"s(i/h)RNA\"\n",
    "    elif (\"https://hanlab.uth.edu/HeRA\" in uri): \n",
    "        retval = \"eRNA\"\n",
    "    elif (\"http://bigdata.ibp.ac.cn/piRBase\" in uri): \n",
    "        retval = \"piRNA\"\n",
    "    elif (\"http://scottgroup.med.usherbrooke.ca/snoDB\" in uri): \n",
    "        retval = \"rRNA\"\n",
    "    elif (\"tRNA\" in uri) or (\"trna\" in uri) or (\"TRNA\" in uri):\n",
    "        retval = \"tRNA\"\n",
    "    elif (\"tRF\" in uri) or (\"trf\" in uri):\n",
    "        retval = \"tRF\"\n",
    "    elif (\"tsRNA\" in uri): \n",
    "        retval = \"tsRNA\"\n",
    "    elif (\"https://go.drugbank.com/drugs/\" in uri): \n",
    "        retval = \"RNA drug\"\n",
    "    elif (\"https://eskip-finder.org\" in uri): \n",
    "        retval = \"ASO\"\n",
    "    elif (\"https://www.aptagen.com/aptamer-details\" in uri): \n",
    "        retval = \"Aptamer\"\n",
    "    elif (\"retained_intron\" in uri): \n",
    "        retval = \"Retained intron\"\n",
    "    elif (\"tbdb.io/tboxes/\" in uri) or (\"penchovsky\" in uri):\n",
    "        retval = \"Riboswitch\"\n",
    "    elif (\"http://rfamlive.xfam.org/\" in uri):\n",
    "        retval = \"Ribozyme\"\n",
    "    elif not RNAonly:    \n",
    "        if (\"http://purl.obolibrary.org/obo/MONDO\" in uri) or (\"purl.obolibrary.org/obo/DOID\" in uri) or (\"ghr.nlm.nih.gov/condition\" in uri) or (\"rarediseases.info.nih.gov/diseases\" in uri):\n",
    "            retval = \"Disease\"\n",
    "        elif (\"purl.obolibrary.org/obo/IDO\" in uri):\n",
    "            retval = \"Infectious disease\"\n",
    "        elif (\"purl.obolibrary.org/obo/MFOMD\" in uri):\n",
    "            retval = \"Mental disease\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/GO\" in uri):\n",
    "            retval = \"GO\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/CHR\" in uri):\n",
    "            retval = \"Chromosome\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/SO\" in uri):\n",
    "            retval = \"Sequence\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/VO\" in uri):\n",
    "            retval = \"Vaccine\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/CHEBI\" in uri): \n",
    "            retval = \"Chemical\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/PR\" in uri) or (\"http://purl.obolibrary.org/obo/vo/ontorat/PR\" in uri): \n",
    "            retval = \"Protein\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/PW\" in uri) or (\"https://reactome.org/content/detail/\" in uri): \n",
    "            retval = \"Pathway\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/VO\" in uri): \n",
    "            retval = \"Vaccine\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/FOODON\" in uri): \n",
    "            retval = \"Food\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/MF\" in uri): \n",
    "            retval = \"Mental functioning\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/OGMS\" in uri): \n",
    "            retval = \"General medical science\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/MAXO\" in uri): \n",
    "            retval = \"Medical action\"\n",
    "        elif (\"https://www.ncbi.nlm.nih.gov/snp/\" in uri):\n",
    "            retval = \"Variant (SNP)\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/NBO\" in uri):\n",
    "            retval = \"Neuro behaviour\"\n",
    "        elif (\"https://www.genome.gov/genetics-glossary/\" in uri):\n",
    "            retval = \"Biological role\"\n",
    "        elif  (\"http://purl.obolibrary.org/obo/CARO\" in uri) or (\"http://purl.obolibrary.org/obo/UBERON\" in uri) or (\"http://sig.uw.edu/fma\" in uri) or (\"http://purl.obolibrary.org/obo/FMA\" in uri): \n",
    "            retval = \"Anatomy\"  \n",
    "        elif  (\"http://purl.obolibrary.org/obo/NCIT\" in uri): \n",
    "            retval = \"NCI thesaurus\" \n",
    "        elif (\"http://purl.obolibrary.org/obo/FBbt\" in uri):\n",
    "            retval = \"Drosophila anatomy\" \n",
    "        elif (\"http://purl.obolibrary.org/obo/CL_\" in uri): \n",
    "            retval = \"Cell\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/CLO\" in uri) or (\"http://www.ebi.ac.uk/cellline\" in uri): \n",
    "            retval = \"Cell line\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/HP\" in uri) or (\"http://purl.obolibrary.org/obo/PATO\" in uri) or (\"http://purl.obolibrary.org/obo/UPHENO\" in uri): \n",
    "            retval = \"Phenotype\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/GNO\" in uri): \n",
    "            retval = \"Glycan\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/BFO\" in uri): \n",
    "            retval = \"Basic formal\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/ENVO\" in uri): \n",
    "            retval = \"Environment\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/ECTO\" in uri): \n",
    "            retval = \"Environmental exposure\"\n",
    "        elif (\"www.ncbi.nlm.nih.gov/gene\" in uri) or (\"http://purl.obolibrary.org/obo/OGG\" in uri):\n",
    "            retval = \"Gene\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/OGG\" in uri):\n",
    "            retval = \"Genome\"\n",
    "        elif (\"www.ncbi.nlm.nih.gov/Taxonomy/Browser\" in uri) or (\"purl.obolibrary.org/obo/NCBITaxon\" in uri): \n",
    "            retval = \"Species\"\n",
    "        elif (\"https://www.encodeproject.org/targets\" in uri): \n",
    "            retval = \"Epigenetic modification\"\n",
    "        elif (\"crdd.osdd.net/raghava/dbem?\" in uri): \n",
    "            retval = \"Histone modification\"\n",
    "        elif (\"bigdata.ibp.ac.cn/SmProt/SmProt.php?ID\" in uri): \n",
    "            retval = \"Small protein\"\n",
    "        elif (\"snomedct\" in uri) or (\"SNOMEDCT\" in uri): \n",
    "            retval = \"snomedct\"\n",
    "        elif (\"http://www.ebi.ac.uk/efo/EFO\" in uri): \n",
    "            retval = \"Experimental factor\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/HsapDv\" in uri): \n",
    "            retval = \"Human developmental stage\"\n",
    "        elif (\"http://www.w3.org/2002/07/owl#Nothing\" in uri): \n",
    "            retval = \"owlNothing\"\n",
    "        elif (\"http://purl.obolibrary.org/obo/\" in uri):    # all unmapped obo types are dealt with here\n",
    "            retlist = re.split(r\"obo/\",uri)\n",
    "            retval = retlist[1].split('_', 1)[0]\n",
    "            \n",
    "    else:\n",
    "        retval = None\n",
    "\n",
    "    return retval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "ntypes_list = []\n",
    "for u in tqdm(nodes_df[\"name\"].values):\n",
    "    nty = uri2ntype(u)\n",
    "    ntypes_list.append(nty)\n",
    "\n",
    "nodes_df.loc[:,\"type\"] = ntypes_list\n",
    "nodes_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_df['type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Unassigned node types:\", nodes_df.type.isna().sum())\n",
    "nodes_df.type.fillna('undefined',inplace=True)\n",
    "\n",
    "nodes_df = nodes_df.dropna()\n",
    "nodes_df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build the edges dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract from graph those codes which refer only to edges, maintaining their\n",
    "# id as from the graph numerical codes\n",
    "ety_df = pd.DataFrame(set(fulldata[colnames[1]]),columns=[\"name\"])\n",
    "#ety_df=fulluri[fulluri.id.isin(ecodes)]\n",
    "#ety_df.reset_index(drop=True,inplace=True)\n",
    "ety_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# split camel case strings, e.g., \"overexpressedIn\" --> \"overexpressed in\"\n",
    "def camel_case_split(identifier):\n",
    "    matches = re.finditer('.+?(?:(?<=[a-z])(?=[A-Z])|(?<=[A-Z])(?=[A-Z][a-z])|$)', identifier)\n",
    "    return [m.group(0) for m in matches]\n",
    "\n",
    "# automatically match all OBO items for the specified ontologies\n",
    "hdr = {'Accept': 'application/json'}\n",
    "ontos = [\"ro\",\"bspo\",\"vo\",\"clo\",\"mondo\",\"ogg\",\"cl\",\"mf\"]\n",
    "tomatch = \"http://purl.obolibrary.org/obo/\"\n",
    "\n",
    "def uri2etype(uri: str)->Union[str,None]:\n",
    "    label = None\n",
    "    for oy in ontos:\n",
    "        baseuri = f\"https://www.ebi.ac.uk/ols4/api/ontologies/{oy}/properties?iri={uri[1:-1]}\"\n",
    "        try:\n",
    "            res = requests.get(baseuri,hdr).json()\n",
    "            label=res['_embedded']['properties'][0]['label']\n",
    "            label=label.lower().capitalize()\n",
    "        except:\n",
    "            pass\n",
    "\n",
    "    if (\"http://www.w3.org/1999/02/22-rdf-syntax-ns#type\" in uri):   \n",
    "        label = \"Type\"  \n",
    "    elif (\"http://www.w3.org/2000/01/rdf-schema#subClassOf\" in uri):   \n",
    "        label = \"SubClass of\" \n",
    "    # new manually set edges by splitting over the # symbol of the uri\n",
    "    elif (\"http://semanticscience.org/resource/SIO_000420\" in uri):   \n",
    "        label = \"Has expression\" \n",
    "    elif (\"http://purl.obolibrary.org/obo/CLO_0054408\" in uri):   \n",
    "        label = \"Overexpresses gene\"\n",
    "    elif (\"http://purl.obolibrary.org/obo/CLO_0054409\" in uri):   \n",
    "        label = \"Adenoma formation induced by cell lineage cells in mice\"\n",
    "    elif (\"http://purl.obolibrary.org/obo/uberon/core\" in uri) or \\\n",
    "         (\"http://purl.obolibrary.org/obo/mondo\" in uri) or \\\n",
    "         (\"http://purl.obolibrary.org/obo/so\" in uri) or \\\n",
    "         (\"http://purl.obolibrary.org/obo/envo\" in uri) or \\\n",
    "         (\"http://purl.obolibrary.org/obo/pr\" in uri) or \\\n",
    "         (\"http://purl.obolibrary.org/obo/pato\" in uri) or \\\n",
    "         (\"http://purl.obolibrary.org/obo/pw\" in uri) or \\\n",
    "         (\"http://purl.obolibrary.org/obo/exo.obo\" in uri) or \\\n",
    "         (\"http://purl.obolibrary.org/obo/cl\" in uri) or \\\n",
    "         (\"http://purl.obolibrary.org/obo/nbo\" in uri) or \\\n",
    "         (\"http://purl.obolibrary.org/obo/MF\" in uri) or \\\n",
    "         (\"http://www.obofoundry.org/ro\" in uri) or \\\n",
    "         (\"http://purl.obolibrary.org/obo/chebi\" in uri):\n",
    "            label = uri[1:-1].split(\"#\")[1]\n",
    "            label = '_'.join(camel_case_split(label))\n",
    "            label = label.replace('_',' ').lower().capitalize()\n",
    "    else:\n",
    "        pass\n",
    "    \n",
    "    return label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "etypes_list = []\n",
    "for u in tqdm(ety_df[\"name\"].values):\n",
    "    ety = uri2etype(u)\n",
    "    etypes_list.append(ety)\n",
    "\n",
    "ety_df.loc[:,\"type\"] = etypes_list\n",
    "ety_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Unassigned edge types:\",ety_df.type.isna().sum())\n",
    "#ety_df.type.fillna('undefined',inplace=True)\n",
    "#ety_df = ety_df.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "# add the type column to the original graph structure\n",
    "edges_df = fulldata.copy()\n",
    "\n",
    "efmap = lambda x: ety_df.type[np.where(ety_df['name'].eq(x))[0][0]]\n",
    "edges_df[\"type\"] = edges_df[colnames[1]].progress_apply(efmap)\n",
    "edges_df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# Neo4j  <a class=\"anchor\" id=\"neo4j\"></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "properties = pd.read_csv('testRNA-KG/resources/knowledge_graphs/PheKnowLator_v3.1.1_full_instance_inverseRelations_OWLNETS_NodeLabels.txt',\n",
    "            sep='\\t')\n",
    "node_properties = properties[properties['entity_type'] == 'NODES']\n",
    "node_properties"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "neo4jnodes_df = nodes_df.copy()\n",
    "neo4jnodes_df = pd.merge(neo4jnodes_df, node_properties, left_on='name', right_on='entity_uri', how='outer')\n",
    "neo4jnodes_df = neo4jnodes_df[['name', 'type', 'label', 'description/definition', 'synonym']]\n",
    "neo4jnodes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "neo4jedges_df = ety_df.copy()\n",
    "edge_properties = properties[properties['entity_type'] == 'RELATIONS']\n",
    "neo4jedges_df = pd.merge(neo4jedges_df, edge_properties, left_on='name', right_on='entity_uri', how='outer')\n",
    "neo4jedges_df = neo4jedges_df[['name', 'type', 'description/definition', 'synonym']]\n",
    "neo4jedges_df = neo4jedges_df.rename(columns={'type':'label'})\n",
    "neo4jedges_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "neo4jedges_df = pd.merge(neo4jedges_df, edges_df, left_on='name', right_on='predicate', how='outer')\n",
    "neo4jedges_df = neo4jedges_df[['subject','predicate','object','label','description/definition','synonym']]\n",
    "neo4jedges_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "neo4jnodes_df.to_csv('nodes.csv', index=None)\n",
    "neo4jedges_df.to_csv('relationships.csv', index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# GRAPE  <a class=\"anchor\" id=\"grape\"></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "# load it into a graph\n",
    "graph = Graph.from_pd(\n",
    "    edges_df=edges_df,\n",
    "    nodes_df=nodes_df,\n",
    "    node_name_column=\"name\",\n",
    "    node_type_column=\"type\",\n",
    "    edge_src_column=\"subject\",\n",
    "    edge_dst_column=\"object\",\n",
    "    #edge_weight_column=\"weight\",\n",
    "    edge_type_column=\"type\",\n",
    "    node_types_separator=\"|\",\n",
    "    directed=True,\n",
    "    name=\"graph\",\n",
    ")\n",
    "\n",
    "graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "htmlrep = fpath+\"RNAgraphReport.html\"\n",
    "\n",
    "ff = open(htmlrep,\"w\")\n",
    "ff.write(str(graph))\n",
    "ff.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph.get_diameter()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph = Graph.from_pd(\n",
    "    edges_df=edges_df,\n",
    "    nodes_df=nodes_df,\n",
    "    node_name_column=\"name\",\n",
    "    node_type_column=\"type\",\n",
    "    edge_src_column=\"subject\",\n",
    "    edge_dst_column=\"object\",\n",
    "    #edge_weight_column=\"weight\",\n",
    "    edge_type_column=\"type\",\n",
    "    node_types_separator=\"|\",\n",
    "    directed=False,\n",
    "    name=\"graph\",\n",
    ")\n",
    "\n",
    "engine = Node2VecCBOWEnsmallen(walk_length=5)\n",
    "embedding = engine.fit_transform(graph)\n",
    "vis = GraphVisualizer(graph)\n",
    "\n",
    "vis.fit_edges(embedding)\n",
    "vis.plot_edge_types(k=9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vis.fit_nodes(embedding)\n",
    "vis.plot_node_types(k=9)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Predictions using Node2Vec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "train, test = graph.connected_holdout(train_size=0.7)\n",
    "train.enable()\n",
    "\n",
    "vis = GraphVisualizer(\n",
    "    graph=test,\n",
    "    support=graph\n",
    ")\n",
    "\n",
    "vis.fit_negative_and_positive_edges(embedding)\n",
    "vis.plot_positive_and_negative_edges()"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "authorship_tag": "ABX9TyO6tVy5b4NGs/oOGRYG4jak",
   "collapsed_sections": [],
   "include_colab_link": true,
   "name": "Graph generation.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
