{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# <p style=\"text-align: center;\">RNA Knowledge Graph Build Data Preparation</p>\n",
    "    \n",
    "***\n",
    "***\n",
    "\n",
    "**Authors:** [ECavalleri](https://mail.google.com/mail/u/0/?view=cm&fs=1&tf=1&to=emanuele.cavalleri@studenti.unimi.it), [TJCallahan](https://mail.google.com/mail/u/0/?view=cm&fs=1&tf=1&to=callahantiff@gmail.com), [MMesiti](https://mail.google.com/mail/u/0/?view=cm&fs=1&tf=1&to=marco.mesiti@unimi.it)\n",
    "\n",
    "**GitHub Repositories:** [testRNA-KG](https://github.com/emanuelecavalleri/testRNA-KG), [PheKnowLator](https://github.com/callahantiff/PheKnowLator/)  \n",
    "<!--- **Release:** **[v2.0.0](https://github.com/callahantiff/PheKnowLator/wiki/v2.0.0)** --->\n",
    "  \n",
    "<br>  \n",
    "  \n",
    "**Purpose:** This notebook serves as a script to download, process, map, and clean data in order to build edges for the RNA-based Knowledge Graph. For more information on the data sources utilize within this script, please see the [PheKnowLator Data Sources](https://github.com/callahantiff/PheKnowLator/wiki/v2-Data-Sources) Wiki page.\n",
    "\n",
    "<br>\n",
    "\n",
    "**Assumptions:**   \n",
    "- Edge data downloads ➞ `./resources/edge_data`  \n",
    "- Ontologies ➞ `./resources/ontologies`    \n",
    "- Processed data write location ➞ `./resources/processed_data`  \n",
    "\n",
    "<br>\n",
    "\n",
    "**Dependencies:**   \n",
    "- **Scripts**: This notebook utilizes several helper functions, which are stored in the [`data_utils.py`](https://github.com/callahantiff/PheKnowLator/blob/master/pkt_kg/utils/data_utils.py) and [`kg_utils.py`](https://github.com/callahantiff/PheKnowLator/blob/master/pkt_kg/utils/kg_utils.py) scripts.  \n",
    "- **Data**: All downloaded and generated data sources are provided through [this](https://drive.google.com/drive/folders/1sev5zczMviX7UVqMhTpkFXG43K3nQa9f) dedicated Google Drive repository. <u>This notebook will download everything that is needed for you</u>.  \n",
    "_____\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table of Contents\n",
    "***\n",
    "\n",
    "### [Download Ontologies](#create-ontologies)\n",
    "\n",
    "\n",
    "### [Create Identifier Maps ](#create-identifier-maps)   \n",
    "\n",
    "\n",
    "### [Download and process Edge Datasets](#create-edges)  \n",
    "\n",
    "____\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set-Up Environment\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: Cython>=0.29.14 in /opt/anaconda3/lib/python3.8/site-packages (from -r requirements.txt (line 1)) (0.29.21)\n",
      "Requirement already satisfied: ipywidgets>=7.7.0 in /opt/anaconda3/lib/python3.8/site-packages (from -r requirements.txt (line 2)) (8.0.2)\n",
      "Requirement already satisfied: matplotlib==3.3.2 in /opt/anaconda3/lib/python3.8/site-packages (from -r requirements.txt (line 3)) (3.3.2)\n",
      "Requirement already satisfied: more-itertools>=8.6.0 in /opt/anaconda3/lib/python3.8/site-packages (from -r requirements.txt (line 4)) (8.6.0)\n",
      "Requirement already satisfied: networkx>=2.4 in /opt/anaconda3/lib/python3.8/site-packages (from -r requirements.txt (line 5)) (2.5)\n",
      "Requirement already satisfied: numpy>=1.18.1 in /opt/anaconda3/lib/python3.8/site-packages (from -r requirements.txt (line 6)) (1.19.2)\n",
      "Requirement already satisfied: openpyxl>=3.0.3 in /opt/anaconda3/lib/python3.8/site-packages (from -r requirements.txt (line 7)) (3.0.5)\n",
      "Requirement already satisfied: pandas>=1.0.5 in /opt/anaconda3/lib/python3.8/site-packages (from -r requirements.txt (line 8)) (1.1.3)\n",
      "Requirement already satisfied: psutil>=5.6.3 in /opt/anaconda3/lib/python3.8/site-packages (from -r requirements.txt (line 9)) (5.7.2)\n",
      "Requirement already satisfied: python-json-logger>=2.0.1 in /opt/anaconda3/lib/python3.8/site-packages (from -r requirements.txt (line 10)) (2.0.4)\n",
      "Requirement already satisfied: pyvis==0.3.0 in /opt/anaconda3/lib/python3.8/site-packages (from -r requirements.txt (line 11)) (0.3.0)\n",
      "Requirement already satisfied: ray>=1.1.0 in /opt/anaconda3/lib/python3.8/site-packages (from -r requirements.txt (line 12)) (2.0.0)\n",
      "Requirement already satisfied: rdflib>=4.2.2 in /opt/anaconda3/lib/python3.8/site-packages (from -r requirements.txt (line 13)) (6.2.0)\n",
      "Requirement already satisfied: reactome2py>=0.0.8 in /opt/anaconda3/lib/python3.8/site-packages (from -r requirements.txt (line 14)) (3.0.0)\n",
      "Requirement already satisfied: requests>=2.22.0 in /opt/anaconda3/lib/python3.8/site-packages (from -r requirements.txt (line 15)) (2.24.0)\n",
      "Requirement already satisfied: responses==0.10.12 in /opt/anaconda3/lib/python3.8/site-packages (from -r requirements.txt (line 16)) (0.10.12)\n",
      "Requirement already satisfied: tqdm>=4.47.0 in /opt/anaconda3/lib/python3.8/site-packages (from -r requirements.txt (line 17)) (4.50.2)\n",
      "Requirement already satisfied: urllib3>=1.25.3 in /opt/anaconda3/lib/python3.8/site-packages (from -r requirements.txt (line 18)) (1.25.11)\n",
      "Requirement already satisfied: python-dateutil>=2.1 in /opt/anaconda3/lib/python3.8/site-packages (from matplotlib==3.3.2->-r requirements.txt (line 3)) (2.8.1)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /opt/anaconda3/lib/python3.8/site-packages (from matplotlib==3.3.2->-r requirements.txt (line 3)) (1.3.0)\n",
      "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.3 in /opt/anaconda3/lib/python3.8/site-packages (from matplotlib==3.3.2->-r requirements.txt (line 3)) (2.4.7)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /opt/anaconda3/lib/python3.8/site-packages (from matplotlib==3.3.2->-r requirements.txt (line 3)) (8.0.1)\n",
      "Requirement already satisfied: certifi>=2020.06.20 in /opt/anaconda3/lib/python3.8/site-packages (from matplotlib==3.3.2->-r requirements.txt (line 3)) (2020.6.20)\n",
      "Requirement already satisfied: cycler>=0.10 in /opt/anaconda3/lib/python3.8/site-packages (from matplotlib==3.3.2->-r requirements.txt (line 3)) (0.10.0)\n",
      "Requirement already satisfied: jinja2>=2.9.6 in /opt/anaconda3/lib/python3.8/site-packages (from pyvis==0.3.0->-r requirements.txt (line 11)) (2.11.2)\n",
      "Requirement already satisfied: jsonpickle>=1.4.1 in /opt/anaconda3/lib/python3.8/site-packages (from pyvis==0.3.0->-r requirements.txt (line 11)) (2.2.0)\n",
      "Requirement already satisfied: ipython>=5.3.0 in /opt/anaconda3/lib/python3.8/site-packages (from pyvis==0.3.0->-r requirements.txt (line 11)) (7.19.0)\n",
      "Requirement already satisfied: six in /opt/anaconda3/lib/python3.8/site-packages (from responses==0.10.12->-r requirements.txt (line 16)) (1.15.0)\n",
      "Requirement already satisfied: jupyterlab-widgets~=3.0 in /opt/anaconda3/lib/python3.8/site-packages (from ipywidgets>=7.7.0->-r requirements.txt (line 2)) (3.0.3)\n",
      "Requirement already satisfied: ipykernel>=4.5.1 in /opt/anaconda3/lib/python3.8/site-packages (from ipywidgets>=7.7.0->-r requirements.txt (line 2)) (5.3.4)\n",
      "Requirement already satisfied: widgetsnbextension~=4.0 in /opt/anaconda3/lib/python3.8/site-packages (from ipywidgets>=7.7.0->-r requirements.txt (line 2)) (4.0.3)\n",
      "Requirement already satisfied: traitlets>=4.3.1 in /opt/anaconda3/lib/python3.8/site-packages (from ipywidgets>=7.7.0->-r requirements.txt (line 2)) (5.0.5)\n",
      "Requirement already satisfied: decorator>=4.3.0 in /opt/anaconda3/lib/python3.8/site-packages (from networkx>=2.4->-r requirements.txt (line 5)) (4.4.2)\n",
      "Requirement already satisfied: et-xmlfile in /opt/anaconda3/lib/python3.8/site-packages (from openpyxl>=3.0.3->-r requirements.txt (line 7)) (1.0.1)\n",
      "Requirement already satisfied: jdcal in /opt/anaconda3/lib/python3.8/site-packages (from openpyxl>=3.0.3->-r requirements.txt (line 7)) (1.4.1)\n",
      "Requirement already satisfied: pytz>=2017.2 in /opt/anaconda3/lib/python3.8/site-packages (from pandas>=1.0.5->-r requirements.txt (line 8)) (2020.1)\n",
      "Requirement already satisfied: msgpack<2.0.0,>=1.0.0 in /opt/anaconda3/lib/python3.8/site-packages (from ray>=1.1.0->-r requirements.txt (line 12)) (1.0.0)\n",
      "Requirement already satisfied: click<=8.0.4,>=7.0 in /opt/anaconda3/lib/python3.8/site-packages (from ray>=1.1.0->-r requirements.txt (line 12)) (7.1.2)\n",
      "Requirement already satisfied: pyyaml in /opt/anaconda3/lib/python3.8/site-packages (from ray>=1.1.0->-r requirements.txt (line 12)) (5.3.1)\n",
      "Requirement already satisfied: aiosignal in /opt/anaconda3/lib/python3.8/site-packages (from ray>=1.1.0->-r requirements.txt (line 12)) (1.2.0)\n",
      "Requirement already satisfied: frozenlist in /opt/anaconda3/lib/python3.8/site-packages (from ray>=1.1.0->-r requirements.txt (line 12)) (1.3.1)\n",
      "Requirement already satisfied: protobuf<4.0.0,>=3.15.3 in /opt/anaconda3/lib/python3.8/site-packages (from ray>=1.1.0->-r requirements.txt (line 12)) (3.20.3)\n",
      "Requirement already satisfied: jsonschema in /opt/anaconda3/lib/python3.8/site-packages (from ray>=1.1.0->-r requirements.txt (line 12)) (3.2.0)\n",
      "Requirement already satisfied: virtualenv in /opt/anaconda3/lib/python3.8/site-packages (from ray>=1.1.0->-r requirements.txt (line 12)) (20.16.5)\n",
      "Requirement already satisfied: filelock in /opt/anaconda3/lib/python3.8/site-packages (from ray>=1.1.0->-r requirements.txt (line 12)) (3.8.0)\n",
      "Requirement already satisfied: grpcio<=1.43.0,>=1.28.1 in /opt/anaconda3/lib/python3.8/site-packages (from ray>=1.1.0->-r requirements.txt (line 12)) (1.43.0)\n",
      "Requirement already satisfied: attrs in /opt/anaconda3/lib/python3.8/site-packages (from ray>=1.1.0->-r requirements.txt (line 12)) (20.3.0)\n",
      "Requirement already satisfied: isodate in /opt/anaconda3/lib/python3.8/site-packages (from rdflib>=4.2.2->-r requirements.txt (line 13)) (0.6.1)\n",
      "Requirement already satisfied: setuptools in /opt/anaconda3/lib/python3.8/site-packages (from rdflib>=4.2.2->-r requirements.txt (line 13)) (50.3.1.post20201107)\n",
      "Requirement already satisfied: json5>=0.8.4 in /opt/anaconda3/lib/python3.8/site-packages (from reactome2py>=0.0.8->-r requirements.txt (line 14)) (0.9.5)\n",
      "Requirement already satisfied: chardet<4,>=3.0.2 in /opt/anaconda3/lib/python3.8/site-packages (from requests>=2.22.0->-r requirements.txt (line 15)) (3.0.4)\n",
      "Requirement already satisfied: idna<3,>=2.5 in /opt/anaconda3/lib/python3.8/site-packages (from requests>=2.22.0->-r requirements.txt (line 15)) (2.10)\n",
      "Requirement already satisfied: appnope in /opt/anaconda3/lib/python3.8/site-packages (from ipykernel>=4.5.1->ipywidgets>=7.7.0->-r requirements.txt (line 2)) (0.1.0)\n",
      "Requirement already satisfied: tornado>=4.2 in /opt/anaconda3/lib/python3.8/site-packages (from ipykernel>=4.5.1->ipywidgets>=7.7.0->-r requirements.txt (line 2)) (6.0.4)\n",
      "Requirement already satisfied: jupyter-client in /opt/anaconda3/lib/python3.8/site-packages (from ipykernel>=4.5.1->ipywidgets>=7.7.0->-r requirements.txt (line 2)) (6.1.7)\n",
      "Requirement already satisfied: pickleshare in /opt/anaconda3/lib/python3.8/site-packages (from ipython>=5.3.0->pyvis==0.3.0->-r requirements.txt (line 11)) (0.7.5)\n",
      "Requirement already satisfied: backcall in /opt/anaconda3/lib/python3.8/site-packages (from ipython>=5.3.0->pyvis==0.3.0->-r requirements.txt (line 11)) (0.2.0)\n",
      "Requirement already satisfied: pygments in /opt/anaconda3/lib/python3.8/site-packages (from ipython>=5.3.0->pyvis==0.3.0->-r requirements.txt (line 11)) (2.7.2)\n",
      "Requirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in /opt/anaconda3/lib/python3.8/site-packages (from ipython>=5.3.0->pyvis==0.3.0->-r requirements.txt (line 11)) (3.0.8)\n",
      "Requirement already satisfied: pexpect>4.3 in /opt/anaconda3/lib/python3.8/site-packages (from ipython>=5.3.0->pyvis==0.3.0->-r requirements.txt (line 11)) (4.8.0)\n",
      "Requirement already satisfied: jedi>=0.10 in /opt/anaconda3/lib/python3.8/site-packages (from ipython>=5.3.0->pyvis==0.3.0->-r requirements.txt (line 11)) (0.17.1)\n",
      "Requirement already satisfied: MarkupSafe>=0.23 in /opt/anaconda3/lib/python3.8/site-packages (from jinja2>=2.9.6->pyvis==0.3.0->-r requirements.txt (line 11)) (1.1.1)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: ipython-genutils in /opt/anaconda3/lib/python3.8/site-packages (from traitlets>=4.3.1->ipywidgets>=7.7.0->-r requirements.txt (line 2)) (0.2.0)\n",
      "Requirement already satisfied: pyrsistent>=0.14.0 in /opt/anaconda3/lib/python3.8/site-packages (from jsonschema->ray>=1.1.0->-r requirements.txt (line 12)) (0.17.3)\n",
      "Requirement already satisfied: distlib<1,>=0.3.5 in /opt/anaconda3/lib/python3.8/site-packages (from virtualenv->ray>=1.1.0->-r requirements.txt (line 12)) (0.3.6)\n",
      "Requirement already satisfied: platformdirs<3,>=2.4 in /opt/anaconda3/lib/python3.8/site-packages (from virtualenv->ray>=1.1.0->-r requirements.txt (line 12)) (2.5.2)\n",
      "Requirement already satisfied: parso<0.8.0,>=0.7.0 in /opt/anaconda3/lib/python3.8/site-packages (from jedi>=0.10->ipython>=5.3.0->pyvis==0.3.0->-r requirements.txt (line 11)) (0.7.0)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /opt/anaconda3/lib/python3.8/site-packages (from pexpect>4.3->ipython>=5.3.0->pyvis==0.3.0->-r requirements.txt (line 11)) (0.6.0)\n",
      "Requirement already satisfied: wcwidth in /opt/anaconda3/lib/python3.8/site-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->ipython>=5.3.0->pyvis==0.3.0->-r requirements.txt (line 11)) (0.2.5)\n",
      "Requirement already satisfied: jupyter-core>=4.6.0 in /opt/anaconda3/lib/python3.8/site-packages (from jupyter-client->ipykernel>=4.5.1->ipywidgets>=7.7.0->-r requirements.txt (line 2)) (4.6.3)\n",
      "Requirement already satisfied: pyzmq>=13 in /opt/anaconda3/lib/python3.8/site-packages (from jupyter-client->ipykernel>=4.5.1->ipywidgets>=7.7.0->-r requirements.txt (line 2)) (19.0.2)\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip available: \u001b[0m\u001b[31;49m22.3.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.0.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49m/opt/anaconda3/bin/python -m pip install --upgrade pip\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "!{sys.executable} -m pip install -r requirements.txt\n",
    "sys.path.append('../')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TODO: Next run, I'll check if all these libraries are needed here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import needed libraries\n",
    "import datetime\n",
    "import glob\n",
    "import itertools\n",
    "import networkx\n",
    "import numpy\n",
    "import os\n",
    "import openpyxl\n",
    "import pandas\n",
    "import pickle\n",
    "import re\n",
    "import requests\n",
    "import tarfile\n",
    "import pandas as pd\n",
    "\n",
    "from collections import Counter\n",
    "from functools import reduce\n",
    "from rdflib import Graph, Namespace, URIRef, BNode, Literal\n",
    "from rdflib.namespace import OWL, RDF, RDFS\n",
    "from reactome2py import content\n",
    "from tqdm import tqdm\n",
    "from typing import Dict\n",
    "from typing import Tuple\n",
    "\n",
    "from pkt_kg.utils import * \n",
    "from builds.ontology_cleaning import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Define Global Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# directory to store resources\n",
    "resource_data_location = '../resources/'\n",
    "\n",
    "# directory to use for processing data\n",
    "processed_data_location = '../resources/processed_data/'\n",
    "\n",
    "# directory to write ontology data to\n",
    "ontology_data_location = '../resources/ontologies/'\n",
    "\n",
    "# directory to write edges data to\n",
    "edge_data_location = '../resources/edge_data/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download data function for already processed data\n",
    "def download(name, path):\n",
    "    url = 'https://storage.googleapis.com/pheknowlator/current_build/data/processed_data/'+name\n",
    "    if not os.path.exists(path + name):\n",
    "        data_downloader(url, path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('84103', ['SO_0001217']),\n",
       " ('84690', ['SO_0001217']),\n",
       " ('3579', ['SO_0001217']),\n",
       " ('54514', ['SO_0001217']),\n",
       " ('7159', ['SO_0001217'])]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# KG construction approach dictionary (for non-Ontology data), provided by PKL ecosystem\n",
    "download('subclass_construction_map.pkl', '../resources/construction_approach/')\n",
    "\n",
    "# Load data, print row count, and preview it\n",
    "nonO_data = pd.read_pickle(r'../resources/construction_approach/'+'subclass_construction_map.pkl')\n",
    "\n",
    "# For instance, ncbi IDs are mapped to appropriate SO Ontology entries\n",
    "list(nonO_data.items())[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 667 RO Relations and Labels\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Label</th>\n",
       "      <th>Relation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>helper property (not for use in curation)</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_0002464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>developmentally replaces</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_0002285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>is approximately equivalent to</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_0002603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>has intracellular endoparasite</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_0002641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>supplies</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_0002178</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       Label  \\\n",
       "0  helper property (not for use in curation)   \n",
       "1                   developmentally replaces   \n",
       "2             is approximately equivalent to   \n",
       "3             has intracellular endoparasite   \n",
       "4                                   supplies   \n",
       "\n",
       "                                    Relation  \n",
       "0  http://purl.obolibrary.org/obo/RO_0002464  \n",
       "1  http://purl.obolibrary.org/obo/RO_0002285  \n",
       "2  http://purl.obolibrary.org/obo/RO_0002603  \n",
       "3  http://purl.obolibrary.org/obo/RO_0002641  \n",
       "4  http://purl.obolibrary.org/obo/RO_0002178  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Relations labels are already provided by PKL ecosystem\n",
    "download('RELATIONS_LABELS.txt', '../resources/relations_data/')\n",
    "\n",
    "# Load data, print row count, and preview it\n",
    "ro_data_label = pandas.read_csv('../resources/relations_data/'+'RELATIONS_LABELS.txt', header=0, delimiter='\\t')\n",
    "\n",
    "print('There are {edge_count} RO Relations and Labels'.format(edge_count=len(ro_data_label)))\n",
    "ro_data_label.head(n=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "***\n",
    "### DOWNLOAD ONTOLOGIES  <a class=\"anchor\" id=\"create-ontologies\"></a>\n",
    "***\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Relation Ontology ([RO](https://www.ebi.ac.uk/ols/ontologies/ro))\n",
    "The OBO Relations Ontology (RO) is a collection of OWL relations (ObjectProperties) intended for use across a wide variety of biological ontologies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(ontology_data_location + 'ro_with_imports.owl'):\n",
    "    command = '{} {} --merge-import-closure -o {}'\n",
    "    os.system(command.format(owltools_location, 'http://purl.obolibrary.org/obo/ro.owl',\n",
    "                             ontology_data_location + 'ro_with_imports.owl'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Mondo Disease Ontology ([MONDO](https://www.ebi.ac.uk/ols/ontologies/ro))\n",
    "A semi-automatically constructed ontology that merges in multiple disease resources to yield a coherent merged ontology."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(ontology_data_location + 'mondo_with_imports.owl'):\n",
    "    command = '{} {} --merge-import-closure -o {}'\n",
    "    os.system(command.format(owltools_location, 'http://purl.obolibrary.org/obo/mondo.owl',\n",
    "                             ontology_data_location + 'mondo_with_imports.owl'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Non-Coding RNA Ontology ([NCRO](https://www.ebi.ac.uk/ols/ontologies/ro))\n",
    "An ontology for non-coding RNA, both of biological origin, and engineered."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(ontology_data_location + 'ncro_with_imports.owl'):\n",
    "    command = '{} {} --merge-import-closure -o {}'\n",
    "    os.system(command.format(owltools_location, 'http://purl.obolibrary.org/obo/ncro.owl',\n",
    "                             ontology_data_location + 'ncro_with_imports.owl'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "***\n",
    "### CREATE MAPPING DATASETS  <a class=\"anchor\" id=\"create-identifier-maps\"></a>\n",
    "***\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ensembl Gene-Entrez Gene <a class=\"anchor\" id=\"ensemblgene-entrezgene\"></a>\n",
    "\n",
    "\n",
    "**Purpose:** To map Ensembl gene identifiers to Entrez gene identifiers\n",
    "\n",
    "**Output:** `ENSEMBL_GENE_ENTREZ_GENE_MAP.txt`\n",
    "\n",
    "Already provided by PKL ecosystem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000171241</td>\n",
       "      <td>79801</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000131149</td>\n",
       "      <td>23199</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000096092</td>\n",
       "      <td>28978</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000222691</td>\n",
       "      <td>106479891</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>not protein-coding</td>\n",
       "      <td>not protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000230052</td>\n",
       "      <td>100873180</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>not protein-coding</td>\n",
       "      <td>not protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42283</th>\n",
       "      <td>ENSG00000175699</td>\n",
       "      <td>256369</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42284</th>\n",
       "      <td>ENSG00000251308</td>\n",
       "      <td>359776</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>not protein-coding</td>\n",
       "      <td>not protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42285</th>\n",
       "      <td>ENSG00000108479</td>\n",
       "      <td>2584</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42286</th>\n",
       "      <td>ENSG00000167371</td>\n",
       "      <td>112476</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42287</th>\n",
       "      <td>ENSG00000228938</td>\n",
       "      <td>119358</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>not protein-coding</td>\n",
       "      <td>not protein-coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>42288 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     0          1                       2               3  \\\n",
       "0      ENSG00000171241      79801          protein-coding  protein-coding   \n",
       "1      ENSG00000131149      23199          protein-coding  protein-coding   \n",
       "2      ENSG00000096092      28978          protein-coding  protein-coding   \n",
       "3      ENSG00000222691  106479891                   snRNA      pseudogene   \n",
       "4      ENSG00000230052  100873180  unprocessed_pseudogene      pseudogene   \n",
       "...                ...        ...                     ...             ...   \n",
       "42283  ENSG00000175699     256369          protein-coding  protein-coding   \n",
       "42284  ENSG00000251308     359776    processed_pseudogene      pseudogene   \n",
       "42285  ENSG00000108479       2584          protein-coding  protein-coding   \n",
       "42286  ENSG00000167371     112476          protein-coding  protein-coding   \n",
       "42287  ENSG00000228938     119358    processed_pseudogene      pseudogene   \n",
       "\n",
       "                        4                   5  \n",
       "0          protein-coding      protein-coding  \n",
       "1          protein-coding      protein-coding  \n",
       "2          protein-coding      protein-coding  \n",
       "3      not protein-coding  not protein-coding  \n",
       "4      not protein-coding  not protein-coding  \n",
       "...                   ...                 ...  \n",
       "42283      protein-coding      protein-coding  \n",
       "42284  not protein-coding  not protein-coding  \n",
       "42285      protein-coding      protein-coding  \n",
       "42286      protein-coding      protein-coding  \n",
       "42287  not protein-coding  not protein-coding  \n",
       "\n",
       "[42288 rows x 6 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "download('ENSEMBL_GENE_ENTREZ_GENE_MAP.txt', processed_data_location)\n",
    "\n",
    "ensEntrez = pd.read_csv(processed_data_location + 'ENSEMBL_GENE_ENTREZ_GENE_MAP.txt', sep=\"\\t\", header=None)\n",
    "ensEntrez"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Disease Ontology (DO) - MONDO mapping <a class=\"anchor\" id=\"ensemblgene-entrezgene\"></a>\n",
    "\n",
    "\n",
    "**Purpose:** To map DO identifiers to MONDO identifiers\n",
    "\n",
    "**Output:** `DISEASE_DOID_MONDO_Map.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mondo_graph = Graph().parse(ontology_data_location + 'mondo_with_imports.owl')\n",
    "\n",
    "dbxref_res = gets_ontology_class_dbxrefs(mondo_graph)[0]\n",
    "\n",
    "# Fix DOIDs (substitute : with _) and upper case them\n",
    "mondo_dict = {str(k).replace(':','_').upper(): {str(i).split('/')[-1].replace(':','_') for i in v}\n",
    "              for k, v in dbxref_res.items() if 'doid' in str(k)}\n",
    "\n",
    "with open(processed_data_location + 'DISEASE_DOID_MONDO_Map.txt', 'w') as outfile:\n",
    "    for k, v in mondo_dict.items():\n",
    "        outfile.write(str(k) + '\\t' + str(v).replace('{','').replace('\\'','').replace('}','') + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DOID_0060503</td>\n",
       "      <td>MONDO_0000778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DOID_0111559</td>\n",
       "      <td>MONDO_0032728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DOID_3635</td>\n",
       "      <td>MONDO_0018940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DOID_0110321</td>\n",
       "      <td>MONDO_0013200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DOID_10444</td>\n",
       "      <td>MONDO_0001037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9786</th>\n",
       "      <td>DOID_0111153</td>\n",
       "      <td>MONDO_0016558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9787</th>\n",
       "      <td>DOID_13404</td>\n",
       "      <td>MONDO_0007011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9788</th>\n",
       "      <td>DOID_0111341</td>\n",
       "      <td>MONDO_0007434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9789</th>\n",
       "      <td>DOID_14271</td>\n",
       "      <td>MONDO_0001930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9790</th>\n",
       "      <td>DOID_8464</td>\n",
       "      <td>MONDO_0004578</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9791 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 0              1\n",
       "0     DOID_0060503  MONDO_0000778\n",
       "1     DOID_0111559  MONDO_0032728\n",
       "2        DOID_3635  MONDO_0018940\n",
       "3     DOID_0110321  MONDO_0013200\n",
       "4       DOID_10444  MONDO_0001037\n",
       "...            ...            ...\n",
       "9786  DOID_0111153  MONDO_0016558\n",
       "9787    DOID_13404  MONDO_0007011\n",
       "9788  DOID_0111341  MONDO_0007434\n",
       "9789    DOID_14271  MONDO_0001930\n",
       "9790     DOID_8464  MONDO_0004578\n",
       "\n",
       "[9791 rows x 2 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "doidMondo = pd.read_csv(processed_data_location + 'DISEASE_DOID_MONDO_Map.txt', sep=\"\\t\", header=None)\n",
    "doidMondo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Disease description from DO - DO mapping <a class=\"anchor\" id=\"ensemblgene-entrezgene\"></a>\n",
    "\n",
    "\n",
    "**Purpose:** To map Disease descriptions from DO to DO identifiers\n",
    "\n",
    "**Output:** None, this mapping will be used only internally.\n",
    "\n",
    "Provided by [mir2Disease](http://watson.compbio.iupui.edu:8080/miR2Disease/)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Data from http://watson.compbio.iupui.edu:8080/miR2Disease/download/diseaseList.txt\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease name in original paper</th>\n",
       "      <th>disease ontology ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Abdominal Aortic Aneurysm</td>\n",
       "      <td>DOID:7693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>acute lymphoblastic leukemia (ALL)</td>\n",
       "      <td>DOID:9952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>acute myeloid leukemia (AML)</td>\n",
       "      <td>DOID:9119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>acute myocardial infarction</td>\n",
       "      <td>DOID:9408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>acute promyelocytic leukemia (APL)</td>\n",
       "      <td>DOID:9119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>uterine leiomyoma (ULM)</td>\n",
       "      <td>DOID:13223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>uveal melanoma</td>\n",
       "      <td>DOID:1909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>vascular disease</td>\n",
       "      <td>DOID:178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>vesicular stomatitis</td>\n",
       "      <td>DOID:10881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>Waldenstrom Macroglobulinemia (WM)</td>\n",
       "      <td>DOID:9080</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>174 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         disease name in original paper disease ontology ID\n",
       "0             Abdominal Aortic Aneurysm           DOID:7693\n",
       "1    acute lymphoblastic leukemia (ALL)           DOID:9952\n",
       "2          acute myeloid leukemia (AML)           DOID:9119\n",
       "3           acute myocardial infarction           DOID:9408\n",
       "4    acute promyelocytic leukemia (APL)           DOID:9119\n",
       "..                                  ...                 ...\n",
       "169             uterine leiomyoma (ULM)          DOID:13223\n",
       "170                      uveal melanoma           DOID:1909\n",
       "171                    vascular disease            DOID:178\n",
       "172                vesicular stomatitis          DOID:10881\n",
       "173  Waldenstrom Macroglobulinemia (WM)           DOID:9080\n",
       "\n",
       "[174 rows x 2 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_downloader('http://watson.compbio.iupui.edu:8080/miR2Disease/download/diseaseList.txt', processed_data_location)\n",
    " \n",
    "descDOmap = pd.read_csv(processed_data_location + 'diseaseList.txt', sep=\"\\t\")\n",
    "descDOmap"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then fix it to be appropriate used by PKL ecosystem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>desc</th>\n",
       "      <th>doid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>abdominal aortic aneurysm</td>\n",
       "      <td>DOID_7693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>acute lymphoblastic leukemia (all)</td>\n",
       "      <td>DOID_9952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>acute myeloid leukemia (aml)</td>\n",
       "      <td>DOID_9119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>acute myocardial infarction</td>\n",
       "      <td>DOID_9408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>acute promyelocytic leukemia (apl)</td>\n",
       "      <td>DOID_9119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>uterine leiomyoma (ulm)</td>\n",
       "      <td>DOID_13223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>uveal melanoma</td>\n",
       "      <td>DOID_1909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>vascular disease</td>\n",
       "      <td>DOID_178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>vesicular stomatitis</td>\n",
       "      <td>DOID_10881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>waldenstrom macroglobulinemia (wm)</td>\n",
       "      <td>DOID_9080</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>174 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   desc        doid\n",
       "0             abdominal aortic aneurysm   DOID_7693\n",
       "1    acute lymphoblastic leukemia (all)   DOID_9952\n",
       "2          acute myeloid leukemia (aml)   DOID_9119\n",
       "3           acute myocardial infarction   DOID_9408\n",
       "4    acute promyelocytic leukemia (apl)   DOID_9119\n",
       "..                                  ...         ...\n",
       "169             uterine leiomyoma (ulm)  DOID_13223\n",
       "170                      uveal melanoma   DOID_1909\n",
       "171                    vascular disease    DOID_178\n",
       "172                vesicular stomatitis  DOID_10881\n",
       "173  waldenstrom macroglobulinemia (wm)   DOID_9080\n",
       "\n",
       "[174 rows x 2 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "descDOmap.columns = ['desc', 'doid']\n",
    "descDOmap['desc'] = descDOmap['desc'].str.lower()\n",
    "descDOmap['doid'] = descDOmap['doid'].str.upper()\n",
    "descDOmap['doid'] = descDOmap['doid'].str.replace(':', '_')\n",
    "descDOmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove original file\n",
    "os.remove(processed_data_location + 'diseaseList.txt')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### miRBase ID - Non-Coding RNA Ontology (NCRO) mapping\n",
    "\n",
    "\n",
    "**Purpose:** To map miRNA identifiers from miRBase to NCRO identifiers\n",
    "\n",
    "**Output:** `MIRBASE_ID_NCRO_MAP.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helper function to get dbxrefs for all ontology classes' label\n",
    "def gets_ontology_class_label(graph: Graph) -> Tuple:\n",
    "    dbx_uris: Dict = dict()\n",
    "    dbx = [x for x in graph if 'label' in str(x[1]).lower() if isinstance(x[0], URIRef)]\n",
    "    for x in dbx:\n",
    "        if str(x[2]).lower() in dbx_uris.keys(): dbx_uris[str(x[2]).lower()].append(str(x[0]))\n",
    "        else: dbx_uris[str(x[2]).lower()] = [str(x[0])]\n",
    "    dbx_type = {str(x[2]).lower(): 'DbXref' for x in dbx}\n",
    "\n",
    "    ex_uris: Dict = dict()\n",
    "    ex = [x for x in graph if 'exactmatch' in str(x[1]).lower() if isinstance([0], URIRef)]\n",
    "    for x in ex:\n",
    "        if str(x[2]).lower() in ex_uris.keys(): ex_uris[str(x[2]).lower()].append(str(x[0]))\n",
    "        else: ex_uris[str(x[2]).lower()] = [str(x[0])]\n",
    "    ex_type = {str(x[2]).lower(): 'ExactMatch' for x in ex}\n",
    "\n",
    "    return {**dbx_uris, **ex_uris}, {**dbx_type, **ex_type}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read data into RDFLib graph object\n",
    "ncro_graph = Graph().parse(ontology_data_location + 'ncro_with_imports.owl')\n",
    "\n",
    "dbxref_res = gets_ontology_class_label(ncro_graph)[0]\n",
    "\n",
    "# Fix string patterns\n",
    "ncro_dict = {str(k): {str(i).split('/')[-1].replace(':','_') for i in v}\n",
    "             for k, v in dbxref_res.items() if 'NCRO' in str(v) and 'mir-' in str(k) and 'hsa' in str(k)}\n",
    "ncro_dict2 = {'hsa-'+str(k): {str(i).split('/')[-1].replace(':','_') for i in v} for k, v in dbxref_res.items()\n",
    "              if 'NCRO' in str(v) and 'mir-' in str(k) and 'hsa' not in str(k)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(processed_data_location + 'MIRBASE_ID_NCRO_MAP.txt', 'w') as outfile:\n",
    "    for k, v in {**ncro_dict, **ncro_dict2}.items():\n",
    "        outfile.write(str(k) + '\\t' + str(v).replace('{','').replace('\\'','').replace('}','') + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-mir-1302-8</td>\n",
       "      <td>NCRO_0002185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-mir-3617</td>\n",
       "      <td>NCRO_0003197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-mir-4539</td>\n",
       "      <td>NCRO_0001132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-mir-210</td>\n",
       "      <td>NCRO_0002758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-mir-6895</td>\n",
       "      <td>NCRO_0001579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014</th>\n",
       "      <td>hsa-mir-9</td>\n",
       "      <td>NCRO_0002701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2015</th>\n",
       "      <td>hsa-mir-1302</td>\n",
       "      <td>NCRO_0002885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2016</th>\n",
       "      <td>hsa-mir-3180</td>\n",
       "      <td>NCRO_0003084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2017</th>\n",
       "      <td>hsa-mir-10</td>\n",
       "      <td>NCRO_0002713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2018</th>\n",
       "      <td>hsa-mir-194</td>\n",
       "      <td>NCRO_0002729</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2019 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0             1\n",
       "0     hsa-mir-1302-8  NCRO_0002185\n",
       "1       hsa-mir-3617  NCRO_0003197\n",
       "2       hsa-mir-4539  NCRO_0001132\n",
       "3        hsa-mir-210  NCRO_0002758\n",
       "4       hsa-mir-6895  NCRO_0001579\n",
       "...              ...           ...\n",
       "2014       hsa-mir-9  NCRO_0002701\n",
       "2015    hsa-mir-1302  NCRO_0002885\n",
       "2016    hsa-mir-3180  NCRO_0003084\n",
       "2017      hsa-mir-10  NCRO_0002713\n",
       "2018     hsa-mir-194  NCRO_0002729\n",
       "\n",
       "[2019 rows x 2 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mirbaseNcro = pd.read_csv(processed_data_location + 'MIRBASE_ID_NCRO_MAP.txt', sep=\"\\t\", header=None)\n",
    "mirbaseNcro"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "***\n",
    "### DOWNLOAD EDGES  <a class=\"anchor\" id=\"create-edges\"></a>\n",
    "***\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### gene-disease from PKL itself"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Data from https://storage.googleapis.com/pheknowlator/current_build/data/original_data/curated_gene_disease_associations.tsv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneId</th>\n",
       "      <th>geneSymbol</th>\n",
       "      <th>DSI</th>\n",
       "      <th>DPI</th>\n",
       "      <th>diseaseId</th>\n",
       "      <th>diseaseName</th>\n",
       "      <th>diseaseType</th>\n",
       "      <th>diseaseClass</th>\n",
       "      <th>diseaseSemanticType</th>\n",
       "      <th>score</th>\n",
       "      <th>EI</th>\n",
       "      <th>YearInitial</th>\n",
       "      <th>YearFinal</th>\n",
       "      <th>NofPmids</th>\n",
       "      <th>NofSnps</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.700</td>\n",
       "      <td>0.538</td>\n",
       "      <td>C0019209</td>\n",
       "      <td>Hepatomegaly</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>C23;C06</td>\n",
       "      <td>Finding</td>\n",
       "      <td>0.30</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.700</td>\n",
       "      <td>0.538</td>\n",
       "      <td>C0036341</td>\n",
       "      <td>Schizophrenia</td>\n",
       "      <td>disease</td>\n",
       "      <td>F03</td>\n",
       "      <td>Mental or Behavioral Dysfunction</td>\n",
       "      <td>0.30</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>0.529</td>\n",
       "      <td>0.769</td>\n",
       "      <td>C0002395</td>\n",
       "      <td>Alzheimer's Disease</td>\n",
       "      <td>disease</td>\n",
       "      <td>C10;F03</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.769</td>\n",
       "      <td>1998.0</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>0.529</td>\n",
       "      <td>0.769</td>\n",
       "      <td>C0007102</td>\n",
       "      <td>Malignant tumor of colon</td>\n",
       "      <td>disease</td>\n",
       "      <td>C06;C04</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.31</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>0.529</td>\n",
       "      <td>0.769</td>\n",
       "      <td>C0009375</td>\n",
       "      <td>Colonic Neoplasms</td>\n",
       "      <td>group</td>\n",
       "      <td>C06;C04</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.30</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84033</th>\n",
       "      <td>109580095</td>\n",
       "      <td>HBB-LCR</td>\n",
       "      <td>0.743</td>\n",
       "      <td>0.115</td>\n",
       "      <td>C0002875</td>\n",
       "      <td>Cooley's anemia</td>\n",
       "      <td>disease</td>\n",
       "      <td>C16;C15</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84034</th>\n",
       "      <td>109580095</td>\n",
       "      <td>HBB-LCR</td>\n",
       "      <td>0.743</td>\n",
       "      <td>0.115</td>\n",
       "      <td>C0005283</td>\n",
       "      <td>beta Thalassemia</td>\n",
       "      <td>disease</td>\n",
       "      <td>C16;C15</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84035</th>\n",
       "      <td>109580095</td>\n",
       "      <td>HBB-LCR</td>\n",
       "      <td>0.743</td>\n",
       "      <td>0.115</td>\n",
       "      <td>C0019025</td>\n",
       "      <td>Hemoglobin F Disease</td>\n",
       "      <td>disease</td>\n",
       "      <td>C16;C15</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84036</th>\n",
       "      <td>109580095</td>\n",
       "      <td>HBB-LCR</td>\n",
       "      <td>0.743</td>\n",
       "      <td>0.115</td>\n",
       "      <td>C0085578</td>\n",
       "      <td>Thalassemia Minor</td>\n",
       "      <td>disease</td>\n",
       "      <td>C16;C15</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84037</th>\n",
       "      <td>109580095</td>\n",
       "      <td>HBB-LCR</td>\n",
       "      <td>0.743</td>\n",
       "      <td>0.115</td>\n",
       "      <td>C0271979</td>\n",
       "      <td>Thalassemia Intermedia</td>\n",
       "      <td>disease</td>\n",
       "      <td>C16;C15</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>84038 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          geneId geneSymbol    DSI    DPI diseaseId               diseaseName  \\\n",
       "0              1       A1BG  0.700  0.538  C0019209              Hepatomegaly   \n",
       "1              1       A1BG  0.700  0.538  C0036341             Schizophrenia   \n",
       "2              2        A2M  0.529  0.769  C0002395       Alzheimer's Disease   \n",
       "3              2        A2M  0.529  0.769  C0007102  Malignant tumor of colon   \n",
       "4              2        A2M  0.529  0.769  C0009375         Colonic Neoplasms   \n",
       "...          ...        ...    ...    ...       ...                       ...   \n",
       "84033  109580095    HBB-LCR  0.743  0.115  C0002875           Cooley's anemia   \n",
       "84034  109580095    HBB-LCR  0.743  0.115  C0005283          beta Thalassemia   \n",
       "84035  109580095    HBB-LCR  0.743  0.115  C0019025      Hemoglobin F Disease   \n",
       "84036  109580095    HBB-LCR  0.743  0.115  C0085578         Thalassemia Minor   \n",
       "84037  109580095    HBB-LCR  0.743  0.115  C0271979    Thalassemia Intermedia   \n",
       "\n",
       "      diseaseType diseaseClass               diseaseSemanticType  score  \\\n",
       "0       phenotype      C23;C06                           Finding   0.30   \n",
       "1         disease          F03  Mental or Behavioral Dysfunction   0.30   \n",
       "2         disease      C10;F03               Disease or Syndrome   0.50   \n",
       "3         disease      C06;C04                Neoplastic Process   0.31   \n",
       "4           group      C06;C04                Neoplastic Process   0.30   \n",
       "...           ...          ...                               ...    ...   \n",
       "84033     disease      C16;C15               Disease or Syndrome   0.30   \n",
       "84034     disease      C16;C15               Disease or Syndrome   0.30   \n",
       "84035     disease      C16;C15               Disease or Syndrome   0.30   \n",
       "84036     disease      C16;C15               Disease or Syndrome   0.30   \n",
       "84037     disease      C16;C15               Disease or Syndrome   0.30   \n",
       "\n",
       "          EI  YearInitial  YearFinal  NofPmids  NofSnps     source  \n",
       "0      1.000       2017.0     2017.0         1        0  CTD_human  \n",
       "1      1.000       2015.0     2015.0         1        0  CTD_human  \n",
       "2      0.769       1998.0     2018.0         3        0  CTD_human  \n",
       "3      1.000       2004.0     2019.0         1        0  CTD_human  \n",
       "4      1.000       2004.0     2004.0         1        0  CTD_human  \n",
       "...      ...          ...        ...       ...      ...        ...  \n",
       "84033    NaN          NaN        NaN         0        0  CTD_human  \n",
       "84034    NaN          NaN        NaN         0        0  CTD_human  \n",
       "84035    NaN          NaN        NaN         0        0  CTD_human  \n",
       "84036    NaN          NaN        NaN         0        0  CTD_human  \n",
       "84037    NaN          NaN        NaN         0        0  CTD_human  \n",
       "\n",
       "[84038 rows x 16 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_downloader('https://storage.googleapis.com/pheknowlator/current_build/data/original_data/curated_gene_disease_associations.tsv',\n",
    "                edge_data_location)\n",
    "\n",
    "# Rename file adding relationship's identifier\n",
    "os.rename(edge_data_location+'curated_gene_disease_associations.tsv',\n",
    "          edge_data_location+'gene-disease_curated_gene_disease_associations.tsv')\n",
    "\n",
    "with open(edge_data_location + 'gene-disease_curated_gene_disease_associations.tsv') as f:\n",
    "    data = f.read()\n",
    "\n",
    "data = pd.read_csv(edge_data_location + 'gene-disease_curated_gene_disease_associations.tsv', sep=\"\\t\")  \n",
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### gene-miRNA from [TarBase](https://dianalab.e-ce.uth.gr/html/diana/web/index.php?r=tarbasev8/index)\n",
    "DIANA-TarBase v8 is a reference database devoted to the indexing of experimentally supported microRNA (miRNA) targets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Gzipped Data from https://dianalab.e-ce.uth.gr/downloads/tarbase_v8_data.tar.gz\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneId</th>\n",
       "      <th>geneName</th>\n",
       "      <th>mirna</th>\n",
       "      <th>species</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>tissue</th>\n",
       "      <th>category</th>\n",
       "      <th>method</th>\n",
       "      <th>positive_negative</th>\n",
       "      <th>direct_indirect</th>\n",
       "      <th>up_down</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0910001A06Rik(mmu)</td>\n",
       "      <td>0910001A06Rik(mmu)</td>\n",
       "      <td>mmu-miR-124-3p</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>UP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1200004M23Rik(mmu)</td>\n",
       "      <td>1200004M23Rik(mmu)</td>\n",
       "      <td>mmu-miR-124-3p</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1700027J05Rik(mmu)</td>\n",
       "      <td>1700027J05Rik(mmu)</td>\n",
       "      <td>mmu-miR-124-3p</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>UP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1810015A11Rik(mmu)</td>\n",
       "      <td>1810015A11Rik(mmu)</td>\n",
       "      <td>mmu-miR-124-3p</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2310047A01Rik(mmu)</td>\n",
       "      <td>2310047A01Rik(mmu)</td>\n",
       "      <td>mmu-miR-124-3p</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927114</th>\n",
       "      <td>uPA(hsa)</td>\n",
       "      <td>uPA(hsa)</td>\n",
       "      <td>hsa-miR-23b-3p</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Western Blot</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927115</th>\n",
       "      <td>vimentin(hsa)</td>\n",
       "      <td>vimentin(hsa)</td>\n",
       "      <td>hsa-miR-9-5p</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Western Blot</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927116</th>\n",
       "      <td>Â Â Â Â (PTPRG)(hsa)</td>\n",
       "      <td>Â Â Â Â (PTPRG)(hsa)</td>\n",
       "      <td>hsa-miR-146a-5p</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>UP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927117</th>\n",
       "      <td>Î±-actin(mmu)</td>\n",
       "      <td>Î±-actin(mmu)</td>\n",
       "      <td>mmu-miR-24-3p</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>qPCR</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>UP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927118</th>\n",
       "      <td>Î±-actin(mmu)</td>\n",
       "      <td>Î±-actin(mmu)</td>\n",
       "      <td>mmu-miR-24-3p</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Western Blot</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>UP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>927119 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      geneId              geneName            mirna  \\\n",
       "0         0910001A06Rik(mmu)    0910001A06Rik(mmu)   mmu-miR-124-3p   \n",
       "1         1200004M23Rik(mmu)    1200004M23Rik(mmu)   mmu-miR-124-3p   \n",
       "2         1700027J05Rik(mmu)    1700027J05Rik(mmu)   mmu-miR-124-3p   \n",
       "3         1810015A11Rik(mmu)    1810015A11Rik(mmu)   mmu-miR-124-3p   \n",
       "4         2310047A01Rik(mmu)    2310047A01Rik(mmu)   mmu-miR-124-3p   \n",
       "...                      ...                   ...              ...   \n",
       "927114              uPA(hsa)              uPA(hsa)   hsa-miR-23b-3p   \n",
       "927115         vimentin(hsa)         vimentin(hsa)     hsa-miR-9-5p   \n",
       "927116  Â Â Â Â (PTPRG)(hsa)  Â Â Â Â (PTPRG)(hsa)  hsa-miR-146a-5p   \n",
       "927117         Î±-actin(mmu)         Î±-actin(mmu)    mmu-miR-24-3p   \n",
       "927118         Î±-actin(mmu)         Î±-actin(mmu)    mmu-miR-24-3p   \n",
       "\n",
       "             species cell_line tissue          category        method  \\\n",
       "0       Mus musculus      <NA>    NaN  Cancer/Malignant   Microarrays   \n",
       "1       Mus musculus      <NA>    NaN  Cancer/Malignant   Microarrays   \n",
       "2       Mus musculus      <NA>    NaN  Cancer/Malignant   Microarrays   \n",
       "3       Mus musculus      <NA>    NaN  Cancer/Malignant   Microarrays   \n",
       "4       Mus musculus      <NA>    NaN  Cancer/Malignant   Microarrays   \n",
       "...              ...       ...    ...               ...           ...   \n",
       "927114  Homo sapiens      <NA>    NaN  Cancer/Malignant  Western Blot   \n",
       "927115  Homo sapiens      <NA>    NaN  Cancer/Malignant  Western Blot   \n",
       "927116  Homo sapiens      <NA>    NaN  Cancer/Malignant   Microarrays   \n",
       "927117  Mus musculus      <NA>    NaN  Cancer/Malignant          qPCR   \n",
       "927118  Mus musculus      <NA>    NaN  Cancer/Malignant  Western Blot   \n",
       "\n",
       "       positive_negative direct_indirect up_down condition  \n",
       "0               POSITIVE        INDIRECT      UP       NaN  \n",
       "1               POSITIVE        INDIRECT    DOWN       NaN  \n",
       "2               POSITIVE        INDIRECT      UP       NaN  \n",
       "3               POSITIVE        INDIRECT    DOWN       NaN  \n",
       "4               POSITIVE        INDIRECT    DOWN       NaN  \n",
       "...                  ...             ...     ...       ...  \n",
       "927114          POSITIVE        INDIRECT    DOWN       NaN  \n",
       "927115          NEGATIVE        INDIRECT     NaN       NaN  \n",
       "927116          POSITIVE        INDIRECT      UP       NaN  \n",
       "927117          POSITIVE        INDIRECT      UP       NaN  \n",
       "927118          POSITIVE        INDIRECT      UP       NaN  \n",
       "\n",
       "[927119 rows x 12 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_downloader('https://dianalab.e-ce.uth.gr/downloads/tarbase_v8_data.tar.gz', edge_data_location)\n",
    "\n",
    "with tarfile.TarFile(edge_data_location+'tarbase_v8_data.tar', 'r') as tar_ref:\n",
    "    tar_ref.extractall(edge_data_location)\n",
    "    \n",
    "# Remove tar file\n",
    "os.remove(edge_data_location+'tarbase_v8_data.tar')\n",
    "    \n",
    "# Rename file adding relationship's identifier\n",
    "os.rename(edge_data_location+'TarBase_v8_download.txt',\n",
    "          edge_data_location+'gene-miRNA_TarBase_v8_download.txt')   \n",
    "\n",
    "with open(edge_data_location + 'gene-miRNA_TarBase_v8_download.txt') as f:\n",
    "    data = f.read()\n",
    "\n",
    "data = pd.read_csv(edge_data_location + 'gene-miRNA_TarBase_v8_download.txt', sep=\"\\t\", dtype={\"cell_line\": \"string\"})  \n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneId</th>\n",
       "      <th>geneName</th>\n",
       "      <th>mirna</th>\n",
       "      <th>species</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>tissue</th>\n",
       "      <th>category</th>\n",
       "      <th>method</th>\n",
       "      <th>positive_negative</th>\n",
       "      <th>direct_indirect</th>\n",
       "      <th>up_down</th>\n",
       "      <th>condition</th>\n",
       "      <th>p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>296946</th>\n",
       "      <td>ENSG00000138756</td>\n",
       "      <td>BMP2K</td>\n",
       "      <td>hsa-mir-22</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HUVEC</td>\n",
       "      <td>Umbilical Vein</td>\n",
       "      <td>Normal/Primary</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34726</th>\n",
       "      <td>ENSG00000062650</td>\n",
       "      <td>WAPAL</td>\n",
       "      <td>hsa-mir-3657</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>293S</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>Embryonic/Fetal</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>treatment: emetine</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145176</th>\n",
       "      <td>ENSG00000109670</td>\n",
       "      <td>FBXW7</td>\n",
       "      <td>hsa-mir-4782</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HEK293</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>Embryonic/Fetal</td>\n",
       "      <td>Luciferase Reporter Assay</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109953</th>\n",
       "      <td>ENSG00000102978</td>\n",
       "      <td>POLR2C</td>\n",
       "      <td>hsa-mir-29c</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>MKN74</td>\n",
       "      <td>Stomach</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24hrs post-transfection, Overexpression</td>\n",
       "      <td>3p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>446282</th>\n",
       "      <td>ENSG00000167904</td>\n",
       "      <td>TMEM68</td>\n",
       "      <td>hsa-mir-95</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>MCF7</td>\n",
       "      <td>Mammary Gland</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>24hrs post-transfection, Overexpression</td>\n",
       "      <td>3p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246354</th>\n",
       "      <td>ENSG00000130726</td>\n",
       "      <td>TRIM28</td>\n",
       "      <td>hsa-mir-30c-1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>293S</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>Embryonic/Fetal</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>treatment:hippuristanol</td>\n",
       "      <td>3p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105785</th>\n",
       "      <td>ENSG00000101974</td>\n",
       "      <td>ATP11C</td>\n",
       "      <td>hsa-mir-3176</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>293S</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>Embryonic/Fetal</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>no treatment (control)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496197</th>\n",
       "      <td>ENSG00000176903</td>\n",
       "      <td>PNMA1</td>\n",
       "      <td>hsa-mir-23a</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>BC1</td>\n",
       "      <td>Pleura</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>PAR-CLIP</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>426081</th>\n",
       "      <td>ENSG00000165359</td>\n",
       "      <td>DDX26B</td>\n",
       "      <td>hsa-mir-182</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Melanoma metastatic Liver Cells</td>\n",
       "      <td>Liver</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Administered twice per week over 4 w, Silencing</td>\n",
       "      <td>5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42158</th>\n",
       "      <td>ENSG00000068024</td>\n",
       "      <td>HDAC4</td>\n",
       "      <td>hsa-mir-140</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>BETA Cells</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Normal/Primary</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5p</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>50000 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 geneId geneName          mirna       species  \\\n",
       "296946  ENSG00000138756    BMP2K     hsa-mir-22  Homo sapiens   \n",
       "34726   ENSG00000062650    WAPAL   hsa-mir-3657  Homo sapiens   \n",
       "145176  ENSG00000109670    FBXW7   hsa-mir-4782  Homo sapiens   \n",
       "109953  ENSG00000102978   POLR2C    hsa-mir-29c  Homo sapiens   \n",
       "446282  ENSG00000167904   TMEM68     hsa-mir-95  Homo sapiens   \n",
       "...                 ...      ...            ...           ...   \n",
       "246354  ENSG00000130726   TRIM28  hsa-mir-30c-1  Homo sapiens   \n",
       "105785  ENSG00000101974   ATP11C   hsa-mir-3176  Homo sapiens   \n",
       "496197  ENSG00000176903    PNMA1    hsa-mir-23a  Homo sapiens   \n",
       "426081  ENSG00000165359   DDX26B    hsa-mir-182  Homo sapiens   \n",
       "42158   ENSG00000068024    HDAC4    hsa-mir-140  Homo sapiens   \n",
       "\n",
       "                              cell_line          tissue          category  \\\n",
       "296946                            HUVEC  Umbilical Vein    Normal/Primary   \n",
       "34726                              293S          Kidney   Embryonic/Fetal   \n",
       "145176                           HEK293          Kidney   Embryonic/Fetal   \n",
       "109953                            MKN74         Stomach  Cancer/Malignant   \n",
       "446282                             MCF7   Mammary Gland  Cancer/Malignant   \n",
       "...                                 ...             ...               ...   \n",
       "246354                             293S          Kidney   Embryonic/Fetal   \n",
       "105785                             293S          Kidney   Embryonic/Fetal   \n",
       "496197                              BC1          Pleura  Cancer/Malignant   \n",
       "426081  Melanoma metastatic Liver Cells           Liver  Cancer/Malignant   \n",
       "42158                        BETA Cells        Pancreas    Normal/Primary   \n",
       "\n",
       "                           method positive_negative direct_indirect up_down  \\\n",
       "296946                  HITS-CLIP          POSITIVE          DIRECT    DOWN   \n",
       "34726                   HITS-CLIP          POSITIVE          DIRECT    DOWN   \n",
       "145176  Luciferase Reporter Assay          POSITIVE          DIRECT    DOWN   \n",
       "109953                Microarrays          NEGATIVE        INDIRECT     NaN   \n",
       "446282                Microarrays          POSITIVE        INDIRECT    DOWN   \n",
       "...                           ...               ...             ...     ...   \n",
       "246354                  HITS-CLIP          POSITIVE          DIRECT    DOWN   \n",
       "105785                  HITS-CLIP          POSITIVE          DIRECT    DOWN   \n",
       "496197                   PAR-CLIP          POSITIVE          DIRECT    DOWN   \n",
       "426081                Microarrays          NEGATIVE        INDIRECT     NaN   \n",
       "42158                   HITS-CLIP          POSITIVE          DIRECT    DOWN   \n",
       "\n",
       "                                              condition    p  \n",
       "296946                                              NaN   3p  \n",
       "34726                                treatment: emetine  NaN  \n",
       "145176                                              NaN   3p  \n",
       "109953          24hrs post-transfection, Overexpression   3p  \n",
       "446282          24hrs post-transfection, Overexpression   3p  \n",
       "...                                                 ...  ...  \n",
       "246354                          treatment:hippuristanol   3p  \n",
       "105785                           no treatment (control)  NaN  \n",
       "496197                                              NaN   3p  \n",
       "426081  Administered twice per week over 4 w, Silencing   5p  \n",
       "42158                                               NaN   5p  \n",
       "\n",
       "[50000 rows x 13 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For the time being, we keep only Homo sapiens rows\n",
    "data = data[data['species'].str.contains(\"Homo sapiens\")]\n",
    "\n",
    "# Moreover, we keep only 50k (random) rows to reduce input size\n",
    "data = data.sample(n=50000)\n",
    "\n",
    "# NCRO Ontology ignores if a transcript is \"3p\" or \"5p\", so we store this information as additional column\n",
    "data['p'] = data[data['mirna'].str.contains(\"p\")]['mirna']\n",
    "data[\"p\"] = data[\"p\"].str[-2:]\n",
    "data[\"mirna\"] = data[\"mirna\"].str.replace(r'-3p$', '')\n",
    "data[\"mirna\"] = data[\"mirna\"].str.replace(r'-5p$', '')\n",
    "\n",
    "# If you want to learn more about why we have 5p or 3p human miRNAs:\n",
    "# https://www.mirbase.org/blog/category/nomenclature/\n",
    "\n",
    "# Again, for NCRO IDs compatability\n",
    "data['mirna'] = data['mirna'].str.lower()\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "data.to_csv(edge_data_location + 'gene-miRNA_TarBase_v8_download.txt', header=None, sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### miRNA-disease from [miR2Disease](http://watson.compbio.iupui.edu:8080/miR2Disease/)\n",
    "miR2Disease, a manually curated database, aims at providing a comprehensive resource of miRNA deregulation in various human diseases."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Data from http://watson.compbio.iupui.edu:8080/miR2Disease/download/AllEntries.txt\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7f-2</td>\n",
       "      <td>kidney cancer</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>microarray</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>Micro-RNA profiling in kidney and bladder canc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-let-7g</td>\n",
       "      <td>hepatocellular carcinoma (HCC)</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>Identification of metastasis-related microRNAs...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7g</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>The let-7 microRNA represses cell proliferatio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-let-7g</td>\n",
       "      <td>non-small cell lung cancer (NSCLC)</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>Suppression of non-small cell lung tumor devel...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-let-7g</td>\n",
       "      <td>ovarian cancer (OC)</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>Let-7 expression defines two differentiation s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2897</th>\n",
       "      <td>hsa-mir-21</td>\n",
       "      <td>glioblastoma multiforme (GBM)</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>miR-124 and miR-137 inhibit proliferation of g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2898</th>\n",
       "      <td>hsa-mir-21</td>\n",
       "      <td>glioma</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>MicroRNA 21 promotes glioma invasion by target...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2899</th>\n",
       "      <td>hsa-mir-21</td>\n",
       "      <td>hepatocellular carcinoma (HCC)</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>Downregulation of miR-122 in the rodent and hu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2900</th>\n",
       "      <td>hsa-mir-21</td>\n",
       "      <td>Inclusion body myositis (IBM)</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>microarray</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>Distinctive patterns of microRNA expression in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2901</th>\n",
       "      <td>hsa-mir-21</td>\n",
       "      <td>limb-girdle muscular dystrophies types 2A (LGM...</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>microarray</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>Distinctive patterns of microRNA expression in...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2902 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 0                                                  1  \\\n",
       "0     hsa-let-7f-2                                      kidney cancer   \n",
       "1       hsa-let-7g                     hepatocellular carcinoma (HCC)   \n",
       "2       hsa-let-7g                                        lung cancer   \n",
       "3       hsa-let-7g                 non-small cell lung cancer (NSCLC)   \n",
       "4       hsa-let-7g                                ovarian cancer (OC)   \n",
       "...            ...                                                ...   \n",
       "2897    hsa-mir-21                      glioblastoma multiforme (GBM)   \n",
       "2898    hsa-mir-21                                             glioma   \n",
       "2899    hsa-mir-21                     hepatocellular carcinoma (HCC)   \n",
       "2900    hsa-mir-21                      Inclusion body myositis (IBM)   \n",
       "2901    hsa-mir-21  limb-girdle muscular dystrophies types 2A (LGM...   \n",
       "\n",
       "                   2                           3       4  \\\n",
       "0       up-regulated                  microarray  2007.0   \n",
       "1     down-regulated  Northern blot, qRT-PCR etc  2008.0   \n",
       "2     down-regulated  Northern blot, qRT-PCR etc  2007.0   \n",
       "3     down-regulated  Northern blot, qRT-PCR etc  2008.0   \n",
       "4     down-regulated  Northern blot, qRT-PCR etc  2007.0   \n",
       "...              ...                         ...     ...   \n",
       "2897    up-regulated  Northern blot, qRT-PCR etc  2008.0   \n",
       "2898    up-regulated  Northern blot, qRT-PCR etc  2008.0   \n",
       "2899    up-regulated  Northern blot, qRT-PCR etc  2006.0   \n",
       "2900    up-regulated                  microarray  2007.0   \n",
       "2901    up-regulated                  microarray  2007.0   \n",
       "\n",
       "                                                      5  \n",
       "0     Micro-RNA profiling in kidney and bladder canc...  \n",
       "1     Identification of metastasis-related microRNAs...  \n",
       "2     The let-7 microRNA represses cell proliferatio...  \n",
       "3     Suppression of non-small cell lung tumor devel...  \n",
       "4     Let-7 expression defines two differentiation s...  \n",
       "...                                                 ...  \n",
       "2897  miR-124 and miR-137 inhibit proliferation of g...  \n",
       "2898  MicroRNA 21 promotes glioma invasion by target...  \n",
       "2899  Downregulation of miR-122 in the rodent and hu...  \n",
       "2900  Distinctive patterns of microRNA expression in...  \n",
       "2901  Distinctive patterns of microRNA expression in...  \n",
       "\n",
       "[2902 rows x 6 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_downloader('http://watson.compbio.iupui.edu:8080/miR2Disease/download/AllEntries.txt', edge_data_location)\n",
    "    \n",
    "# Rename file adding relationship's identifier\n",
    "os.rename(edge_data_location+'AllEntries.txt',\n",
    "          edge_data_location+'miRNA-disease_miR2Disease.txt') \n",
    "\n",
    "with open(edge_data_location + 'miRNA-disease_mir2Disease.txt') as f:\n",
    "    data = f.read()\n",
    "\n",
    "data = pd.read_csv(edge_data_location + 'miRNA-disease_miR2Disease.txt', sep=\"\\t\", header=None)  \n",
    "# For NCRO IDs compatability\n",
    "data[0] = data[0].str.lower()\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>desc</th>\n",
       "      <th>doid</th>\n",
       "      <th>mirna</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>adenoma</td>\n",
       "      <td>DOID_657</td>\n",
       "      <td>hsa-let-7a</td>\n",
       "      <td>normal</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>Disrupting the pairing between let-7 and Hmga2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>adrenocortical carcinoma</td>\n",
       "      <td>DOID_3948</td>\n",
       "      <td>hsa-mir-184</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>microarray</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>Integrative molecular-bioinformatics study of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>adrenocortical carcinoma</td>\n",
       "      <td>DOID_3948</td>\n",
       "      <td>hsa-mir-503</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>microarray</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>Integrative molecular-bioinformatics study of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>adrenocortical carcinoma</td>\n",
       "      <td>DOID_3948</td>\n",
       "      <td>hsa-mir-511</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>microarray</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>Integrative molecular-bioinformatics study of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>adrenocortical carcinoma</td>\n",
       "      <td>DOID_3948</td>\n",
       "      <td>hsa-mir-214</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>microarray</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>Integrative molecular-bioinformatics study of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1290</th>\n",
       "      <td>vascular disease</td>\n",
       "      <td>DOID_178</td>\n",
       "      <td>hsa-mir-21</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>MicroRNA expression signature and antisense-me...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1291</th>\n",
       "      <td>vascular disease</td>\n",
       "      <td>DOID_178</td>\n",
       "      <td>hsa-mir-352</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>MicroRNA expression signature and antisense-me...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1292</th>\n",
       "      <td>vascular disease</td>\n",
       "      <td>DOID_178</td>\n",
       "      <td>hsa-mir-365</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>MicroRNA expression signature and antisense-me...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1293</th>\n",
       "      <td>vesicular stomatitis</td>\n",
       "      <td>DOID_10881</td>\n",
       "      <td>hsa-mir-93</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>Hypersusceptibility to vesicular stomatitis vi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1294</th>\n",
       "      <td>vesicular stomatitis</td>\n",
       "      <td>DOID_10881</td>\n",
       "      <td>hsa-mir-24</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>Hypersusceptibility to vesicular stomatitis vi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1295 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          desc        doid        mirna               2  \\\n",
       "0                      adenoma    DOID_657   hsa-let-7a          normal   \n",
       "1     adrenocortical carcinoma   DOID_3948  hsa-mir-184    up-regulated   \n",
       "2     adrenocortical carcinoma   DOID_3948  hsa-mir-503    up-regulated   \n",
       "3     adrenocortical carcinoma   DOID_3948  hsa-mir-511  down-regulated   \n",
       "4     adrenocortical carcinoma   DOID_3948  hsa-mir-214  down-regulated   \n",
       "...                        ...         ...          ...             ...   \n",
       "1290          vascular disease    DOID_178   hsa-mir-21    up-regulated   \n",
       "1291          vascular disease    DOID_178  hsa-mir-352    up-regulated   \n",
       "1292          vascular disease    DOID_178  hsa-mir-365  down-regulated   \n",
       "1293      vesicular stomatitis  DOID_10881   hsa-mir-93  down-regulated   \n",
       "1294      vesicular stomatitis  DOID_10881   hsa-mir-24  down-regulated   \n",
       "\n",
       "                               3       4  \\\n",
       "0     Northern blot, qRT-PCR etc  2007.0   \n",
       "1                     microarray  2009.0   \n",
       "2                     microarray  2009.0   \n",
       "3                     microarray  2009.0   \n",
       "4                     microarray  2009.0   \n",
       "...                          ...     ...   \n",
       "1290  Northern blot, qRT-PCR etc  2007.0   \n",
       "1291  Northern blot, qRT-PCR etc  2007.0   \n",
       "1292  Northern blot, qRT-PCR etc  2007.0   \n",
       "1293  Northern blot, qRT-PCR etc  2007.0   \n",
       "1294  Northern blot, qRT-PCR etc  2007.0   \n",
       "\n",
       "                                                      5  \n",
       "0     Disrupting the pairing between let-7 and Hmga2...  \n",
       "1     Integrative molecular-bioinformatics study of ...  \n",
       "2     Integrative molecular-bioinformatics study of ...  \n",
       "3     Integrative molecular-bioinformatics study of ...  \n",
       "4     Integrative molecular-bioinformatics study of ...  \n",
       "...                                                 ...  \n",
       "1290  MicroRNA expression signature and antisense-me...  \n",
       "1291  MicroRNA expression signature and antisense-me...  \n",
       "1292  MicroRNA expression signature and antisense-me...  \n",
       "1293  Hypersusceptibility to vesicular stomatitis vi...  \n",
       "1294  Hypersusceptibility to vesicular stomatitis vi...  \n",
       "\n",
       "[1295 rows x 7 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We need to move from description to DO Ontology IDs\n",
    "data.columns = ['mirna', 'desc', 2,3,4,5]\n",
    "disease2mirna = pd.merge(descDOmap, data, on=['desc'])\n",
    "disease2mirna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "disease2mirna.to_csv(edge_data_location + 'miRNA-disease_mir2Disease.txt', header=None, sep='\\t', index=None)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "pycharm": {
   "stem_cell": {
    "cell_type": "raw",
    "metadata": {
     "collapsed": false
    },
    "source": []
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
