{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# <p style=\"text-align: center;\">RNA Knowledge Graph Build Data Preparation</p>\n",
    "    \n",
    "***\n",
    "***\n",
    "\n",
    "**Authors:** [ECavalleri](https://mail.google.com/mail/u/0/?view=cm&fs=1&tf=1&to=emanuele.cavalleri@unimi.it), [TJCallahan](https://mail.google.com/mail/u/0/?view=cm&fs=1&tf=1&to=callahantiff@gmail.com), [MMesiti](https://mail.google.com/mail/u/0/?view=cm&fs=1&tf=1&to=marco.mesiti@unimi.it)\n",
    "\n",
    "**GitHub Repositories:** [testRNA-KG](https://github.com/emanuelecavalleri/testRNA-KG), [PheKnowLator](https://github.com/callahantiff/PheKnowLator/) (PKT)\n",
    "<!--- **Release:** **[v2.0.0](https://github.com/callahantiff/PheKnowLator/wiki/v2.0.0)** --->\n",
    "  \n",
    "<br>  \n",
    "  \n",
    "**Purpose:** This notebook serves as a script to download, process, map, and clean data in order to build edges for the simplified RNA-centered Knowledge Graph.\n",
    "\n",
    "<br>\n",
    "\n",
    "**Assumptions:**   \n",
    "- Edge data downloads ➞ `./resources/edge_data`  \n",
    "- Ontologies ➞ `./resources/ontologies`    \n",
    "- Processed data write location ➞ `./resources/processed_data`  \n",
    "\n",
    "<br>\n",
    "\n",
    "**Dependencies:**   \n",
    "- **Scripts**: This notebook utilizes several helper functions, which are stored in the [`data_utils.py`](https://github.com/callahantiff/PheKnowLator/blob/master/pkt_kg/utils/data_utils.py) and [`kg_utils.py`](https://github.com/callahantiff/PheKnowLator/blob/master/pkt_kg/utils/kg_utils.py) scripts.  \n",
    "- **Data**: All downloaded and generated data sources are provided through [this](https://drive.google.com/drive/folders/1sev5zczMviX7UVqMhTpkFXG43K3nQa9f) dedicated Google Drive repository. <u>This notebook will download everything that is needed for you</u>.  \n",
    "_____\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table of Contents\n",
    "***\n",
    "\n",
    "### [Download Ontologies](#create-ontologies)\n",
    "\n",
    "\n",
    "### [Create Identifier Maps ](#create-identifier-maps)   \n",
    "\n",
    "\n",
    "### [Download and process Edge Datasets](#create-edges)  \n",
    "\n",
    "____\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set-Up Environment\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "import sys\n",
    "!{sys.executable} -m pip install -r requirements.txt\n",
    "sys.path.append('../')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import needed libraries\n",
    "import os\n",
    "import pickle\n",
    "import tarfile\n",
    "import pandas as pd\n",
    "import gffpandas.gffpandas as gffpd\n",
    "from rdflib import Graph\n",
    "\n",
    "from pkt_kg.utils import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Define Global Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# directory to store resources\n",
    "resource_data_location = '../resources/'\n",
    "\n",
    "# directory to use for processing data\n",
    "processed_data_location = '../resources/processed_data/'\n",
    "\n",
    "# directory to write ontology data to\n",
    "ontology_data_location = '../resources/ontologies/'\n",
    "\n",
    "# directory to write edges data to\n",
    "edge_data_location = '../resources/edge_data/'\n",
    "\n",
    "# owltools location\n",
    "owltools_location = '../pkt_kg/libs/owltools'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download data function for already processed data\n",
    "def download(name, path):\n",
    "    url = 'https://storage.googleapis.com/pheknowlator/current_build/data/processed_data/' + name\n",
    "    #if not os.path.exists(path + name):\n",
    "    data_downloader(url, path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "***\n",
    "### DOWNLOAD ONTOLOGIES  <a class=\"anchor\" id=\"create-ontologies\"></a>\n",
    "We must establish a unified standard for identifying entities within our simplified RNA-centered KG. Entities are relations, diseases, miRNAs, and genes. While well-reputed bio-ontologies provide terms for relations and diseases, miRNAs and genes lack direct correspondences.\n",
    "***\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Relation Ontology ([RO](https://www.ebi.ac.uk/ols/ontologies/ro))\n",
    "The OBO Relations Ontology (RO) is a collection of OWL relations (ObjectProperties) intended for use across a wide variety of biological ontologies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: An illegal reflective access operation has occurred\n",
      "WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$2 (file:/home/ecavalleri/testRNA-KG/pkt_kg/libs/owltools) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)\n",
      "WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$2\n",
      "WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations\n",
      "WARNING: All illegal access operations will be denied in a future release\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "command = '{} {} --merge-import-closure -o {}'\n",
    "os.system(command.format(owltools_location, 'http://purl.obolibrary.org/obo/ro.owl',\n",
    "                         ontology_data_location + 'ro_with_imports.owl'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Data from https://storage.googleapis.com/pheknowlator/current_build/data/processed_data/RELATIONS_LABELS.txt\n",
      "Downloading Data from https://storage.googleapis.com/pheknowlator/current_build/data/processed_data/INVERSE_RELATIONS.txt\n",
      "There are 667 RO Relations and Labels\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Label</th>\n",
       "      <th>Relation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>helper property (not for use in curation)</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_0002464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>developmentally replaces</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_0002285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>is approximately equivalent to</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_0002603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>has intracellular endoparasite</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_0002641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>supplies</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_0002178</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       Label   \n",
       "0  helper property (not for use in curation)  \\\n",
       "1                   developmentally replaces   \n",
       "2             is approximately equivalent to   \n",
       "3             has intracellular endoparasite   \n",
       "4                                   supplies   \n",
       "\n",
       "                                    Relation  \n",
       "0  http://purl.obolibrary.org/obo/RO_0002464  \n",
       "1  http://purl.obolibrary.org/obo/RO_0002285  \n",
       "2  http://purl.obolibrary.org/obo/RO_0002603  \n",
       "3  http://purl.obolibrary.org/obo/RO_0002641  \n",
       "4  http://purl.obolibrary.org/obo/RO_0002178  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Relations labels are already provided by PKT ecosystem\n",
    "download('RELATIONS_LABELS.txt', '../resources/relations_data/')\n",
    "download('INVERSE_RELATIONS.txt', '../resources/relations_data/')\n",
    "\n",
    "# Load data, print row count, and preview it\n",
    "ro_data_label = pandas.read_csv('../resources/relations_data/'+'RELATIONS_LABELS.txt', header=0, delimiter='\\t')\n",
    "\n",
    "# We also specify symmetric relations (e.g., RO_0002434 inverseOf RO_0002434)\n",
    "symmetric_relation = pd.DataFrame({'Relation': 'RO_0002434', 'Inverse_Relation': 'RO_0002434'}, index=[0])\n",
    "ro_data_inv = pandas.read_csv('../resources/relations_data/'+'INVERSE_RELATIONS.txt', header=0, delimiter='\\t')\n",
    "ro_data_inv = pd.concat([ro_data_inv, symmetric_relation], ignore_index=True)\n",
    "\n",
    "print('There are {edge_count} RO Relations and Labels'.format(edge_count=len(ro_data_label)))\n",
    "ro_data_label.head(n=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Mondo Disease Ontology ([Mondo](https://www.ebi.ac.uk/ols/ontologies/ro))\n",
    "A semi-automatically constructed ontology that merges in multiple disease resources to yield a coherent merged ontology."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: An illegal reflective access operation has occurred\n",
      "WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$2 (file:/home/ecavalleri/testRNA-KG/pkt_kg/libs/owltools) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)\n",
      "WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$2\n",
      "WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations\n",
      "WARNING: All illegal access operations will be denied in a future release\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "command = '{} {} --merge-import-closure -o {}'\n",
    "os.system(command.format(owltools_location, 'http://purl.obolibrary.org/obo/mondo.owl',\n",
    "                         ontology_data_location + 'mondo_with_imports.owl'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At this point, please run the [<tt>Ontology_Cleaning.ipynb</tt>](https://github.com/callahantiff/PheKnowLator/blob/master/notebooks/Ontology_Cleaning.ipynb) notebook provided by PKT."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "***\n",
    "### DOWNLOAD EDGES  <a class=\"anchor\" id=\"create-edges\"></a>\n",
    "***\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### gene-disease from [Human Disease Molecular Mechanisms](https://github.com/callahantiff/PheKnowLator/wiki/Building-a-KG-of-Human-Disease-Molecular-Mechanisms) (PKT-built)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Data from https://storage.googleapis.com/pheknowlator/current_build/data/original_data/curated_gene_disease_associations.tsv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneId</th>\n",
       "      <th>geneSymbol</th>\n",
       "      <th>DSI</th>\n",
       "      <th>DPI</th>\n",
       "      <th>diseaseId</th>\n",
       "      <th>diseaseName</th>\n",
       "      <th>diseaseType</th>\n",
       "      <th>diseaseClass</th>\n",
       "      <th>diseaseSemanticType</th>\n",
       "      <th>score</th>\n",
       "      <th>EI</th>\n",
       "      <th>YearInitial</th>\n",
       "      <th>YearFinal</th>\n",
       "      <th>NofPmids</th>\n",
       "      <th>NofSnps</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.700</td>\n",
       "      <td>0.538</td>\n",
       "      <td>C0019209</td>\n",
       "      <td>Hepatomegaly</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>C23;C06</td>\n",
       "      <td>Finding</td>\n",
       "      <td>0.30</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.700</td>\n",
       "      <td>0.538</td>\n",
       "      <td>C0036341</td>\n",
       "      <td>Schizophrenia</td>\n",
       "      <td>disease</td>\n",
       "      <td>F03</td>\n",
       "      <td>Mental or Behavioral Dysfunction</td>\n",
       "      <td>0.30</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>0.529</td>\n",
       "      <td>0.769</td>\n",
       "      <td>C0002395</td>\n",
       "      <td>Alzheimer's Disease</td>\n",
       "      <td>disease</td>\n",
       "      <td>C10;F03</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.769</td>\n",
       "      <td>1998.0</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>0.529</td>\n",
       "      <td>0.769</td>\n",
       "      <td>C0007102</td>\n",
       "      <td>Malignant tumor of colon</td>\n",
       "      <td>disease</td>\n",
       "      <td>C06;C04</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.31</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>0.529</td>\n",
       "      <td>0.769</td>\n",
       "      <td>C0009375</td>\n",
       "      <td>Colonic Neoplasms</td>\n",
       "      <td>group</td>\n",
       "      <td>C06;C04</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.30</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84033</th>\n",
       "      <td>109580095</td>\n",
       "      <td>HBB-LCR</td>\n",
       "      <td>0.743</td>\n",
       "      <td>0.115</td>\n",
       "      <td>C0002875</td>\n",
       "      <td>Cooley's anemia</td>\n",
       "      <td>disease</td>\n",
       "      <td>C16;C15</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84034</th>\n",
       "      <td>109580095</td>\n",
       "      <td>HBB-LCR</td>\n",
       "      <td>0.743</td>\n",
       "      <td>0.115</td>\n",
       "      <td>C0005283</td>\n",
       "      <td>beta Thalassemia</td>\n",
       "      <td>disease</td>\n",
       "      <td>C16;C15</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84035</th>\n",
       "      <td>109580095</td>\n",
       "      <td>HBB-LCR</td>\n",
       "      <td>0.743</td>\n",
       "      <td>0.115</td>\n",
       "      <td>C0019025</td>\n",
       "      <td>Hemoglobin F Disease</td>\n",
       "      <td>disease</td>\n",
       "      <td>C16;C15</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84036</th>\n",
       "      <td>109580095</td>\n",
       "      <td>HBB-LCR</td>\n",
       "      <td>0.743</td>\n",
       "      <td>0.115</td>\n",
       "      <td>C0085578</td>\n",
       "      <td>Thalassemia Minor</td>\n",
       "      <td>disease</td>\n",
       "      <td>C16;C15</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84037</th>\n",
       "      <td>109580095</td>\n",
       "      <td>HBB-LCR</td>\n",
       "      <td>0.743</td>\n",
       "      <td>0.115</td>\n",
       "      <td>C0271979</td>\n",
       "      <td>Thalassemia Intermedia</td>\n",
       "      <td>disease</td>\n",
       "      <td>C16;C15</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>84038 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          geneId geneSymbol    DSI    DPI diseaseId               diseaseName   \n",
       "0              1       A1BG  0.700  0.538  C0019209              Hepatomegaly  \\\n",
       "1              1       A1BG  0.700  0.538  C0036341             Schizophrenia   \n",
       "2              2        A2M  0.529  0.769  C0002395       Alzheimer's Disease   \n",
       "3              2        A2M  0.529  0.769  C0007102  Malignant tumor of colon   \n",
       "4              2        A2M  0.529  0.769  C0009375         Colonic Neoplasms   \n",
       "...          ...        ...    ...    ...       ...                       ...   \n",
       "84033  109580095    HBB-LCR  0.743  0.115  C0002875           Cooley's anemia   \n",
       "84034  109580095    HBB-LCR  0.743  0.115  C0005283          beta Thalassemia   \n",
       "84035  109580095    HBB-LCR  0.743  0.115  C0019025      Hemoglobin F Disease   \n",
       "84036  109580095    HBB-LCR  0.743  0.115  C0085578         Thalassemia Minor   \n",
       "84037  109580095    HBB-LCR  0.743  0.115  C0271979    Thalassemia Intermedia   \n",
       "\n",
       "      diseaseType diseaseClass               diseaseSemanticType  score   \n",
       "0       phenotype      C23;C06                           Finding   0.30  \\\n",
       "1         disease          F03  Mental or Behavioral Dysfunction   0.30   \n",
       "2         disease      C10;F03               Disease or Syndrome   0.50   \n",
       "3         disease      C06;C04                Neoplastic Process   0.31   \n",
       "4           group      C06;C04                Neoplastic Process   0.30   \n",
       "...           ...          ...                               ...    ...   \n",
       "84033     disease      C16;C15               Disease or Syndrome   0.30   \n",
       "84034     disease      C16;C15               Disease or Syndrome   0.30   \n",
       "84035     disease      C16;C15               Disease or Syndrome   0.30   \n",
       "84036     disease      C16;C15               Disease or Syndrome   0.30   \n",
       "84037     disease      C16;C15               Disease or Syndrome   0.30   \n",
       "\n",
       "          EI  YearInitial  YearFinal  NofPmids  NofSnps     source  \n",
       "0      1.000       2017.0     2017.0         1        0  CTD_human  \n",
       "1      1.000       2015.0     2015.0         1        0  CTD_human  \n",
       "2      0.769       1998.0     2018.0         3        0  CTD_human  \n",
       "3      1.000       2004.0     2019.0         1        0  CTD_human  \n",
       "4      1.000       2004.0     2004.0         1        0  CTD_human  \n",
       "...      ...          ...        ...       ...      ...        ...  \n",
       "84033    NaN          NaN        NaN         0        0  CTD_human  \n",
       "84034    NaN          NaN        NaN         0        0  CTD_human  \n",
       "84035    NaN          NaN        NaN         0        0  CTD_human  \n",
       "84036    NaN          NaN        NaN         0        0  CTD_human  \n",
       "84037    NaN          NaN        NaN         0        0  CTD_human  \n",
       "\n",
       "[84038 rows x 16 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_downloader('https://storage.googleapis.com/pheknowlator/current_build/data/original_data/curated_gene_disease_associations.tsv',\n",
    "                edge_data_location)\n",
    "\n",
    "# Rename file adding relationship's identifier\n",
    "os.rename(edge_data_location+'curated_gene_disease_associations.tsv',\n",
    "          edge_data_location+'gene-disease_curated_gene_disease_associations.tsv')\n",
    "\n",
    "with open(edge_data_location + 'gene-disease_curated_gene_disease_associations.tsv') as f:\n",
    "    data = f.read()\n",
    "\n",
    "data = pd.read_csv(edge_data_location + 'gene-disease_curated_gene_disease_associations.tsv', sep=\"\\t\")  \n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We keep only 50k (random) rows to reduce input size\n",
    "data = data.sample(n=50000)\n",
    "data.to_csv(edge_data_location + 'gene-disease_curated_gene_disease_associations.tsv', header=None, sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For representing genes, we can use NCBI Entrez Gene identifiers (<tt>geneID</tt> column). It is worth noting that symbols could have been a viable choice as well. For denoting diseases (<tt>diseaseID</tt> column), we can notice the original tsv adopts DisGeNET identifiers."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### gene-miRNA from [TarBase](https://dianalab.e-ce.uth.gr/html/diana/web/index.php?r=tarbasev8/index)\n",
    "DIANA-TarBase v8 is a reference database devoted to the indexing of experimentally supported microRNA (miRNA) targets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Gzipped Data from https://dianalab.e-ce.uth.gr/downloads/tarbase_v8_data.tar.gz\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneId</th>\n",
       "      <th>geneName</th>\n",
       "      <th>mirna</th>\n",
       "      <th>species</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>tissue</th>\n",
       "      <th>category</th>\n",
       "      <th>method</th>\n",
       "      <th>positive_negative</th>\n",
       "      <th>direct_indirect</th>\n",
       "      <th>up_down</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0910001A06Rik(mmu)</td>\n",
       "      <td>0910001A06Rik(mmu)</td>\n",
       "      <td>mmu-miR-124-3p</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>UP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1200004M23Rik(mmu)</td>\n",
       "      <td>1200004M23Rik(mmu)</td>\n",
       "      <td>mmu-miR-124-3p</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1700027J05Rik(mmu)</td>\n",
       "      <td>1700027J05Rik(mmu)</td>\n",
       "      <td>mmu-miR-124-3p</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>UP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1810015A11Rik(mmu)</td>\n",
       "      <td>1810015A11Rik(mmu)</td>\n",
       "      <td>mmu-miR-124-3p</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2310047A01Rik(mmu)</td>\n",
       "      <td>2310047A01Rik(mmu)</td>\n",
       "      <td>mmu-miR-124-3p</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927114</th>\n",
       "      <td>uPA(hsa)</td>\n",
       "      <td>uPA(hsa)</td>\n",
       "      <td>hsa-miR-23b-3p</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Western Blot</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927115</th>\n",
       "      <td>vimentin(hsa)</td>\n",
       "      <td>vimentin(hsa)</td>\n",
       "      <td>hsa-miR-9-5p</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Western Blot</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927116</th>\n",
       "      <td>Â Â Â Â (PTPRG)(hsa)</td>\n",
       "      <td>Â Â Â Â (PTPRG)(hsa)</td>\n",
       "      <td>hsa-miR-146a-5p</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>UP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927117</th>\n",
       "      <td>Î±-actin(mmu)</td>\n",
       "      <td>Î±-actin(mmu)</td>\n",
       "      <td>mmu-miR-24-3p</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>qPCR</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>UP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927118</th>\n",
       "      <td>Î±-actin(mmu)</td>\n",
       "      <td>Î±-actin(mmu)</td>\n",
       "      <td>mmu-miR-24-3p</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Western Blot</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>UP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>927119 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      geneId              geneName            mirna   \n",
       "0         0910001A06Rik(mmu)    0910001A06Rik(mmu)   mmu-miR-124-3p  \\\n",
       "1         1200004M23Rik(mmu)    1200004M23Rik(mmu)   mmu-miR-124-3p   \n",
       "2         1700027J05Rik(mmu)    1700027J05Rik(mmu)   mmu-miR-124-3p   \n",
       "3         1810015A11Rik(mmu)    1810015A11Rik(mmu)   mmu-miR-124-3p   \n",
       "4         2310047A01Rik(mmu)    2310047A01Rik(mmu)   mmu-miR-124-3p   \n",
       "...                      ...                   ...              ...   \n",
       "927114              uPA(hsa)              uPA(hsa)   hsa-miR-23b-3p   \n",
       "927115         vimentin(hsa)         vimentin(hsa)     hsa-miR-9-5p   \n",
       "927116  Â Â Â Â (PTPRG)(hsa)  Â Â Â Â (PTPRG)(hsa)  hsa-miR-146a-5p   \n",
       "927117         Î±-actin(mmu)         Î±-actin(mmu)    mmu-miR-24-3p   \n",
       "927118         Î±-actin(mmu)         Î±-actin(mmu)    mmu-miR-24-3p   \n",
       "\n",
       "             species cell_line tissue          category        method   \n",
       "0       Mus musculus      <NA>    NaN  Cancer/Malignant   Microarrays  \\\n",
       "1       Mus musculus      <NA>    NaN  Cancer/Malignant   Microarrays   \n",
       "2       Mus musculus      <NA>    NaN  Cancer/Malignant   Microarrays   \n",
       "3       Mus musculus      <NA>    NaN  Cancer/Malignant   Microarrays   \n",
       "4       Mus musculus      <NA>    NaN  Cancer/Malignant   Microarrays   \n",
       "...              ...       ...    ...               ...           ...   \n",
       "927114  Homo sapiens      <NA>    NaN  Cancer/Malignant  Western Blot   \n",
       "927115  Homo sapiens      <NA>    NaN  Cancer/Malignant  Western Blot   \n",
       "927116  Homo sapiens      <NA>    NaN  Cancer/Malignant   Microarrays   \n",
       "927117  Mus musculus      <NA>    NaN  Cancer/Malignant          qPCR   \n",
       "927118  Mus musculus      <NA>    NaN  Cancer/Malignant  Western Blot   \n",
       "\n",
       "       positive_negative direct_indirect up_down condition  \n",
       "0               POSITIVE        INDIRECT      UP       NaN  \n",
       "1               POSITIVE        INDIRECT    DOWN       NaN  \n",
       "2               POSITIVE        INDIRECT      UP       NaN  \n",
       "3               POSITIVE        INDIRECT    DOWN       NaN  \n",
       "4               POSITIVE        INDIRECT    DOWN       NaN  \n",
       "...                  ...             ...     ...       ...  \n",
       "927114          POSITIVE        INDIRECT    DOWN       NaN  \n",
       "927115          NEGATIVE        INDIRECT     NaN       NaN  \n",
       "927116          POSITIVE        INDIRECT      UP       NaN  \n",
       "927117          POSITIVE        INDIRECT      UP       NaN  \n",
       "927118          POSITIVE        INDIRECT      UP       NaN  \n",
       "\n",
       "[927119 rows x 12 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_downloader('https://dianalab.e-ce.uth.gr/downloads/tarbase_v8_data.tar.gz', edge_data_location)\n",
    "\n",
    "with tarfile.TarFile(edge_data_location+'tarbase_v8_data.tar', 'r') as tar_ref:\n",
    "    tar_ref.extractall(edge_data_location)\n",
    "    \n",
    "# Remove tar file\n",
    "os.remove(edge_data_location+'tarbase_v8_data.tar')\n",
    "    \n",
    "# Rename file adding relationship's identifier\n",
    "os.rename(edge_data_location+'TarBase_v8_download.txt',\n",
    "          edge_data_location+'gene-miRNA_TarBase_v8_download.txt')   \n",
    "\n",
    "with open(edge_data_location + 'gene-miRNA_TarBase_v8_download.txt') as f:\n",
    "    data = f.read()\n",
    "\n",
    "data = pd.read_csv(edge_data_location + 'gene-miRNA_TarBase_v8_download.txt', sep=\"\\t\", dtype={\"cell_line\": \"string\"})  \n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneId</th>\n",
       "      <th>geneName</th>\n",
       "      <th>mirna</th>\n",
       "      <th>species</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>tissue</th>\n",
       "      <th>category</th>\n",
       "      <th>method</th>\n",
       "      <th>positive_negative</th>\n",
       "      <th>direct_indirect</th>\n",
       "      <th>up_down</th>\n",
       "      <th>condition</th>\n",
       "      <th>p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23496</th>\n",
       "      <td>ENSG00000044446</td>\n",
       "      <td>PHKA2</td>\n",
       "      <td>hsa-mir-766</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>293S</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>Embryonic/Fetal</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>no treatment (control)</td>\n",
       "      <td>5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>439409</th>\n",
       "      <td>ENSG00000166946</td>\n",
       "      <td>CCNDBP1</td>\n",
       "      <td>hsa-let-7b</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HELA</td>\n",
       "      <td>Cervix</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>8hrs post-transfection, Overexpression</td>\n",
       "      <td>5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202972</th>\n",
       "      <td>ENSG00000119866</td>\n",
       "      <td>BCL11A</td>\n",
       "      <td>hsa-mir-30a</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>TZMBL</td>\n",
       "      <td>Cervix</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>PAR-CLIP</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>shRNAs against HIV-1</td>\n",
       "      <td>5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112380</th>\n",
       "      <td>ENSG00000103495</td>\n",
       "      <td>MAZ</td>\n",
       "      <td>hsa-mir-29b</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>293S</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>Embryonic/Fetal</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>treatment: arsenite</td>\n",
       "      <td>3p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322689</th>\n",
       "      <td>ENSG00000143499</td>\n",
       "      <td>SMYD2</td>\n",
       "      <td>hsa-mir-301b</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HEK293</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>Embryonic/Fetal</td>\n",
       "      <td>PAR-CLIP</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>mild MNase digestion</td>\n",
       "      <td>3p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198909</th>\n",
       "      <td>ENSG00000119048</td>\n",
       "      <td>UBE2B</td>\n",
       "      <td>hsa-mir-20a</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HEK293</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>Embryonic/Fetal</td>\n",
       "      <td>PAR-CLIP</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>mild MNase digestion</td>\n",
       "      <td>5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27773</th>\n",
       "      <td>ENSG00000052802</td>\n",
       "      <td>MSMO1</td>\n",
       "      <td>hsa-mir-23a</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>293S</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>Embryonic/Fetal</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>treatment: emetine</td>\n",
       "      <td>3p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346983</th>\n",
       "      <td>ENSG00000148730</td>\n",
       "      <td>EIF4EBP2</td>\n",
       "      <td>hsa-mir-22</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HUVEC</td>\n",
       "      <td>Umbilical Vein</td>\n",
       "      <td>Normal/Primary</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297095</th>\n",
       "      <td>ENSG00000138757</td>\n",
       "      <td>G3BP2</td>\n",
       "      <td>hsa-mir-27a</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>BETA Cells</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Normal/Primary</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>560015</th>\n",
       "      <td>ENSG00000197965</td>\n",
       "      <td>MPZL1</td>\n",
       "      <td>hsa-mir-24</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>293S</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>Embryonic/Fetal</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>treatment: emetine</td>\n",
       "      <td>3p</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>50000 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 geneId  geneName         mirna       species   cell_line   \n",
       "23496   ENSG00000044446     PHKA2   hsa-mir-766  Homo sapiens        293S  \\\n",
       "439409  ENSG00000166946   CCNDBP1    hsa-let-7b  Homo sapiens        HELA   \n",
       "202972  ENSG00000119866    BCL11A   hsa-mir-30a  Homo sapiens       TZMBL   \n",
       "112380  ENSG00000103495       MAZ   hsa-mir-29b  Homo sapiens        293S   \n",
       "322689  ENSG00000143499     SMYD2  hsa-mir-301b  Homo sapiens      HEK293   \n",
       "...                 ...       ...           ...           ...         ...   \n",
       "198909  ENSG00000119048     UBE2B   hsa-mir-20a  Homo sapiens      HEK293   \n",
       "27773   ENSG00000052802     MSMO1   hsa-mir-23a  Homo sapiens        293S   \n",
       "346983  ENSG00000148730  EIF4EBP2    hsa-mir-22  Homo sapiens       HUVEC   \n",
       "297095  ENSG00000138757     G3BP2   hsa-mir-27a  Homo sapiens  BETA Cells   \n",
       "560015  ENSG00000197965     MPZL1    hsa-mir-24  Homo sapiens        293S   \n",
       "\n",
       "                tissue          category       method positive_negative   \n",
       "23496           Kidney   Embryonic/Fetal    HITS-CLIP          POSITIVE  \\\n",
       "439409          Cervix  Cancer/Malignant  Microarrays          POSITIVE   \n",
       "202972          Cervix  Cancer/Malignant     PAR-CLIP          POSITIVE   \n",
       "112380          Kidney   Embryonic/Fetal    HITS-CLIP          POSITIVE   \n",
       "322689          Kidney   Embryonic/Fetal     PAR-CLIP          POSITIVE   \n",
       "...                ...               ...          ...               ...   \n",
       "198909          Kidney   Embryonic/Fetal     PAR-CLIP          POSITIVE   \n",
       "27773           Kidney   Embryonic/Fetal    HITS-CLIP          POSITIVE   \n",
       "346983  Umbilical Vein    Normal/Primary    HITS-CLIP          POSITIVE   \n",
       "297095        Pancreas    Normal/Primary    HITS-CLIP          POSITIVE   \n",
       "560015          Kidney   Embryonic/Fetal    HITS-CLIP          POSITIVE   \n",
       "\n",
       "       direct_indirect up_down                               condition   p  \n",
       "23496           DIRECT    DOWN                  no treatment (control)  5p  \n",
       "439409        INDIRECT    DOWN  8hrs post-transfection, Overexpression  5p  \n",
       "202972          DIRECT    DOWN                    shRNAs against HIV-1  5p  \n",
       "112380          DIRECT    DOWN                     treatment: arsenite  3p  \n",
       "322689          DIRECT    DOWN                    mild MNase digestion  3p  \n",
       "...                ...     ...                                     ...  ..  \n",
       "198909          DIRECT    DOWN                    mild MNase digestion  5p  \n",
       "27773           DIRECT    DOWN                      treatment: emetine  3p  \n",
       "346983          DIRECT    DOWN                                     NaN  3p  \n",
       "297095          DIRECT    DOWN                                     NaN  3p  \n",
       "560015          DIRECT    DOWN                      treatment: emetine  3p  \n",
       "\n",
       "[50000 rows x 13 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For the time being, we keep only Homo sapiens rows\n",
    "data = data[data['species'].str.contains(\"Homo sapiens\")]\n",
    "\n",
    "# Moreover, we keep only 50k (random) rows to reduce input size\n",
    "data = data.sample(n=50000)\n",
    "\n",
    "# This simplified KG ignores if a transcript is \"3p\" or \"5p\", so we store this information as additional column\n",
    "data['p'] = data[data['mirna'].str.contains(\"p\")]['mirna']\n",
    "data[\"p\"] = data[\"p\"].str[-2:]\n",
    "data['mirna'] = data['mirna'].str.replace(r'-[35]p$', '', regex=True)\n",
    "data['mirna'] = data['mirna'].str.lower()\n",
    "\n",
    "# If you're interested in understanding why Homo sapiens has -3p and -5p miRNAs:\n",
    "# https://pubmed.ncbi.nlm.nih.gov/12592000/\n",
    "# Putting aside the -(3/5)p information, we are essentially dealing with non-mature (aka hairpin) miRNA:\n",
    "# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3616697/\n",
    "\n",
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this case, we have only symbols and ENSG identifiers (<tt>geneId</tt> and <tt>geneName</tt> columns) for identifying genes, we'll require an additional mapping to link these symbols or ENSG to NCBI Entrez Gene identifiers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "data.to_csv(edge_data_location + 'gene-miRNA_TarBase_v8_download.txt', header=None, sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### miRNA-disease from [miR2Disease](http://watson.compbio.iupui.edu:8080/miR2Disease/)\n",
    "miR2Disease, a manually curated database, aims at providing a comprehensive resource of miRNA deregulation in various human diseases."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Data from http://watson.compbio.iupui.edu:8080/miR2Disease/download/AllEntries.txt\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7f-2</td>\n",
       "      <td>kidney cancer</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>microarray</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>Micro-RNA profiling in kidney and bladder canc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-let-7g</td>\n",
       "      <td>hepatocellular carcinoma (HCC)</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>Identification of metastasis-related microRNAs...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7g</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>The let-7 microRNA represses cell proliferatio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-let-7g</td>\n",
       "      <td>non-small cell lung cancer (NSCLC)</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>Suppression of non-small cell lung tumor devel...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-let-7g</td>\n",
       "      <td>ovarian cancer (OC)</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>Let-7 expression defines two differentiation s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2897</th>\n",
       "      <td>hsa-miR-21</td>\n",
       "      <td>glioblastoma multiforme (GBM)</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>miR-124 and miR-137 inhibit proliferation of g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2898</th>\n",
       "      <td>hsa-miR-21</td>\n",
       "      <td>glioma</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>MicroRNA 21 promotes glioma invasion by target...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2899</th>\n",
       "      <td>hsa-miR-21</td>\n",
       "      <td>hepatocellular carcinoma (HCC)</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>Downregulation of miR-122 in the rodent and hu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2900</th>\n",
       "      <td>hsa-miR-21</td>\n",
       "      <td>Inclusion body myositis (IBM)</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>microarray</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>Distinctive patterns of microRNA expression in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2901</th>\n",
       "      <td>hsa-miR-21</td>\n",
       "      <td>limb-girdle muscular dystrophies types 2A (LGM...</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>microarray</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>Distinctive patterns of microRNA expression in...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2902 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 0                                                  1   \n",
       "0     hsa-let-7f-2                                      kidney cancer  \\\n",
       "1       hsa-let-7g                     hepatocellular carcinoma (HCC)   \n",
       "2       hsa-let-7g                                        lung cancer   \n",
       "3       hsa-let-7g                 non-small cell lung cancer (NSCLC)   \n",
       "4       hsa-let-7g                                ovarian cancer (OC)   \n",
       "...            ...                                                ...   \n",
       "2897    hsa-miR-21                      glioblastoma multiforme (GBM)   \n",
       "2898    hsa-miR-21                                             glioma   \n",
       "2899    hsa-miR-21                     hepatocellular carcinoma (HCC)   \n",
       "2900    hsa-miR-21                      Inclusion body myositis (IBM)   \n",
       "2901    hsa-miR-21  limb-girdle muscular dystrophies types 2A (LGM...   \n",
       "\n",
       "                   2                           3       4   \n",
       "0       up-regulated                  microarray  2007.0  \\\n",
       "1     down-regulated  Northern blot, qRT-PCR etc  2008.0   \n",
       "2     down-regulated  Northern blot, qRT-PCR etc  2007.0   \n",
       "3     down-regulated  Northern blot, qRT-PCR etc  2008.0   \n",
       "4     down-regulated  Northern blot, qRT-PCR etc  2007.0   \n",
       "...              ...                         ...     ...   \n",
       "2897    up-regulated  Northern blot, qRT-PCR etc  2008.0   \n",
       "2898    up-regulated  Northern blot, qRT-PCR etc  2008.0   \n",
       "2899    up-regulated  Northern blot, qRT-PCR etc  2006.0   \n",
       "2900    up-regulated                  microarray  2007.0   \n",
       "2901    up-regulated                  microarray  2007.0   \n",
       "\n",
       "                                                      5  \n",
       "0     Micro-RNA profiling in kidney and bladder canc...  \n",
       "1     Identification of metastasis-related microRNAs...  \n",
       "2     The let-7 microRNA represses cell proliferatio...  \n",
       "3     Suppression of non-small cell lung tumor devel...  \n",
       "4     Let-7 expression defines two differentiation s...  \n",
       "...                                                 ...  \n",
       "2897  miR-124 and miR-137 inhibit proliferation of g...  \n",
       "2898  MicroRNA 21 promotes glioma invasion by target...  \n",
       "2899  Downregulation of miR-122 in the rodent and hu...  \n",
       "2900  Distinctive patterns of microRNA expression in...  \n",
       "2901  Distinctive patterns of microRNA expression in...  \n",
       "\n",
       "[2902 rows x 6 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_downloader('http://watson.compbio.iupui.edu:8080/miR2Disease/download/AllEntries.txt', edge_data_location)\n",
    "\n",
    "data = pd.read_csv(edge_data_location + 'AllEntries.txt', sep=\"\\t\", header=None)  \n",
    "os.remove(edge_data_location + 'AllEntries.txt')\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Data from http://watson.compbio.iupui.edu:8080/miR2Disease/download/diseaseList.txt\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease name in original paper</th>\n",
       "      <th>disease ontology ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Abdominal Aortic Aneurysm</td>\n",
       "      <td>DOID:7693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>acute lymphoblastic leukemia (ALL)</td>\n",
       "      <td>DOID:9952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>acute myeloid leukemia (AML)</td>\n",
       "      <td>DOID:9119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>acute myocardial infarction</td>\n",
       "      <td>DOID:9408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>acute promyelocytic leukemia (APL)</td>\n",
       "      <td>DOID:9119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>uterine leiomyoma (ULM)</td>\n",
       "      <td>DOID:13223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>uveal melanoma</td>\n",
       "      <td>DOID:1909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>vascular disease</td>\n",
       "      <td>DOID:178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>vesicular stomatitis</td>\n",
       "      <td>DOID:10881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>Waldenstrom Macroglobulinemia (WM)</td>\n",
       "      <td>DOID:9080</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>174 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         disease name in original paper disease ontology ID\n",
       "0             Abdominal Aortic Aneurysm           DOID:7693\n",
       "1    acute lymphoblastic leukemia (ALL)           DOID:9952\n",
       "2          acute myeloid leukemia (AML)           DOID:9119\n",
       "3           acute myocardial infarction           DOID:9408\n",
       "4    acute promyelocytic leukemia (APL)           DOID:9119\n",
       "..                                  ...                 ...\n",
       "169             uterine leiomyoma (ULM)          DOID:13223\n",
       "170                      uveal melanoma           DOID:1909\n",
       "171                    vascular disease            DOID:178\n",
       "172                vesicular stomatitis          DOID:10881\n",
       "173  Waldenstrom Macroglobulinemia (WM)           DOID:9080\n",
       "\n",
       "[174 rows x 2 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# miR2Disease provides a look-up table for mapping disease names to DO terms\n",
    "data_downloader('http://watson.compbio.iupui.edu:8080/miR2Disease/download/diseaseList.txt', processed_data_location)\n",
    "\n",
    "descDOmap = pd.read_csv(processed_data_location + 'diseaseList.txt', sep=\"\\t\")  \n",
    "os.remove(processed_data_location + 'diseaseList.txt')\n",
    "descDOmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease ontology ID</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DOID_9952</td>\n",
       "      <td>hsa-let-7e</td>\n",
       "      <td>acute lymphoblastic leukemia (ALL)</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>microarray</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>MicroRNA expression signatures accurately disc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DOID_9952</td>\n",
       "      <td>hsa-mir-125a</td>\n",
       "      <td>acute lymphoblastic leukemia (ALL)</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>microarray</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>MicroRNA expression signatures accurately disc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DOID_9952</td>\n",
       "      <td>hsa-mir-151*</td>\n",
       "      <td>acute lymphoblastic leukemia (ALL)</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>microarray</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>MicroRNA expression signatures accurately disc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DOID_9952</td>\n",
       "      <td>hsa-mir-210</td>\n",
       "      <td>acute lymphoblastic leukemia (ALL)</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>microarray</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>MicroRNA expression signatures accurately disc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DOID_9952</td>\n",
       "      <td>hsa-mir-22</td>\n",
       "      <td>acute lymphoblastic leukemia (ALL)</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>northern blot, qRT-PCR etc</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>Gene silencing of MIR22 in acute lymphoblastic...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2619</th>\n",
       "      <td>DOID_9080</td>\n",
       "      <td>hsa-mir-542-3p</td>\n",
       "      <td>Waldenstrom Macroglobulinemia (WM)</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>microRNA expression in the biology, prognosis ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2620</th>\n",
       "      <td>DOID_9080</td>\n",
       "      <td>hsa-mir-9*</td>\n",
       "      <td>Waldenstrom Macroglobulinemia (WM)</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>microRNA expression in the biology, prognosis ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2621</th>\n",
       "      <td>DOID_9080</td>\n",
       "      <td>hsa-mir-155</td>\n",
       "      <td>Waldenstrom Macroglobulinemia (WM)</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>microRNA expression in the biology, prognosis ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2622</th>\n",
       "      <td>DOID_9080</td>\n",
       "      <td>hsa-mir-184</td>\n",
       "      <td>Waldenstrom Macroglobulinemia (WM)</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>microRNA expression in the biology, prognosis ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2623</th>\n",
       "      <td>DOID_9080</td>\n",
       "      <td>hsa-mir-206</td>\n",
       "      <td>Waldenstrom Macroglobulinemia (WM)</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>microRNA expression in the biology, prognosis ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2624 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     disease ontology ID               0                                   1   \n",
       "0              DOID_9952      hsa-let-7e  acute lymphoblastic leukemia (ALL)  \\\n",
       "1              DOID_9952    hsa-mir-125a  acute lymphoblastic leukemia (ALL)   \n",
       "2              DOID_9952    hsa-mir-151*  acute lymphoblastic leukemia (ALL)   \n",
       "3              DOID_9952     hsa-mir-210  acute lymphoblastic leukemia (ALL)   \n",
       "4              DOID_9952      hsa-mir-22  acute lymphoblastic leukemia (ALL)   \n",
       "...                  ...             ...                                 ...   \n",
       "2619           DOID_9080  hsa-mir-542-3p  Waldenstrom Macroglobulinemia (WM)   \n",
       "2620           DOID_9080      hsa-mir-9*  Waldenstrom Macroglobulinemia (WM)   \n",
       "2621           DOID_9080     hsa-mir-155  Waldenstrom Macroglobulinemia (WM)   \n",
       "2622           DOID_9080     hsa-mir-184  Waldenstrom Macroglobulinemia (WM)   \n",
       "2623           DOID_9080     hsa-mir-206  Waldenstrom Macroglobulinemia (WM)   \n",
       "\n",
       "                   2                           3       4   \n",
       "0     down-regulated                  microarray  2007.0  \\\n",
       "1     down-regulated                  microarray  2007.0   \n",
       "2       up-regulated                  microarray  2007.0   \n",
       "3       up-regulated                  microarray  2007.0   \n",
       "4     down-regulated  northern blot, qRT-PCR etc  2009.0   \n",
       "...              ...                         ...     ...   \n",
       "2619    up-regulated  Northern blot, qRT-PCR etc  2009.0   \n",
       "2620  down-regulated  Northern blot, qRT-PCR etc  2009.0   \n",
       "2621    up-regulated  Northern blot, qRT-PCR etc  2009.0   \n",
       "2622    up-regulated  Northern blot, qRT-PCR etc  2009.0   \n",
       "2623    up-regulated  Northern blot, qRT-PCR etc  2009.0   \n",
       "\n",
       "                                                      5  \n",
       "0     MicroRNA expression signatures accurately disc...  \n",
       "1     MicroRNA expression signatures accurately disc...  \n",
       "2     MicroRNA expression signatures accurately disc...  \n",
       "3     MicroRNA expression signatures accurately disc...  \n",
       "4     Gene silencing of MIR22 in acute lymphoblastic...  \n",
       "...                                                 ...  \n",
       "2619  microRNA expression in the biology, prognosis ...  \n",
       "2620  microRNA expression in the biology, prognosis ...  \n",
       "2621  microRNA expression in the biology, prognosis ...  \n",
       "2622  microRNA expression in the biology, prognosis ...  \n",
       "2623  microRNA expression in the biology, prognosis ...  \n",
       "\n",
       "[2624 rows x 7 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ontologies are represented in OWL files that make use of _ for URIs\n",
    "descDOmap['disease ontology ID'] = descDOmap['disease ontology ID'].astype(str).str.replace(':', '_')\n",
    "\n",
    "disease2mirna = pd.merge(descDOmap, data, left_on=['disease name in original paper'], right_on=[1]).drop(\n",
    "    columns=['disease name in original paper'])\n",
    "disease2mirna[0] = disease2mirna[0].str.lower()\n",
    "disease2mirna"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We lost 2,902-2,624=278 rows during mapping (278/2,902<10%)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "disease2mirna.to_csv(edge_data_location + 'miRNA-disease_miR2Disease.txt', header=None, sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "***\n",
    "### CREATE MAPPING DATASETS  <a class=\"anchor\" id=\"create-identifier-maps\"></a>\n",
    "***\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ensembl Gene-Entrez Gene <a class=\"anchor\" id=\"ensemblgene-entrezgene\"></a>\n",
    "\n",
    "\n",
    "**Purpose:** To map Ensembl gene identifiers to Entrez gene identifiers\n",
    "\n",
    "**Output:** `ENSEMBL_GENE_ENTREZ_GENE_MAP.txt`\n",
    "\n",
    "Already provided by PKT ecosystem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Data from https://storage.googleapis.com/pheknowlator/current_build/data/processed_data/ENSEMBL_GENE_ENTREZ_GENE_MAP.txt\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000171241</td>\n",
       "      <td>79801</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000131149</td>\n",
       "      <td>23199</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000096092</td>\n",
       "      <td>28978</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000222691</td>\n",
       "      <td>106479891</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>not protein-coding</td>\n",
       "      <td>not protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000230052</td>\n",
       "      <td>100873180</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>not protein-coding</td>\n",
       "      <td>not protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42283</th>\n",
       "      <td>ENSG00000175699</td>\n",
       "      <td>256369</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42284</th>\n",
       "      <td>ENSG00000251308</td>\n",
       "      <td>359776</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>not protein-coding</td>\n",
       "      <td>not protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42285</th>\n",
       "      <td>ENSG00000108479</td>\n",
       "      <td>2584</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42286</th>\n",
       "      <td>ENSG00000167371</td>\n",
       "      <td>112476</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42287</th>\n",
       "      <td>ENSG00000228938</td>\n",
       "      <td>119358</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>not protein-coding</td>\n",
       "      <td>not protein-coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>42288 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     0          1                       2               3   \n",
       "0      ENSG00000171241      79801          protein-coding  protein-coding  \\\n",
       "1      ENSG00000131149      23199          protein-coding  protein-coding   \n",
       "2      ENSG00000096092      28978          protein-coding  protein-coding   \n",
       "3      ENSG00000222691  106479891                   snRNA      pseudogene   \n",
       "4      ENSG00000230052  100873180  unprocessed_pseudogene      pseudogene   \n",
       "...                ...        ...                     ...             ...   \n",
       "42283  ENSG00000175699     256369          protein-coding  protein-coding   \n",
       "42284  ENSG00000251308     359776    processed_pseudogene      pseudogene   \n",
       "42285  ENSG00000108479       2584          protein-coding  protein-coding   \n",
       "42286  ENSG00000167371     112476          protein-coding  protein-coding   \n",
       "42287  ENSG00000228938     119358    processed_pseudogene      pseudogene   \n",
       "\n",
       "                        4                   5  \n",
       "0          protein-coding      protein-coding  \n",
       "1          protein-coding      protein-coding  \n",
       "2          protein-coding      protein-coding  \n",
       "3      not protein-coding  not protein-coding  \n",
       "4      not protein-coding  not protein-coding  \n",
       "...                   ...                 ...  \n",
       "42283      protein-coding      protein-coding  \n",
       "42284  not protein-coding  not protein-coding  \n",
       "42285      protein-coding      protein-coding  \n",
       "42286      protein-coding      protein-coding  \n",
       "42287  not protein-coding  not protein-coding  \n",
       "\n",
       "[42288 rows x 6 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "download('ENSEMBL_GENE_ENTREZ_GENE_MAP.txt', processed_data_location)\n",
    "\n",
    "ensEntrez = pd.read_csv(processed_data_location + 'ENSEMBL_GENE_ENTREZ_GENE_MAP.txt', sep=\"\\t\", header=None)\n",
    "ensEntrez"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### DisGeNET-Mondo <a class=\"anchor\" id=\"DisGeNET-Mondo\"></a>\n",
    "\n",
    "\n",
    "**Purpose:** To map DisGeNET identifiers to Mondo identifiers\n",
    "\n",
    "**Output:** `DISEASE_MONDO_MAP.txt`\n",
    "\n",
    "Already provided by PKT ecosystem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Data from https://storage.googleapis.com/pheknowlator/current_build/data/processed_data/DISEASE_MONDO_MAP.txt\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0001816</td>\n",
       "      <td>MONDO_0016982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0002116</td>\n",
       "      <td>MONDO_0005085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0014667</td>\n",
       "      <td>MONDO_0005066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0040084</td>\n",
       "      <td>MONDO_0005972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0040085</td>\n",
       "      <td>MONDO_0005229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182942</th>\n",
       "      <td>D017098</td>\n",
       "      <td>MONDO_0001341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182943</th>\n",
       "      <td>C131452</td>\n",
       "      <td>MONDO_0013874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182944</th>\n",
       "      <td>19995004</td>\n",
       "      <td>MONDO_0022013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182945</th>\n",
       "      <td>268173</td>\n",
       "      <td>MONDO_0017053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182946</th>\n",
       "      <td>C0152179</td>\n",
       "      <td>MONDO_0001535</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>182947 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0              1\n",
       "0        0001816  MONDO_0016982\n",
       "1        0002116  MONDO_0005085\n",
       "2        0014667  MONDO_0005066\n",
       "3        0040084  MONDO_0005972\n",
       "4        0040085  MONDO_0005229\n",
       "...          ...            ...\n",
       "182942   D017098  MONDO_0001341\n",
       "182943   C131452  MONDO_0013874\n",
       "182944  19995004  MONDO_0022013\n",
       "182945    268173  MONDO_0017053\n",
       "182946  C0152179  MONDO_0001535\n",
       "\n",
       "[182947 rows x 2 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "download('DISEASE_MONDO_MAP.txt', processed_data_location)\n",
    "\n",
    "disMondo = pd.read_csv(processed_data_location + 'DISEASE_MONDO_MAP.txt', sep=\"\\t\", header=None)\n",
    "disMondo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Disease Ontology (DO) - Mondo mapping <a class=\"anchor\" id=\"ensemblgene-entrezgene\"></a>\n",
    "\n",
    "\n",
    "**Purpose:** To map DO identifiers to Mondo identifiers\n",
    "\n",
    "**Output:** `DISEASE_DOID_MONDO_Map.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "mondo_graph = Graph().parse(ontology_data_location + 'mondo_with_imports.owl')\n",
    "\n",
    "dbxref_res = gets_ontology_class_dbxrefs(mondo_graph)[0]\n",
    "\n",
    "# Fix DOIDs (substitute : with _) and upper case them\n",
    "mondo_dict = {str(k).replace(':','_').upper(): {str(i).split('/')[-1].replace(':','_') for i in v}\n",
    "              for k, v in dbxref_res.items() if 'doid' in str(k)}\n",
    "\n",
    "with open(processed_data_location + 'DISEASE_DOID_MONDO_Map.txt', 'w') as outfile:\n",
    "    for k, v in mondo_dict.items():\n",
    "        outfile.write(str(k) + '\\t' + str(v).replace('{','').replace('\\'','').replace('}','') + '\\n')\n",
    " \n",
    "doidMondo = pd.read_csv(processed_data_location + 'DISEASE_DOID_MONDO_Map.txt', sep=\"\\t\", header=None)\n",
    "doidMondo[1] = doidMondo[1].str.split(',')\n",
    "doidMondo = doidMondo.explode(1)\n",
    "doidMondo.to_csv(processed_data_location + 'DISEASE_DOID_MONDO_Map.txt', header=None, sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DOID_0060231</td>\n",
       "      <td>MONDO_0017195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DOID_0081065</td>\n",
       "      <td>MONDO_0850469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DOID_3255</td>\n",
       "      <td>MONDO_0002578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DOID_6428</td>\n",
       "      <td>MONDO_0006132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DOID_0080563</td>\n",
       "      <td>MONDO_0012052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11281</th>\n",
       "      <td>DOID_0080627</td>\n",
       "      <td>MONDO_0008756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11282</th>\n",
       "      <td>DOID_6559</td>\n",
       "      <td>MONDO_0003923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11283</th>\n",
       "      <td>DOID_13810</td>\n",
       "      <td>MONDO_0005439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11284</th>\n",
       "      <td>DOID_999</td>\n",
       "      <td>MONDO_0015691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11285</th>\n",
       "      <td>DOID_0050649</td>\n",
       "      <td>MONDO_0008846</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11286 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  0              1\n",
       "0      DOID_0060231  MONDO_0017195\n",
       "1      DOID_0081065  MONDO_0850469\n",
       "2         DOID_3255  MONDO_0002578\n",
       "3         DOID_6428  MONDO_0006132\n",
       "4      DOID_0080563  MONDO_0012052\n",
       "...             ...            ...\n",
       "11281  DOID_0080627  MONDO_0008756\n",
       "11282     DOID_6559  MONDO_0003923\n",
       "11283    DOID_13810  MONDO_0005439\n",
       "11284      DOID_999  MONDO_0015691\n",
       "11285  DOID_0050649  MONDO_0008846\n",
       "\n",
       "[11286 rows x 2 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "doidMondo = pd.read_csv(processed_data_location + 'DISEASE_DOID_MONDO_Map.txt', sep=\"\\t\", header=None)\n",
    "doidMondo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### miRBase ID - miRBase accession\n",
    "\n",
    "\n",
    "**Purpose:** To map miRNA identifiers from miRBase to miRBase accession (it guarantees a standard identification of miRNA molecules via URIs)\n",
    "\n",
    "**Output:** `MIRBASE_ID_ACCESSION_MAP.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Data from https://www.mirbase.org/download/hsa.gff3\n",
      "##gff-version 3\n",
      "##date 2018-3-5\n",
      "#\n",
      "# Chromosomal coordinates of Homo sapiens microRNAs\n",
      "# microRNAs:               miRBase v22\n",
      "# genome-build-id:         GRCh38\n",
      "# genome-build-accession:  NCBI_Assembly:GCA_000001405.15\n",
      "#\n",
      "# Hairpin precursor sequences have type \"miRNA_primary_transcript\". \n",
      "# Note, these sequences do not represent the full primary transcript, \n",
      "# rather a predicted stem-loop portion that includes the precursor \n",
      "# miRNA. Mature sequences have type \"miRNA\".\n",
      "#\n",
      "\n",
      "     seq_id source                      type     start       end score strand   \n",
      "0      chr1      .  miRNA_primary_transcript     17369     17436     .      -  \\\n",
      "1      chr1      .                     miRNA     17409     17431     .      -   \n",
      "2      chr1      .                     miRNA     17369     17391     .      -   \n",
      "3      chr1      .  miRNA_primary_transcript     30366     30503     .      +   \n",
      "4      chr1      .                     miRNA     30438     30458     .      +   \n",
      "...     ...    ...                       ...       ...       ...   ...    ...   \n",
      "4796   chrY      .                     miRNA   2609229   2609252     .      +   \n",
      "4797   chrY      .  miRNA_primary_transcript   4606120   4606228     .      +   \n",
      "4798   chrY      .                     miRNA   4606140   4606158     .      +   \n",
      "4799   chrY      .  miRNA_primary_transcript  13479177  13479266     .      +   \n",
      "4800   chrY      .                     miRNA  13479189  13479214     .      +   \n",
      "\n",
      "     phase                                         attributes  \n",
      "0        .   ID=MI0022705;Alias=MI0022705;Name=hsa-mir-6859-1  \n",
      "1        .  ID=MIMAT0027618;Alias=MIMAT0027618;Name=hsa-mi...  \n",
      "2        .  ID=MIMAT0027619;Alias=MIMAT0027619;Name=hsa-mi...  \n",
      "3        .   ID=MI0006363;Alias=MI0006363;Name=hsa-mir-1302-2  \n",
      "4        .  ID=MIMAT0005890;Alias=MIMAT0005890;Name=hsa-mi...  \n",
      "...    ...                                                ...  \n",
      "4796     .  ID=MIMAT0023714_1;Alias=MIMAT0023714;Name=hsa-...  \n",
      "4797     .     ID=MI0032313;Alias=MI0032313;Name=hsa-mir-9985  \n",
      "4798     .  ID=MIMAT0039763;Alias=MIMAT0039763;Name=hsa-mi...  \n",
      "4799     .    ID=MI0039722;Alias=MI0039722;Name=hsa-mir-12120  \n",
      "4800     .  ID=MIMAT0049014;Alias=MIMAT0049014;Name=hsa-mi...  \n",
      "\n",
      "[4801 rows x 9 columns]\n"
     ]
    }
   ],
   "source": [
    "data_downloader('https://www.mirbase.org/download/hsa.gff3', processed_data_location)\n",
    "\n",
    "miRBaseMap = gffpd.read_gff3(processed_data_location + 'hsa.gff3')  \n",
    "os.remove(processed_data_location + 'hsa.gff3')\n",
    "print(miRBaseMap.header)\n",
    "print(miRBaseMap.df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ID=MI0022705;Alias=MI0022705;Name=hsa-mir-6859-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ID=MIMAT0027618;Alias=MIMAT0027618;Name=hsa-mi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ID=MIMAT0027619;Alias=MIMAT0027619;Name=hsa-mi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ID=MI0006363;Alias=MI0006363;Name=hsa-mir-1302-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ID=MIMAT0005890;Alias=MIMAT0005890;Name=hsa-mi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4796</th>\n",
       "      <td>ID=MIMAT0023714_1;Alias=MIMAT0023714;Name=hsa-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4797</th>\n",
       "      <td>ID=MI0032313;Alias=MI0032313;Name=hsa-mir-9985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4798</th>\n",
       "      <td>ID=MIMAT0039763;Alias=MIMAT0039763;Name=hsa-mi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4799</th>\n",
       "      <td>ID=MI0039722;Alias=MI0039722;Name=hsa-mir-12120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4800</th>\n",
       "      <td>ID=MIMAT0049014;Alias=MIMAT0049014;Name=hsa-mi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4801 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             attributes\n",
       "0      ID=MI0022705;Alias=MI0022705;Name=hsa-mir-6859-1\n",
       "1     ID=MIMAT0027618;Alias=MIMAT0027618;Name=hsa-mi...\n",
       "2     ID=MIMAT0027619;Alias=MIMAT0027619;Name=hsa-mi...\n",
       "3      ID=MI0006363;Alias=MI0006363;Name=hsa-mir-1302-2\n",
       "4     ID=MIMAT0005890;Alias=MIMAT0005890;Name=hsa-mi...\n",
       "...                                                 ...\n",
       "4796  ID=MIMAT0023714_1;Alias=MIMAT0023714;Name=hsa-...\n",
       "4797     ID=MI0032313;Alias=MI0032313;Name=hsa-mir-9985\n",
       "4798  ID=MIMAT0039763;Alias=MIMAT0039763;Name=hsa-mi...\n",
       "4799    ID=MI0039722;Alias=MI0039722;Name=hsa-mir-12120\n",
       "4800  ID=MIMAT0049014;Alias=MIMAT0049014;Name=hsa-mi...\n",
       "\n",
       "[4801 rows x 1 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "miRBaseMap = miRBaseMap.attributes_to_columns()\n",
    "miRBaseMap = miRBaseMap[['attributes']]\n",
    "miRBaseMap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>2</th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-mir-6859-1</td>\n",
       "      <td>MI0022705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-mir-1302-2</td>\n",
       "      <td>MI0006363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>hsa-mir-6859-2</td>\n",
       "      <td>MI0026420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>hsa-mir-12136</td>\n",
       "      <td>MI0039740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>hsa-mir-200b</td>\n",
       "      <td>MI0000342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4791</th>\n",
       "      <td>hsa-mir-1184-3</td>\n",
       "      <td>MI0015972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4793</th>\n",
       "      <td>hsa-mir-3690-2</td>\n",
       "      <td>MI0023561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4795</th>\n",
       "      <td>hsa-mir-6089-2</td>\n",
       "      <td>MI0023563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4797</th>\n",
       "      <td>hsa-mir-9985</td>\n",
       "      <td>MI0032313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4799</th>\n",
       "      <td>hsa-mir-12120</td>\n",
       "      <td>MI0039722</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1918 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   2          0\n",
       "0     hsa-mir-6859-1  MI0022705\n",
       "3     hsa-mir-1302-2  MI0006363\n",
       "5     hsa-mir-6859-2  MI0026420\n",
       "8      hsa-mir-12136  MI0039740\n",
       "10      hsa-mir-200b  MI0000342\n",
       "...              ...        ...\n",
       "4791  hsa-mir-1184-3  MI0015972\n",
       "4793  hsa-mir-3690-2  MI0023561\n",
       "4795  hsa-mir-6089-2  MI0023563\n",
       "4797    hsa-mir-9985  MI0032313\n",
       "4799   hsa-mir-12120  MI0039722\n",
       "\n",
       "[1918 rows x 2 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "miRBaseMap = miRBaseMap.attributes.str.split(';',expand=True)\n",
    "# Keep only \"ID\" and \"Name\" columns\n",
    "miRBaseMap = miRBaseMap[[2,0]]\n",
    "# Remove substring \"ID=\"\n",
    "miRBaseMap[0] = miRBaseMap[0].str[3:]\n",
    "# Remove substring \"Name=\"\n",
    "miRBaseMap[2] = miRBaseMap[2].str[5:]\n",
    "# Keep only hairpin/stem-loop miRNAs\n",
    "# (those starting with MI and not MIMAT, last one is reserved for mature sequences)\n",
    "miRBaseMap = miRBaseMap[~miRBaseMap[0].str.startswith('MIMAT')]\n",
    "miRBaseMap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "miRBaseMap.to_csv(processed_data_location+'MIRBASE_ID_ACCESSION_MAP.txt', header=None,sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "To represent genes, PKT designates them as subclasses of relevant Sequence Ontology ([SO](https://www.ebi.ac.uk/ols4/ontologies/so)) terms. We add miRNAs as subclasses of [SO_0000647](http://purl.obolibrary.org/obo/SO_0000647) (*miRNA_primary_transcript*)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Data from https://storage.googleapis.com/pheknowlator/current_build/data/processed_data/subclass_construction_map.pkl\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[('84103', ['SO_0001217']),\n",
       " ('84690', ['SO_0001217']),\n",
       " ('3579', ['SO_0001217']),\n",
       " ('54514', ['SO_0001217']),\n",
       " ('7159', ['SO_0001217'])]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# KG construction approach dictionary (for non ontological data), provided by PKT ecosystem\n",
    "download('subclass_construction_map.pkl', '../resources/construction_approach/')\n",
    "\n",
    "# Load data, print row count, and preview it\n",
    "nonO_data = pd.read_pickle(r'../resources/construction_approach/subclass_construction_map.pkl')\n",
    "\n",
    "# For instance, ncbi IDs are mapped to appropriate SO Ontology entries\n",
    "list(nonO_data.items())[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('R-HSA-977442', {'PW_0000848'}),\n",
       " ('R-HSA-977441', {'PW_0000848'}),\n",
       " ('R-HSA-6788656', {'PW_0000051', 'PW_0000052', 'PW_0000053', 'PW_0000075'}),\n",
       " ('R-HSA-168254', {'PW_0001054'}),\n",
       " ('R-HSA-71182', {'PW_0000052', 'PW_0000075'})]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(nonO_data.items())[-5:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('MI0015972', ['SO_0000647']),\n",
       " ('MI0023561', ['SO_0000647']),\n",
       " ('MI0023563', ['SO_0000647']),\n",
       " ('MI0032313', ['SO_0000647']),\n",
       " ('MI0039722', ['SO_0000647'])]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "miRBaseMap['SO'] = [['SO_0000647']] * len(miRBaseMap)\n",
    "\n",
    "mirna_nonO = miRBaseMap.drop(2, axis=1).set_index(0).to_dict()\n",
    "nonO_data = {**nonO_data, **mirna_nonO['SO']}\n",
    "\n",
    "list(nonO_data.items())[len(list(nonO_data.items()))-5:len(list(nonO_data.items()))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../resources/construction_approach/subclass_construction_map.pkl', 'wb') as handle:\n",
    "    pickle.dump(nonO_data, handle, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "PKT also provides node and relation metadata (a.k.a. node properties and relation type attributes)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Data from https://storage.googleapis.com/pheknowlator/current_build/data/processed_data/node_metadata_dict.pkl\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "dict_keys(['nodes', 'relations'])"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# KG metadata, provided by PKT ecosystem\n",
    "download('node_metadata_dict.pkl', '../resources/node_data/')\n",
    "\n",
    "# Load data, print row count, and preview it\n",
    "metadata = pd.read_pickle(r'../resources/node_data/node_metadata_dict.pkl')\n",
    "\n",
    "metadata.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'http://www.ncbi.nlm.nih.gov/gene/1': {'Label': 'A1BG',\n",
       "  'Description': \"A1BG has locus group 'protein-coding' and is located on chromosome 19 (19q13.43).\",\n",
       "  'Synonym': 'HEL-S-163pA|ABG|A1B|epididymis secretory sperm binding protein Li 163pA|GAB|HYST2477alpha-1B-glycoprotein'},\n",
       " 'http://www.ncbi.nlm.nih.gov/gene/2': {'Label': 'A2M',\n",
       "  'Description': \"A2M has locus group 'protein-coding' and is located on chromosome 12 (12p13.31).\",\n",
       "  'Synonym': 'CPAMD5|S863-7alpha-2-macroglobulin|FWP007|C3 and PZP-like alpha-2-macroglobulin domain-containing protein 5|A2MD|alpha-2-M'},\n",
       " 'http://www.ncbi.nlm.nih.gov/gene/3': {'Label': 'A2MP1',\n",
       "  'Description': \"A2MP1 has locus group 'pseudo' and is located on chromosome 12 (12p13.31).\",\n",
       "  'Synonym': 'A2MPpregnancy-zone protein pseudogene'},\n",
       " 'http://www.ncbi.nlm.nih.gov/gene/9': {'Label': 'NAT1',\n",
       "  'Description': \"NAT1 has locus group 'protein-coding' and is located on chromosome 8 (8p22).\",\n",
       "  'Synonym': 'NAT-1|N-acetyltransferase type 1|NATIarylamine N-acetyltransferase 1|arylamide acetylase 1|monomorphic arylamine N-acetyltransferase|MNAT|AAC1|N-acetyltransferase 1 (arylamine N-acetyltransferase)'},\n",
       " 'http://www.ncbi.nlm.nih.gov/gene/10': {'Label': 'NAT2',\n",
       "  'Description': \"NAT2 has locus group 'protein-coding' and is located on chromosome 8 (8p22).\",\n",
       "  'Synonym': 'AAC2|N-acetyltransferase type 2|PNATarylamine N-acetyltransferase 2|NAT-2|arylamide acetylase 2|N-acetyltransferase 2 (arylamine N-acetyltransferase)'}}"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "{k: metadata['nodes'][k] for k in list(metadata['nodes'])[:5]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'http://purl.obolibrary.org/obo/RO_0002533': {'Label': 'sequence atomic unit',\n",
       "  'Description': 'Any individual unit of a collection of like units arranged in a linear order',\n",
       "  'Synonym': 'None'},\n",
       " 'http://purl.obolibrary.org/obo/RO_0002577': {'Label': 'system',\n",
       "  'Description': 'A material entity consisting of multiple components that are causally integrated.',\n",
       "  'Synonym': 'None'},\n",
       " 'http://purl.obolibrary.org/obo/RO_0002534': {'Label': 'sequence bearer',\n",
       "  'Description': 'Any entity that can be divided into parts such that each part is an atomical unit of a sequence',\n",
       "  'Synonym': 'None'},\n",
       " 'http://purl.obolibrary.org/obo/RO_0002532': {'Label': 'sequentially ordered entity',\n",
       "  'Description': 'Any entity that is ordered in discrete units along a linear axis.',\n",
       "  'Synonym': 'None'},\n",
       " 'http://purl.obolibrary.org/obo/RO_0002310': {'Label': 'exposure event or process',\n",
       "  'Description': 'A process occurring within or in the vicinity of an organism that exerts some causal influence on the organism via the interaction between an exposure stimulus and an exposure receptor. The exposure stimulus may be a process, material entity or condition (for example, lack of nutrients). The exposure receptor can be an organism, organism population or a part of an organism.',\n",
       "  'Synonym': 'None'}}"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "{k: metadata['relations'][k] for k in list(metadata['relations'])[:5]}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can add miRNA properties (*Label*, *Description*, and *Synonym*) from miRBase."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Data from https://www.mirbase.org/download/miRNA.dat\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Description</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Comments</th>\n",
       "      <th>References</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MI0000001</td>\n",
       "      <td>Caenorhabditis elegans let-7 stem-loop</td>\n",
       "      <td>UACACUGUGGAUCCGGUGAGGUAGUAGGUUGUAUAGUUUGGAAUAU...</td>\n",
       "      <td>let-7 is found on chromosome X in Caenorhabdit...</td>\n",
       "      <td>([1], '11679671'), ([2], '12672692'), ([3], '1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MI0000002</td>\n",
       "      <td>Caenorhabditis elegans lin-4 stem-loop</td>\n",
       "      <td>AUGCUUCCGGCCUGUUCCCUGAGACCUCAAGUGUGAGUGUACUAUU...</td>\n",
       "      <td>lin-4 is found on chromosome II in Caenorhabdi...</td>\n",
       "      <td>([1], '11679671'), ([2], '10642801'), ([3], '1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MI0000003</td>\n",
       "      <td>Caenorhabditis elegans miR-1 stem-loop</td>\n",
       "      <td>AAAGUGACCGUACCGAGCUGCAUACUUCCUUACAUGCCCAUACUAU...</td>\n",
       "      <td>miR-1 was independently identified in C. elega...</td>\n",
       "      <td>([1], '11679671'), ([2], '11679672'), ([3], '1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MI0000004</td>\n",
       "      <td>Caenorhabditis elegans miR-2 stem-loop</td>\n",
       "      <td>UAAACAGUAUACAGAAAGCCAUCAAAGCGGUGGUUGAUGUGUUGCA...</td>\n",
       "      <td></td>\n",
       "      <td>([1], '11679671'), ([2], '11679672'), ([3], '1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MI0000005</td>\n",
       "      <td>Caenorhabditis elegans miR-34 stem-loop</td>\n",
       "      <td>CGGACAAUGCUCGAGAGGCAGUGUGGUUAGCUGGUUGCAUAUUUCC...</td>\n",
       "      <td></td>\n",
       "      <td>([1], '11679671'), ([2], '12672692'), ([3], '1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38584</th>\n",
       "      <td>MI0041070</td>\n",
       "      <td>Symbiodinium microadriaticum miR-12461 stem-loop</td>\n",
       "      <td>GAGGAUGCUGAUCAUUCACUGGCCCCCUGUGGACACGUGUGUUGCA...</td>\n",
       "      <td></td>\n",
       "      <td>([1], '24119094')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38585</th>\n",
       "      <td>MI0041071</td>\n",
       "      <td>Homo sapiens miR-9902-1 stem-loop</td>\n",
       "      <td>GCAGGGAAAGGGAACCCAGAAAUCUGGUAUGCCAGCAAAGAGAGUA...</td>\n",
       "      <td></td>\n",
       "      <td>([1], '25049417')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38586</th>\n",
       "      <td>MI0041072</td>\n",
       "      <td>Gallus gallus miR-1784b stem-loop</td>\n",
       "      <td>UUCUGCUCCUAUUUAAGUCAAUGGCAGAACUCUCACUGAUUUCAAU...</td>\n",
       "      <td></td>\n",
       "      <td>([1], '29079676')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38587</th>\n",
       "      <td>MI0041073</td>\n",
       "      <td>Monodelphis domestica miR-7385g-1 stem-loop</td>\n",
       "      <td>UAGUCUGAUAUUCCAUGUUUCUAUGUCAUGAAACUUGGAGCAUAGA...</td>\n",
       "      <td></td>\n",
       "      <td>([1], '29079676')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38588</th>\n",
       "      <td>MI0041074</td>\n",
       "      <td>Monodelphis domestica miR-7385g-2 stem-loop</td>\n",
       "      <td>UAGUCUGAUAUUCCAUGUUUCUAGGGUAUCAAGAGACUUGGAGCAU...</td>\n",
       "      <td></td>\n",
       "      <td>([1], '29079676')</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38589 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID                                       Description   \n",
       "0      MI0000001            Caenorhabditis elegans let-7 stem-loop  \\\n",
       "1      MI0000002            Caenorhabditis elegans lin-4 stem-loop   \n",
       "2      MI0000003            Caenorhabditis elegans miR-1 stem-loop   \n",
       "3      MI0000004            Caenorhabditis elegans miR-2 stem-loop   \n",
       "4      MI0000005           Caenorhabditis elegans miR-34 stem-loop   \n",
       "...          ...                                               ...   \n",
       "38584  MI0041070  Symbiodinium microadriaticum miR-12461 stem-loop   \n",
       "38585  MI0041071                 Homo sapiens miR-9902-1 stem-loop   \n",
       "38586  MI0041072                 Gallus gallus miR-1784b stem-loop   \n",
       "38587  MI0041073       Monodelphis domestica miR-7385g-1 stem-loop   \n",
       "38588  MI0041074       Monodelphis domestica miR-7385g-2 stem-loop   \n",
       "\n",
       "                                                Sequence   \n",
       "0      UACACUGUGGAUCCGGUGAGGUAGUAGGUUGUAUAGUUUGGAAUAU...  \\\n",
       "1      AUGCUUCCGGCCUGUUCCCUGAGACCUCAAGUGUGAGUGUACUAUU...   \n",
       "2      AAAGUGACCGUACCGAGCUGCAUACUUCCUUACAUGCCCAUACUAU...   \n",
       "3      UAAACAGUAUACAGAAAGCCAUCAAAGCGGUGGUUGAUGUGUUGCA...   \n",
       "4      CGGACAAUGCUCGAGAGGCAGUGUGGUUAGCUGGUUGCAUAUUUCC...   \n",
       "...                                                  ...   \n",
       "38584  GAGGAUGCUGAUCAUUCACUGGCCCCCUGUGGACACGUGUGUUGCA...   \n",
       "38585  GCAGGGAAAGGGAACCCAGAAAUCUGGUAUGCCAGCAAAGAGAGUA...   \n",
       "38586  UUCUGCUCCUAUUUAAGUCAAUGGCAGAACUCUCACUGAUUUCAAU...   \n",
       "38587  UAGUCUGAUAUUCCAUGUUUCUAUGUCAUGAAACUUGGAGCAUAGA...   \n",
       "38588  UAGUCUGAUAUUCCAUGUUUCUAGGGUAUCAAGAGACUUGGAGCAU...   \n",
       "\n",
       "                                                Comments   \n",
       "0      let-7 is found on chromosome X in Caenorhabdit...  \\\n",
       "1      lin-4 is found on chromosome II in Caenorhabdi...   \n",
       "2      miR-1 was independently identified in C. elega...   \n",
       "3                                                          \n",
       "4                                                          \n",
       "...                                                  ...   \n",
       "38584                                                      \n",
       "38585                                                      \n",
       "38586                                                      \n",
       "38587                                                      \n",
       "38588                                                      \n",
       "\n",
       "                                              References  \n",
       "0      ([1], '11679671'), ([2], '12672692'), ([3], '1...  \n",
       "1      ([1], '11679671'), ([2], '10642801'), ([3], '1...  \n",
       "2      ([1], '11679671'), ([2], '11679672'), ([3], '1...  \n",
       "3      ([1], '11679671'), ([2], '11679672'), ([3], '1...  \n",
       "4      ([1], '11679671'), ([2], '12672692'), ([3], '1...  \n",
       "...                                                  ...  \n",
       "38584                                  ([1], '24119094')  \n",
       "38585                                  ([1], '25049417')  \n",
       "38586                                  ([1], '29079676')  \n",
       "38587                                  ([1], '29079676')  \n",
       "38588                                  ([1], '29079676')  \n",
       "\n",
       "[38589 rows x 5 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "data_downloader('https://www.mirbase.org/download/miRNA.dat', processed_data_location)\n",
    "\n",
    "# Open the EMBL file\n",
    "embl_file = processed_data_location + 'miRNA.dat'\n",
    "\n",
    "# Create empty lists to store the data\n",
    "data = {\n",
    "    \"ID\": [],\n",
    "    \"Description\": [],\n",
    "    \"Sequence\": [],\n",
    "    \"Comments\": [],\n",
    "    \"References\": []\n",
    "}\n",
    "\n",
    "# Iterate through the records in the EMBL file\n",
    "for record in SeqIO.parse(embl_file, \"embl\"):\n",
    "    data[\"ID\"].append(record.id)\n",
    "    data[\"Description\"].append(record.description)\n",
    "    data[\"Sequence\"].append(str(record.seq))\n",
    "    data[\"Comments\"].append(str(record.annotations.get('comment', '')))\n",
    "    references = []\n",
    "    i = 0\n",
    "    for ref in record.annotations.get('references', []):\n",
    "        i = i + 1\n",
    "        references.append(f\"{[i], ref.pubmed_id}\")\n",
    "    data[\"References\"].append(\", \".join(references))\n",
    "\n",
    "df = pd.DataFrame(data)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Description</th>\n",
       "      <th>Synonym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>MI0000060</td>\n",
       "      <td>let-7a-3p cloned in [6] has a 1 nt 3' extensio...</td>\n",
       "      <td>Homo sapiens let-7a-1 stem-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>MI0000061</td>\n",
       "      <td>([1], '11679670'), ([2], '15183728'), ([3], '1...</td>\n",
       "      <td>Homo sapiens let-7a-2 stem-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>MI0000062</td>\n",
       "      <td>let-7a-3p cloned in [6] has a 1 nt 3' extensio...</td>\n",
       "      <td>Homo sapiens let-7a-3 stem-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>MI0000063</td>\n",
       "      <td>([1], '11679670'), ([2], '14573789'), ([3], '1...</td>\n",
       "      <td>Homo sapiens let-7b stem-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>MI0000064</td>\n",
       "      <td>([1], '11679670'), ([2], '14573789'), ([3], '1...</td>\n",
       "      <td>Homo sapiens let-7c stem-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37307</th>\n",
       "      <td>MI0039734</td>\n",
       "      <td>([1], '28471449'). Sequence: UUAACAUCUUUUCCAUC...</td>\n",
       "      <td>Homo sapiens miR-12132 stem-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37308</th>\n",
       "      <td>MI0039735</td>\n",
       "      <td>([1], '28471449'). Sequence: GAAGUGUACUUUUUAAU...</td>\n",
       "      <td>Homo sapiens miR-12133 stem-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37312</th>\n",
       "      <td>MI0039739</td>\n",
       "      <td>([1], '28640911'). Sequence: UGUGGAUAUUCUUUUUU...</td>\n",
       "      <td>Homo sapiens miR-12135 stem-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37313</th>\n",
       "      <td>MI0039740</td>\n",
       "      <td>([1], '28640911'). Sequence: GAAAAAGUCAUGGAGGC...</td>\n",
       "      <td>Homo sapiens miR-12136 stem-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38585</th>\n",
       "      <td>MI0041071</td>\n",
       "      <td>([1], '25049417'). Sequence: GCAGGGAAAGGGAACCC...</td>\n",
       "      <td>Homo sapiens miR-9902-1 stem-loop</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1917 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID                                        Description   \n",
       "57     MI0000060  let-7a-3p cloned in [6] has a 1 nt 3' extensio...  \\\n",
       "58     MI0000061  ([1], '11679670'), ([2], '15183728'), ([3], '1...   \n",
       "59     MI0000062  let-7a-3p cloned in [6] has a 1 nt 3' extensio...   \n",
       "60     MI0000063  ([1], '11679670'), ([2], '14573789'), ([3], '1...   \n",
       "61     MI0000064  ([1], '11679670'), ([2], '14573789'), ([3], '1...   \n",
       "...          ...                                                ...   \n",
       "37307  MI0039734  ([1], '28471449'). Sequence: UUAACAUCUUUUCCAUC...   \n",
       "37308  MI0039735  ([1], '28471449'). Sequence: GAAGUGUACUUUUUAAU...   \n",
       "37312  MI0039739  ([1], '28640911'). Sequence: UGUGGAUAUUCUUUUUU...   \n",
       "37313  MI0039740  ([1], '28640911'). Sequence: GAAAAAGUCAUGGAGGC...   \n",
       "38585  MI0041071  ([1], '25049417'). Sequence: GCAGGGAAAGGGAACCC...   \n",
       "\n",
       "                                 Synonym  \n",
       "57       Homo sapiens let-7a-1 stem-loop  \n",
       "58       Homo sapiens let-7a-2 stem-loop  \n",
       "59       Homo sapiens let-7a-3 stem-loop  \n",
       "60         Homo sapiens let-7b stem-loop  \n",
       "61         Homo sapiens let-7c stem-loop  \n",
       "...                                  ...  \n",
       "37307   Homo sapiens miR-12132 stem-loop  \n",
       "37308   Homo sapiens miR-12133 stem-loop  \n",
       "37312   Homo sapiens miR-12135 stem-loop  \n",
       "37313   Homo sapiens miR-12136 stem-loop  \n",
       "38585  Homo sapiens miR-9902-1 stem-loop  \n",
       "\n",
       "[1917 rows x 3 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reduce to \"ID\", \"Description\", and \"Synonym\" columns\n",
    "df = df[df['Description'].astype(str).str.contains('Homo sapiens')]\n",
    "df['Synonym'] = df['Description']\n",
    "df['Description'] = df['Comments'].astype(str) + df['References'].astype(str) + '. Sequence: ' + df['Sequence'].astype(str)\n",
    "df = df[['ID', 'Description', 'Synonym']]\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we add *Labels* through the map we have previously developed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Description</th>\n",
       "      <th>Synonym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MI0000060</td>\n",
       "      <td>hsa-let-7a-1</td>\n",
       "      <td>let-7a-3p cloned in [6] has a 1 nt 3' extensio...</td>\n",
       "      <td>Homo sapiens let-7a-1 stem-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MI0000061</td>\n",
       "      <td>hsa-let-7a-2</td>\n",
       "      <td>([1], '11679670'), ([2], '15183728'), ([3], '1...</td>\n",
       "      <td>Homo sapiens let-7a-2 stem-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MI0000062</td>\n",
       "      <td>hsa-let-7a-3</td>\n",
       "      <td>let-7a-3p cloned in [6] has a 1 nt 3' extensio...</td>\n",
       "      <td>Homo sapiens let-7a-3 stem-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MI0000063</td>\n",
       "      <td>hsa-let-7b</td>\n",
       "      <td>([1], '11679670'), ([2], '14573789'), ([3], '1...</td>\n",
       "      <td>Homo sapiens let-7b stem-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MI0000064</td>\n",
       "      <td>hsa-let-7c</td>\n",
       "      <td>([1], '11679670'), ([2], '14573789'), ([3], '1...</td>\n",
       "      <td>Homo sapiens let-7c stem-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1908</th>\n",
       "      <td>MI0039734</td>\n",
       "      <td>hsa-mir-12132</td>\n",
       "      <td>([1], '28471449'). Sequence: UUAACAUCUUUUCCAUC...</td>\n",
       "      <td>Homo sapiens miR-12132 stem-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1909</th>\n",
       "      <td>MI0039735</td>\n",
       "      <td>hsa-mir-12133</td>\n",
       "      <td>([1], '28471449'). Sequence: GAAGUGUACUUUUUAAU...</td>\n",
       "      <td>Homo sapiens miR-12133 stem-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1910</th>\n",
       "      <td>MI0039739</td>\n",
       "      <td>hsa-mir-12135</td>\n",
       "      <td>([1], '28640911'). Sequence: UGUGGAUAUUCUUUUUU...</td>\n",
       "      <td>Homo sapiens miR-12135 stem-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1911</th>\n",
       "      <td>MI0039740</td>\n",
       "      <td>hsa-mir-12136</td>\n",
       "      <td>([1], '28640911'). Sequence: GAAAAAGUCAUGGAGGC...</td>\n",
       "      <td>Homo sapiens miR-12136 stem-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1912</th>\n",
       "      <td>MI0041071</td>\n",
       "      <td>hsa-mir-9902-2</td>\n",
       "      <td>([1], '25049417'). Sequence: GCAGGGAAAGGGAACCC...</td>\n",
       "      <td>Homo sapiens miR-9902-1 stem-loop</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1913 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             ID           Label   \n",
       "0     MI0000060    hsa-let-7a-1  \\\n",
       "1     MI0000061    hsa-let-7a-2   \n",
       "2     MI0000062    hsa-let-7a-3   \n",
       "3     MI0000063      hsa-let-7b   \n",
       "4     MI0000064      hsa-let-7c   \n",
       "...         ...             ...   \n",
       "1908  MI0039734   hsa-mir-12132   \n",
       "1909  MI0039735   hsa-mir-12133   \n",
       "1910  MI0039739   hsa-mir-12135   \n",
       "1911  MI0039740   hsa-mir-12136   \n",
       "1912  MI0041071  hsa-mir-9902-2   \n",
       "\n",
       "                                            Description   \n",
       "0     let-7a-3p cloned in [6] has a 1 nt 3' extensio...  \\\n",
       "1     ([1], '11679670'), ([2], '15183728'), ([3], '1...   \n",
       "2     let-7a-3p cloned in [6] has a 1 nt 3' extensio...   \n",
       "3     ([1], '11679670'), ([2], '14573789'), ([3], '1...   \n",
       "4     ([1], '11679670'), ([2], '14573789'), ([3], '1...   \n",
       "...                                                 ...   \n",
       "1908  ([1], '28471449'). Sequence: UUAACAUCUUUUCCAUC...   \n",
       "1909  ([1], '28471449'). Sequence: GAAGUGUACUUUUUAAU...   \n",
       "1910  ([1], '28640911'). Sequence: UGUGGAUAUUCUUUUUU...   \n",
       "1911  ([1], '28640911'). Sequence: GAAAAAGUCAUGGAGGC...   \n",
       "1912  ([1], '25049417'). Sequence: GCAGGGAAAGGGAACCC...   \n",
       "\n",
       "                                Synonym  \n",
       "0       Homo sapiens let-7a-1 stem-loop  \n",
       "1       Homo sapiens let-7a-2 stem-loop  \n",
       "2       Homo sapiens let-7a-3 stem-loop  \n",
       "3         Homo sapiens let-7b stem-loop  \n",
       "4         Homo sapiens let-7c stem-loop  \n",
       "...                                 ...  \n",
       "1908   Homo sapiens miR-12132 stem-loop  \n",
       "1909   Homo sapiens miR-12133 stem-loop  \n",
       "1910   Homo sapiens miR-12135 stem-loop  \n",
       "1911   Homo sapiens miR-12136 stem-loop  \n",
       "1912  Homo sapiens miR-9902-1 stem-loop  \n",
       "\n",
       "[1913 rows x 4 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.merge(df, miRBaseMap, left_on=['ID'], right_on=[0])\n",
    "df['Label'] = df[2]\n",
    "df = df[['ID', 'Label', 'Description', 'Synonym']]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'https://www.mirbase.org/hairpin/MI0000060': {'Label': 'hsa-let-7a-1',\n",
       "  'Description': \"let-7a-3p cloned in [6] has a 1 nt 3' extension (U), which is incompatible\\nwith the genome sequence.([1], '11679670'), ([2], '15183728'), ([3], '12554860'), ([4], '14573789'), ([5], '15325244'), ([6], '17604727'), ([7], '17616659'), ([8], '17989717'), ([9], '20158877'). Sequence: UGGGAUGAGGUAGUAGGUUGUAUAGUUUUAGGGUCACACCCACCACUGGGAGAUAACUAUACAAUCUACUGUCUUUCCUA\",\n",
       "  'Synonym': 'Homo sapiens let-7a-1 stem-loop'},\n",
       " 'https://www.mirbase.org/hairpin/MI0000061': {'Label': 'hsa-let-7a-2',\n",
       "  'Description': \"([1], '11679670'), ([2], '15183728'), ([3], '12554860'), ([4], '14573789'), ([5], '15325244'), ([6], '17604727'), ([7], '17616659'), ([8], '17989717'), ([9], '19015728'), ([10], '20158877'). Sequence: AGGUUGAGGUAGUAGGUUGUAUAGUUUAGAAUUACAUCAAGGGAGAUAACUGUACAGCCUCCUAGCUUUCCU\",\n",
       "  'Synonym': 'Homo sapiens let-7a-2 stem-loop'},\n",
       " 'https://www.mirbase.org/hairpin/MI0000062': {'Label': 'hsa-let-7a-3',\n",
       "  'Description': \"let-7a-3p cloned in [6] has a 1 nt 3' extension (U), which is incompatible\\nwith the genome sequence.([1], '11679670'), ([2], '15183728'), ([3], '12554860'), ([4], '14573789'), ([5], '15325244'), ([6], '17604727'), ([7], '17616659'), ([8], '17989717'), ([9], '20158877'). Sequence: GGGUGAGGUAGUAGGUUGUAUAGUUUGGGGCUCUGCCCUGCUAUGGGAUAACUAUACAAUCUACUGUCUUUCCU\",\n",
       "  'Synonym': 'Homo sapiens let-7a-3 stem-loop'},\n",
       " 'https://www.mirbase.org/hairpin/MI0000063': {'Label': 'hsa-let-7b',\n",
       "  'Description': \"([1], '11679670'), ([2], '14573789'), ([3], '15325244'), ([4], '17604727'), ([5], '17616659'). Sequence: CGGGGUGAGGUAGUAGGUUGUGUGGUUUCAGGGCAGUGAUGUUGCCCCUCGGAAGAUAACUAUACAACCUACUGCCUUCCCUG\",\n",
       "  'Synonym': 'Homo sapiens let-7b stem-loop'},\n",
       " 'https://www.mirbase.org/hairpin/MI0000064': {'Label': 'hsa-let-7c',\n",
       "  'Description': \"([1], '11679670'), ([2], '14573789'), ([3], '17604727'), ([4], '17616659'), ([5], '20158877'), ([6], '23034410'). Sequence: GCAUCCGGGUUGAGGUAGUAGGUUGUAUGGUUUAGAGUUACACCCUGGGAGUUAACUGUACAACCUUCUAGCUUUCCUUGGAGC\",\n",
       "  'Synonym': 'Homo sapiens let-7c stem-loop'}}"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Convert the DataFrame to dictionary format\n",
    "miRNA_dict = {}\n",
    "for index, row in df.iterrows():\n",
    "    # Transform the ID to correspondent URI (KG's ID)\n",
    "    gene_id = f'https://www.mirbase.org/hairpin/{row[\"ID\"]}'\n",
    "    miRNA_dict[gene_id] = {\n",
    "        'Label': row['Label'],\n",
    "        'Description': row['Description'],\n",
    "        'Synonym': row['Synonym']\n",
    "    }\n",
    "    \n",
    "{k: miRNA_dict[k] for k in list(miRNA_dict)[:5]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_dict = {**metadata.get('nodes'), **miRNA_dict}\n",
    "nodes_final_dict = {'nodes': nodes_dict}\n",
    "\n",
    "rel_dict = {'relations': {**metadata.get('relations')}}\n",
    "metadata = {**nodes_final_dict, **rel_dict}\n",
    "\n",
    "with open('../resources/node_data/node_metadata_dict.pkl', 'wb') as handle:\n",
    "    pickle.dump(metadata, handle, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "If you are interested in enhancing this KG, you can explore the entire [RNA-KG](https://github.com/AnacletoLAB/RNA-KG/). Feel free to contact me at [emanuele dot cavalleri at unimi dot it](https://mail.google.com/mail/u/0/?view=cm&fs=1&tf=1&to=emanuele.cavalleri@unimi.it)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "pycharm": {
   "stem_cell": {
    "cell_type": "raw",
    "metadata": {
     "collapsed": false
    },
    "source": []
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
