{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <p style=\"text-align: center;\">Construction of a simplified RNA-based Knowledge Graph</p>\n",
    "\n",
    "**Objective:** [SDM-RDFizer](https://github.com/SDM-TIB/SDM-RDFizer) is an interpreter of mapping rules that allows the transformation of (un)structured data into RDF Knowledge Graphs. In this notebook, we wish to show the use of RDFizer for the generation of a RNA-based Knowledge Graph that involves **gene-disease**, **gene-miRNA**, and **miRNA-disease** relationships. These relationships are made available from the following public repositories:\n",
    "- [gene-disease](https://storage.googleapis.com/pheknowlator/current_build/data/original_data/curated_gene_disease_associations.tsv), provided by PheKnowLator\n",
    "- [gene-miRNA](https://dianalab.e-ce.uth.gr/downloads/tarbase_v8_data.tar.gz), provided by TarBase <!--- (we selected 50k random rows to reduce input size) -->\n",
    "- [miRNA-disease](http://watson.compbio.iupui.edu:8080/miR2Disease/download/AllEntries.txt), provided by miR2Disease\n",
    "\n",
    "The construction of the Knowledge Graph should be compliant with the following Ontologies:\n",
    "- Mondo Merged Disease Ontology ([**Mondo**](https://obofoundry.org/ontology/mondo.html)), a semi-automatically constructed Ontology that merges in multiple disease resources\n",
    "- Relations Ontology ([**RO**](https://obofoundry.org/ontology/ro.html)), a collection of OWL relations (ObjectProperties) intended for use across a wide variety of biological ontologies\n",
    "\n",
    "To achieve this task, we need to map objects in RDF triples by means of RDFizer facilities, we will consider:\n",
    "- [disease-mondo-map](https://storage.googleapis.com/pheknowlator/current_build/data/processed_data/DISEASE_MONDO_MAP.txt), provided by PheKnowLator\n",
    "- [mirnaLabel-mirbase-map](#genemirna), obtained from [miRBase](https://mirbase.org/)\n",
    "- [doid-mondo-map](#mirnadis), obtained by means of PheKnowLator ecosystem's functions\n",
    "\n",
    "All the data required for the execution of this notebook can be downloaded from this [testRNA-KG/RDFizer/files](https://drive.google.com/drive/u/3/folders/1e54vON8b82FpqcNGesiqiVORD3vV_z2R) Google Drive repository.\n",
    "\n",
    "**Authors:** [ECavalleri](https://mail.google.com/mail/u/0/?view=cm&fs=1&tf=1&to=emanuele.cavalleri@unimi.it), [MMesiti](https://mail.google.com/mail/u/0/?view=cm&fs=1&tf=1&to=marco.mesiti@unimi.it)\n",
    "\n",
    "**GitHub Repositories:** [testRNA-KG/RDFizer](https://github.com/emanuelecavalleri/testRNA-KG/tree/main/RDFizer_comparison), [PheKnowLator](https://github.com/callahantiff/PheKnowLator/) (PKT) \n",
    "<!--- **Release:** **[v2.0.0](https://github.com/callahantiff/PheKnowLator/wiki/v2.0.0)** --->\n",
    "  \n",
    "<br>  \n",
    "\n",
    "<a target=\"_blank\" href=\"https://private-user-images.githubusercontent.com/33032169/298911261-e6228042-137d-4241-8971-e674fe41d4f3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcxNzEwMTQsIm5iZiI6MTcwNzE3MDcxNCwicGF0aCI6Ii8zMzAzMjE2OS8yOTg5MTEyNjEtZTYyMjgwNDItMTM3ZC00MjQxLTg5NzEtZTY3NGZlNDFkNGYzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA1VDIyMDUxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgwMzA1MDAxODY4MWIxMjM4YTVmNDExYzQzMTk2YzFhODIzMDdmZjJhZjI5ZDU0NjFkZDdkYmRkZGI4NmMzN2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.knrkWr6nevh4TrujUVwmC3CwNlYAUBYfAt5-lZDtsMA\"> <img src=\"https://private-user-images.githubusercontent.com/33032169/298911261-e6228042-137d-4241-8971-e674fe41d4f3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcxNzEwMTQsIm5iZiI6MTcwNzE3MDcxNCwicGF0aCI6Ii8zMzAzMjE2OS8yOTg5MTEyNjEtZTYyMjgwNDItMTM3ZC00MjQxLTg5NzEtZTY3NGZlNDFkNGYzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA1VDIyMDUxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgwMzA1MDAxODY4MWIxMjM4YTVmNDExYzQzMTk2YzFhODIzMDdmZjJhZjI5ZDU0NjFkZDdkYmRkZGI4NmMzN2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.knrkWr6nevh4TrujUVwmC3CwNlYAUBYfAt5-lZDtsMA\"></a> \n",
    "\n",
    "(*Click Figure to Enlarge Image in Current Browser Tab*)\n",
    "\n",
    "<br>\n",
    "\n",
    "***\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download, load, and map data\n",
    "This section downloads and loads in several different types of data, which are needed to construct the Knowledge Graph. Identifiers have to be mapped to be compliant to Mondo and miRBase. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### gene-disease from [PheKnowLator](https://github.com/callahantiff/PheKnowLator)\n",
    "[PheKnowLator](https://github.com/callahantiff/PheKnowLator) is a tool developed to download, process, map, and clean data in order to build edges for a Knowledge Graph. It provides you a Human Disease Molecular Mechanisms Knowledge Graph that includes gene-disease associations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneId</th>\n",
       "      <th>geneSymbol</th>\n",
       "      <th>DSI</th>\n",
       "      <th>DPI</th>\n",
       "      <th>diseaseId</th>\n",
       "      <th>diseaseName</th>\n",
       "      <th>diseaseType</th>\n",
       "      <th>diseaseClass</th>\n",
       "      <th>diseaseSemanticType</th>\n",
       "      <th>score</th>\n",
       "      <th>EI</th>\n",
       "      <th>YearInitial</th>\n",
       "      <th>YearFinal</th>\n",
       "      <th>NofPmids</th>\n",
       "      <th>NofSnps</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.700</td>\n",
       "      <td>0.538</td>\n",
       "      <td>C0019209</td>\n",
       "      <td>Hepatomegaly</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>C23;C06</td>\n",
       "      <td>Finding</td>\n",
       "      <td>0.30</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.700</td>\n",
       "      <td>0.538</td>\n",
       "      <td>C0036341</td>\n",
       "      <td>Schizophrenia</td>\n",
       "      <td>disease</td>\n",
       "      <td>F03</td>\n",
       "      <td>Mental or Behavioral Dysfunction</td>\n",
       "      <td>0.30</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>0.529</td>\n",
       "      <td>0.769</td>\n",
       "      <td>C0007102</td>\n",
       "      <td>Malignant tumor of colon</td>\n",
       "      <td>disease</td>\n",
       "      <td>C06;C04</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>0.529</td>\n",
       "      <td>0.769</td>\n",
       "      <td>C0009375</td>\n",
       "      <td>Colonic Neoplasms</td>\n",
       "      <td>group</td>\n",
       "      <td>C06;C04</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.30</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>0.529</td>\n",
       "      <td>0.769</td>\n",
       "      <td>C0011265</td>\n",
       "      <td>Presenile dementia</td>\n",
       "      <td>disease</td>\n",
       "      <td>C10;F03</td>\n",
       "      <td>Mental or Behavioral Dysfunction</td>\n",
       "      <td>0.30</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1998.0</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84027</th>\n",
       "      <td>106480868</td>\n",
       "      <td>RN7SKP114</td>\n",
       "      <td>0.931</td>\n",
       "      <td>0.077</td>\n",
       "      <td>C4722327</td>\n",
       "      <td>PROSTATE CANCER, HEREDITARY, 1</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C12</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.30</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84028</th>\n",
       "      <td>106481323</td>\n",
       "      <td>RNU6-456P</td>\n",
       "      <td>0.931</td>\n",
       "      <td>0.077</td>\n",
       "      <td>C2931456</td>\n",
       "      <td>Prostate cancer, familial</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C12</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.30</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84029</th>\n",
       "      <td>106481323</td>\n",
       "      <td>RNU6-456P</td>\n",
       "      <td>0.931</td>\n",
       "      <td>0.077</td>\n",
       "      <td>C4722327</td>\n",
       "      <td>PROSTATE CANCER, HEREDITARY, 1</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C12</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.30</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84031</th>\n",
       "      <td>107075310</td>\n",
       "      <td>MTCO2P12</td>\n",
       "      <td>0.368</td>\n",
       "      <td>0.962</td>\n",
       "      <td>C0268237</td>\n",
       "      <td>Cytochrome-c Oxidase Deficiency</td>\n",
       "      <td>disease</td>\n",
       "      <td>C16;C18</td>\n",
       "      <td>Disease or Syndrome; Congenital Abnormality</td>\n",
       "      <td>0.33</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1999.0</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GENOMICS_ENGLAND</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84032</th>\n",
       "      <td>107305681</td>\n",
       "      <td>DHS6S1</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.077</td>\n",
       "      <td>C0730294</td>\n",
       "      <td>North Carolina macular dystrophy</td>\n",
       "      <td>disease</td>\n",
       "      <td>C16;C11</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.50</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human;ORPHANET</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74279 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          geneId geneSymbol    DSI    DPI diseaseId   \n",
       "0              1       A1BG  0.700  0.538  C0019209  \\\n",
       "1              1       A1BG  0.700  0.538  C0036341   \n",
       "3              2        A2M  0.529  0.769  C0007102   \n",
       "4              2        A2M  0.529  0.769  C0009375   \n",
       "5              2        A2M  0.529  0.769  C0011265   \n",
       "...          ...        ...    ...    ...       ...   \n",
       "84027  106480868  RN7SKP114  0.931  0.077  C4722327   \n",
       "84028  106481323  RNU6-456P  0.931  0.077  C2931456   \n",
       "84029  106481323  RNU6-456P  0.931  0.077  C4722327   \n",
       "84031  107075310   MTCO2P12  0.368  0.962  C0268237   \n",
       "84032  107305681     DHS6S1  1.000  0.077  C0730294   \n",
       "\n",
       "                            diseaseName diseaseType diseaseClass   \n",
       "0                          Hepatomegaly   phenotype      C23;C06  \\\n",
       "1                         Schizophrenia     disease          F03   \n",
       "3              Malignant tumor of colon     disease      C06;C04   \n",
       "4                     Colonic Neoplasms       group      C06;C04   \n",
       "5                    Presenile dementia     disease      C10;F03   \n",
       "...                                 ...         ...          ...   \n",
       "84027    PROSTATE CANCER, HEREDITARY, 1     disease      C04;C12   \n",
       "84028         Prostate cancer, familial     disease      C04;C12   \n",
       "84029    PROSTATE CANCER, HEREDITARY, 1     disease      C04;C12   \n",
       "84031   Cytochrome-c Oxidase Deficiency     disease      C16;C18   \n",
       "84032  North Carolina macular dystrophy     disease      C16;C11   \n",
       "\n",
       "                               diseaseSemanticType  score   EI  YearInitial   \n",
       "0                                          Finding   0.30  1.0       2017.0  \\\n",
       "1                 Mental or Behavioral Dysfunction   0.30  1.0       2015.0   \n",
       "3                               Neoplastic Process   0.31  1.0       2004.0   \n",
       "4                               Neoplastic Process   0.30  1.0       2004.0   \n",
       "5                 Mental or Behavioral Dysfunction   0.30  1.0       1998.0   \n",
       "...                                            ...    ...  ...          ...   \n",
       "84027                           Neoplastic Process   0.30  1.0       2018.0   \n",
       "84028                           Neoplastic Process   0.30  1.0       2018.0   \n",
       "84029                           Neoplastic Process   0.30  1.0       2018.0   \n",
       "84031  Disease or Syndrome; Congenital Abnormality   0.33  1.0       1999.0   \n",
       "84032                          Disease or Syndrome   0.50  1.0       2016.0   \n",
       "\n",
       "       YearFinal  NofPmids  NofSnps              source  \n",
       "0         2017.0         1        0           CTD_human  \n",
       "1         2015.0         1        0           CTD_human  \n",
       "3         2019.0         1        0           CTD_human  \n",
       "4         2004.0         1        0           CTD_human  \n",
       "5         2004.0         3        0           CTD_human  \n",
       "...          ...       ...      ...                 ...  \n",
       "84027     2018.0         1        0           CTD_human  \n",
       "84028     2018.0         1        0           CTD_human  \n",
       "84029     2018.0         1        0           CTD_human  \n",
       "84031     2011.0         0        0    GENOMICS_ENGLAND  \n",
       "84032     2016.0         1        0  CTD_human;ORPHANET  \n",
       "\n",
       "[74279 rows x 16 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "gene_disease = pd.read_csv(\"https://storage.googleapis.com/pheknowlator/current_build/\"+\n",
    "                            \"data/original_data/curated_gene_disease_associations.tsv\", sep='\\t')\n",
    "# The original data is filtered such that only records meeting the following criteria were included:\n",
    "# EI >= 1.0 (90th percentile)\n",
    "gene_disease = gene_disease[gene_disease['EI']>=1.0]\n",
    "gene_disease"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_disease[['geneId','diseaseId']].to_csv(\n",
    "    'files/gene-disease.csv', index=None, header=['gene', 'disease'])\n",
    "\n",
    "gene_disease_map = pd.read_csv(\"https://storage.googleapis.com/pheknowlator/current_build/\"+\n",
    "                                \"data/processed_data/DISEASE_MONDO_MAP.txt\", sep='\\t', header=None)\n",
    "gene_disease_map.to_csv(\n",
    "    'files/DISEASE_MONDO_MAP.csv', index=None, header=['umls', 'disease'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### gene-miRNA from [TarBase](https://dianalab.e-ce.uth.gr/html/diana/web/index.php?r=tarbasev8/index)<a class=\"anchor\" id=\"genemirna\"></a>\n",
    "DIANA-TarBase v8 is a reference database devoted to the indexing of experimentally supported microRNA (miRNA) targets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TarBase_v8_download.txt</th>\n",
       "      <th>geneName</th>\n",
       "      <th>mirna</th>\n",
       "      <th>species</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>tissue</th>\n",
       "      <th>category</th>\n",
       "      <th>method</th>\n",
       "      <th>positive_negative</th>\n",
       "      <th>direct_indirect</th>\n",
       "      <th>up_down</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0910001A06Rik(mmu)</td>\n",
       "      <td>0910001A06Rik(mmu)</td>\n",
       "      <td>mmu-miR-124-3p</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>UP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1200004M23Rik(mmu)</td>\n",
       "      <td>1200004M23Rik(mmu)</td>\n",
       "      <td>mmu-miR-124-3p</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1700027J05Rik(mmu)</td>\n",
       "      <td>1700027J05Rik(mmu)</td>\n",
       "      <td>mmu-miR-124-3p</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>UP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1810015A11Rik(mmu)</td>\n",
       "      <td>1810015A11Rik(mmu)</td>\n",
       "      <td>mmu-miR-124-3p</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2310047A01Rik(mmu)</td>\n",
       "      <td>2310047A01Rik(mmu)</td>\n",
       "      <td>mmu-miR-124-3p</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927115</th>\n",
       "      <td>vimentin(hsa)</td>\n",
       "      <td>vimentin(hsa)</td>\n",
       "      <td>hsa-miR-9-5p</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Western Blot</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927116</th>\n",
       "      <td>Â Â Â Â (PTPRG)(hsa)</td>\n",
       "      <td>Â Â Â Â (PTPRG)(hsa)</td>\n",
       "      <td>hsa-miR-146a-5p</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Microarrays</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>UP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927117</th>\n",
       "      <td>Î±-actin(mmu)</td>\n",
       "      <td>Î±-actin(mmu)</td>\n",
       "      <td>mmu-miR-24-3p</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>qPCR</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>UP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927118</th>\n",
       "      <td>Î±-actin(mmu)</td>\n",
       "      <td>Î±-actin(mmu)</td>\n",
       "      <td>mmu-miR-24-3p</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer/Malignant</td>\n",
       "      <td>Western Blot</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>INDIRECT</td>\n",
       "      <td>UP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927119</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>927120 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       TarBase_v8_download.txt              geneName            mirna   \n",
       "0           0910001A06Rik(mmu)    0910001A06Rik(mmu)   mmu-miR-124-3p  \\\n",
       "1           1200004M23Rik(mmu)    1200004M23Rik(mmu)   mmu-miR-124-3p   \n",
       "2           1700027J05Rik(mmu)    1700027J05Rik(mmu)   mmu-miR-124-3p   \n",
       "3           1810015A11Rik(mmu)    1810015A11Rik(mmu)   mmu-miR-124-3p   \n",
       "4           2310047A01Rik(mmu)    2310047A01Rik(mmu)   mmu-miR-124-3p   \n",
       "...                        ...                   ...              ...   \n",
       "927115           vimentin(hsa)         vimentin(hsa)     hsa-miR-9-5p   \n",
       "927116    Â Â Â Â (PTPRG)(hsa)  Â Â Â Â (PTPRG)(hsa)  hsa-miR-146a-5p   \n",
       "927117           Î±-actin(mmu)         Î±-actin(mmu)    mmu-miR-24-3p   \n",
       "927118           Î±-actin(mmu)         Î±-actin(mmu)    mmu-miR-24-3p   \n",
       "927119                     NaN                   NaN              NaN   \n",
       "\n",
       "             species cell_line tissue          category        method   \n",
       "0       Mus musculus      <NA>    NaN  Cancer/Malignant   Microarrays  \\\n",
       "1       Mus musculus      <NA>    NaN  Cancer/Malignant   Microarrays   \n",
       "2       Mus musculus      <NA>    NaN  Cancer/Malignant   Microarrays   \n",
       "3       Mus musculus      <NA>    NaN  Cancer/Malignant   Microarrays   \n",
       "4       Mus musculus      <NA>    NaN  Cancer/Malignant   Microarrays   \n",
       "...              ...       ...    ...               ...           ...   \n",
       "927115  Homo sapiens      <NA>    NaN  Cancer/Malignant  Western Blot   \n",
       "927116  Homo sapiens      <NA>    NaN  Cancer/Malignant   Microarrays   \n",
       "927117  Mus musculus      <NA>    NaN  Cancer/Malignant          qPCR   \n",
       "927118  Mus musculus      <NA>    NaN  Cancer/Malignant  Western Blot   \n",
       "927119           NaN      <NA>    NaN               NaN           NaN   \n",
       "\n",
       "       positive_negative direct_indirect up_down condition  \n",
       "0               POSITIVE        INDIRECT      UP       NaN  \n",
       "1               POSITIVE        INDIRECT    DOWN       NaN  \n",
       "2               POSITIVE        INDIRECT      UP       NaN  \n",
       "3               POSITIVE        INDIRECT    DOWN       NaN  \n",
       "4               POSITIVE        INDIRECT    DOWN       NaN  \n",
       "...                  ...             ...     ...       ...  \n",
       "927115          NEGATIVE        INDIRECT     NaN       NaN  \n",
       "927116          POSITIVE        INDIRECT      UP       NaN  \n",
       "927117          POSITIVE        INDIRECT      UP       NaN  \n",
       "927118          POSITIVE        INDIRECT      UP       NaN  \n",
       "927119               NaN             NaN     NaN       NaN  \n",
       "\n",
       "[927120 rows x 12 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv('https://dianalab.e-ce.uth.gr/downloads/tarbase_v8_data.tar.gz',\n",
    "                         compression='gzip', sep=\"\\t\", dtype={\"cell_line\": \"string\"})\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Data from https://www.mirbase.org/download/hsa.gff3\n",
      "##gff-version 3\n",
      "##date 2018-3-5\n",
      "#\n",
      "# Chromosomal coordinates of Homo sapiens microRNAs\n",
      "# microRNAs:               miRBase v22\n",
      "# genome-build-id:         GRCh38\n",
      "# genome-build-accession:  NCBI_Assembly:GCA_000001405.15\n",
      "#\n",
      "# Hairpin precursor sequences have type \"miRNA_primary_transcript\". \n",
      "# Note, these sequences do not represent the full primary transcript, \n",
      "# rather a predicted stem-loop portion that includes the precursor \n",
      "# miRNA. Mature sequences have type \"miRNA\".\n",
      "#\n",
      "\n",
      "     seq_id source                      type     start       end score strand   \n",
      "0      chr1      .  miRNA_primary_transcript     17369     17436     .      -  \\\n",
      "1      chr1      .                     miRNA     17409     17431     .      -   \n",
      "2      chr1      .                     miRNA     17369     17391     .      -   \n",
      "3      chr1      .  miRNA_primary_transcript     30366     30503     .      +   \n",
      "4      chr1      .                     miRNA     30438     30458     .      +   \n",
      "...     ...    ...                       ...       ...       ...   ...    ...   \n",
      "4796   chrY      .                     miRNA   2609229   2609252     .      +   \n",
      "4797   chrY      .  miRNA_primary_transcript   4606120   4606228     .      +   \n",
      "4798   chrY      .                     miRNA   4606140   4606158     .      +   \n",
      "4799   chrY      .  miRNA_primary_transcript  13479177  13479266     .      +   \n",
      "4800   chrY      .                     miRNA  13479189  13479214     .      +   \n",
      "\n",
      "     phase                                         attributes  \n",
      "0        .   ID=MI0022705;Alias=MI0022705;Name=hsa-mir-6859-1  \n",
      "1        .  ID=MIMAT0027618;Alias=MIMAT0027618;Name=hsa-mi...  \n",
      "2        .  ID=MIMAT0027619;Alias=MIMAT0027619;Name=hsa-mi...  \n",
      "3        .   ID=MI0006363;Alias=MI0006363;Name=hsa-mir-1302-2  \n",
      "4        .  ID=MIMAT0005890;Alias=MIMAT0005890;Name=hsa-mi...  \n",
      "...    ...                                                ...  \n",
      "4796     .  ID=MIMAT0023714_1;Alias=MIMAT0023714;Name=hsa-...  \n",
      "4797     .     ID=MI0032313;Alias=MI0032313;Name=hsa-mir-9985  \n",
      "4798     .  ID=MIMAT0039763;Alias=MIMAT0039763;Name=hsa-mi...  \n",
      "4799     .    ID=MI0039722;Alias=MI0039722;Name=hsa-mir-12120  \n",
      "4800     .  ID=MIMAT0049014;Alias=MIMAT0049014;Name=hsa-mi...  \n",
      "\n",
      "[4801 rows x 9 columns]\n"
     ]
    }
   ],
   "source": [
    "# For the time being, we keep only Homo sapiens rows\n",
    "data = data.dropna()[data.dropna()['species'].str.contains(\"Homo sapiens\")]\n",
    "\n",
    "# Moreover, we keep only 50k (random) rows to reduce input size\n",
    "#data = data.sample(n=50000)\n",
    "\n",
    "# This simplified KG ignores if a transcript is \"3p\" or \"5p\", so we store this information as additional column\n",
    "data['p'] = data[data['mirna'].str.contains(\"p\")]['mirna']\n",
    "data[\"p\"] = data[\"p\"].str[-2:]\n",
    "data['mirna'] = data['mirna'].str.replace(r'-[35]p$', '', regex=True)\n",
    "data['mirna'] = data['mirna'].str.lower()\n",
    "\n",
    "# If you're interested in understanding why Homo sapiens has -3p and -5p miRNAs:\n",
    "# https://pubmed.ncbi.nlm.nih.gov/12592000/\n",
    "# Putting aside the -(3/5)p information, we are essentially dealing with non-mature (aka hairpin) miRNA:\n",
    "# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3616697/\n",
    "\n",
    "from pkt_kg.utils import data_downloader\n",
    "import gffpandas.gffpandas as gffpd\n",
    "import os\n",
    "data_downloader('https://www.mirbase.org/download/hsa.gff3', '')\n",
    "\n",
    "miRBaseMap = gffpd.read_gff3('hsa.gff3')  \n",
    "os.remove('hsa.gff3')\n",
    "print(miRBaseMap.header)\n",
    "print(miRBaseMap.df)\n",
    "miRBaseMap = miRBaseMap.attributes_to_columns()\n",
    "miRBaseMap = miRBaseMap[['attributes']]\n",
    "miRBaseMap\n",
    "miRBaseMap = miRBaseMap.attributes.str.split(';',expand=True)\n",
    "# Keep only \"ID\" and \"Name\" columns\n",
    "miRBaseMap = miRBaseMap[[2,0]]\n",
    "# Remove substring \"ID=\"\n",
    "miRBaseMap[0] = miRBaseMap[0].str[3:]\n",
    "# Remove substring \"Name=\"\n",
    "miRBaseMap[2] = miRBaseMap[2].str[5:]\n",
    "# Keep only hairpin/stem-loop miRNAs\n",
    "# (those starting with MI and not MIMAT, last one is reserved for mature sequences)\n",
    "miRBaseMap = miRBaseMap[~miRBaseMap[0].str.startswith('MIMAT')]\n",
    "miRBaseMap.to_csv('files/MIRBASE_ID_ACCESSION_MAP.csv', header=['mirna','mirbase'], index=None)\n",
    "\n",
    "gene_entrez_map = pd.read_csv('https://storage.googleapis.com/pheknowlator/current_build/'+\n",
    "                               'data/processed_data/ENSEMBL_GENE_ENTREZ_GENE_MAP.txt', sep='\\t', header=None)\n",
    "\n",
    "data.merge(gene_entrez_map, left_on=['TarBase_v8_download.txt'], right_on=[0])[[1,'mirna']].to_csv('files/gene-miRNA.csv', index=None, header=['gene', 'mirna'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### miRNA-disease from [miR2Disease](http://watson.compbio.iupui.edu:8080/miR2Disease/)<a class=\"anchor\" id=\"mirnadis\"></a>\n",
    "miR2Disease, a manually curated database, aims at providing a comprehensive resource of miRNA deregulation in various human diseases."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>desc</th>\n",
       "      <th>doid</th>\n",
       "      <th>mirna</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>acute lymphoblastic leukemia (ALL)</td>\n",
       "      <td>DOID_9952</td>\n",
       "      <td>hsa-let-7e</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>microarray</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>MicroRNA expression signatures accurately disc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>acute lymphoblastic leukemia (ALL)</td>\n",
       "      <td>DOID_9952</td>\n",
       "      <td>hsa-mir-125a</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>microarray</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>MicroRNA expression signatures accurately disc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>acute lymphoblastic leukemia (ALL)</td>\n",
       "      <td>DOID_9952</td>\n",
       "      <td>hsa-mir-151*</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>microarray</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>MicroRNA expression signatures accurately disc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>acute lymphoblastic leukemia (ALL)</td>\n",
       "      <td>DOID_9952</td>\n",
       "      <td>hsa-mir-210</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>microarray</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>MicroRNA expression signatures accurately disc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>acute lymphoblastic leukemia (ALL)</td>\n",
       "      <td>DOID_9952</td>\n",
       "      <td>hsa-mir-22</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>northern blot, qRT-PCR etc</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>Gene silencing of MIR22 in acute lymphoblastic...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2619</th>\n",
       "      <td>Waldenstrom Macroglobulinemia (WM)</td>\n",
       "      <td>DOID_9080</td>\n",
       "      <td>hsa-mir-542-3p</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>microRNA expression in the biology, prognosis ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2620</th>\n",
       "      <td>Waldenstrom Macroglobulinemia (WM)</td>\n",
       "      <td>DOID_9080</td>\n",
       "      <td>hsa-mir-9*</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>microRNA expression in the biology, prognosis ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2621</th>\n",
       "      <td>Waldenstrom Macroglobulinemia (WM)</td>\n",
       "      <td>DOID_9080</td>\n",
       "      <td>hsa-mir-155</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>microRNA expression in the biology, prognosis ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2622</th>\n",
       "      <td>Waldenstrom Macroglobulinemia (WM)</td>\n",
       "      <td>DOID_9080</td>\n",
       "      <td>hsa-mir-184</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>microRNA expression in the biology, prognosis ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2623</th>\n",
       "      <td>Waldenstrom Macroglobulinemia (WM)</td>\n",
       "      <td>DOID_9080</td>\n",
       "      <td>hsa-mir-206</td>\n",
       "      <td>up-regulated</td>\n",
       "      <td>Northern blot, qRT-PCR etc</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>microRNA expression in the biology, prognosis ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2624 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    desc       doid           mirna   \n",
       "0     acute lymphoblastic leukemia (ALL)  DOID_9952      hsa-let-7e  \\\n",
       "1     acute lymphoblastic leukemia (ALL)  DOID_9952    hsa-mir-125a   \n",
       "2     acute lymphoblastic leukemia (ALL)  DOID_9952    hsa-mir-151*   \n",
       "3     acute lymphoblastic leukemia (ALL)  DOID_9952     hsa-mir-210   \n",
       "4     acute lymphoblastic leukemia (ALL)  DOID_9952      hsa-mir-22   \n",
       "...                                  ...        ...             ...   \n",
       "2619  Waldenstrom Macroglobulinemia (WM)  DOID_9080  hsa-mir-542-3p   \n",
       "2620  Waldenstrom Macroglobulinemia (WM)  DOID_9080      hsa-mir-9*   \n",
       "2621  Waldenstrom Macroglobulinemia (WM)  DOID_9080     hsa-mir-155   \n",
       "2622  Waldenstrom Macroglobulinemia (WM)  DOID_9080     hsa-mir-184   \n",
       "2623  Waldenstrom Macroglobulinemia (WM)  DOID_9080     hsa-mir-206   \n",
       "\n",
       "                   2                           3       4   \n",
       "0     down-regulated                  microarray  2007.0  \\\n",
       "1     down-regulated                  microarray  2007.0   \n",
       "2       up-regulated                  microarray  2007.0   \n",
       "3       up-regulated                  microarray  2007.0   \n",
       "4     down-regulated  northern blot, qRT-PCR etc  2009.0   \n",
       "...              ...                         ...     ...   \n",
       "2619    up-regulated  Northern blot, qRT-PCR etc  2009.0   \n",
       "2620  down-regulated  Northern blot, qRT-PCR etc  2009.0   \n",
       "2621    up-regulated  Northern blot, qRT-PCR etc  2009.0   \n",
       "2622    up-regulated  Northern blot, qRT-PCR etc  2009.0   \n",
       "2623    up-regulated  Northern blot, qRT-PCR etc  2009.0   \n",
       "\n",
       "                                                      5  \n",
       "0     MicroRNA expression signatures accurately disc...  \n",
       "1     MicroRNA expression signatures accurately disc...  \n",
       "2     MicroRNA expression signatures accurately disc...  \n",
       "3     MicroRNA expression signatures accurately disc...  \n",
       "4     Gene silencing of MIR22 in acute lymphoblastic...  \n",
       "...                                                 ...  \n",
       "2619  microRNA expression in the biology, prognosis ...  \n",
       "2620  microRNA expression in the biology, prognosis ...  \n",
       "2621  microRNA expression in the biology, prognosis ...  \n",
       "2622  microRNA expression in the biology, prognosis ...  \n",
       "2623  microRNA expression in the biology, prognosis ...  \n",
       "\n",
       "[2624 rows x 7 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mirna_disease = pd.read_csv('http://watson.compbio.iupui.edu:8080/miR2Disease/download/AllEntries.txt', sep=\"\\t\", header=None)\n",
    "mirna_disease[0] = mirna_disease[0].str.lower()\n",
    "\n",
    "descDOmap = pd.read_csv('http://watson.compbio.iupui.edu:8080/miR2Disease/download/diseaseList.txt', sep=\"\\t\")\n",
    "descDOmap.columns = ['desc', 'doid']\n",
    "descDOmap['doid'] = descDOmap['doid'].str.upper()\n",
    "descDOmap['doid'] = descDOmap['doid'].str.replace(':', '_')\n",
    "\n",
    "mirna_disease.columns = ['mirna', 'desc', 2,3,4,5]\n",
    "mirna_disease = pd.merge(descDOmap, mirna_disease, on=['desc'])\n",
    "mirna_disease"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Tuple\n",
    "from rdflib import Graph, URIRef\n",
    "\n",
    "def gets_ontology_class_dbxrefs(graph: Graph) -> Tuple:\n",
    "    dbx_uris: Dict = dict()\n",
    "    dbx = [x for x in graph if 'hasdbxref' in str(x[1]).lower() if isinstance(x[0], URIRef)]\n",
    "    for x in dbx:\n",
    "        if str(x[2]).lower() in dbx_uris.keys(): dbx_uris[str(x[2]).lower()].append(str(x[0]))\n",
    "        else: dbx_uris[str(x[2]).lower()] = [str(x[0])]\n",
    "    dbx_type = {str(x[2]).lower(): 'DbXref' for x in dbx}\n",
    "\n",
    "    ex_uris: Dict = dict()\n",
    "    ex = [x for x in graph if 'exactmatch' in str(x[1]).lower() if isinstance([0], URIRef)]\n",
    "    for x in ex:\n",
    "        if str(x[2]).lower() in ex_uris.keys(): ex_uris[str(x[2]).lower()].append(str(x[0]))\n",
    "        else: ex_uris[str(x[2]).lower()] = [str(x[0])]\n",
    "    ex_type = {str(x[2]).lower(): 'ExactMatch' for x in ex}\n",
    "\n",
    "    return {**dbx_uris, **ex_uris}, {**dbx_type, **ex_type}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "mondo_graph = Graph().parse(\"http://purl.obolibrary.org/obo/mondo.owl\", format='application/rdf+xml')\n",
    "dbxref_res = gets_ontology_class_dbxrefs(mondo_graph)[0]\n",
    "\n",
    "# Fix DOIDs (substitute : with _) and upper case them\n",
    "mondo_dict = {str(k).replace(':','_').upper(): {str(i).split('/')[-1].replace(':','_') for i in v}\n",
    "              for k, v in dbxref_res.items() if 'doid' in str(k)}\n",
    "\n",
    "with open('files/DOID_MONDO_MAP.csv', 'w') as outfile:\n",
    "    outfile.write(\"doid,disease\\n\")\n",
    "    for k, v in mondo_dict.items():\n",
    "        outfile.write(str(k) + ',' + str(v).replace('{','').replace(', ','').replace('\\'','').replace('}','') + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "mirbase_mirna_map = pd.read_csv(\"files/MIRBASE_ID_ACCESSION_MAP.csv\")\n",
    "mirna_disease.merge(mirbase_mirna_map, on=['mirna'])[['mirbase', 'doid']].to_csv('files/mirna-disease.csv', index=None, header=['mirbase', 'doid'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "We can copy-paste `mapping.ttl` content in [**rdf-grapher**](https://www.ldf.fi/service/rdf-grapher) to graphically visualize it.\n",
    "\n",
    "<a target=\"_blank\" href=https://github.com/emanuelecavalleri/testRNA-KG/assets/33032169/b0a7b4d6-85ac-427f-8034-3b2da07fed7c> <img src=https://github.com/emanuelecavalleri/testRNA-KG/assets/33032169/b0a7b4d6-85ac-427f-8034-3b2da07fed7c></a> \n",
    "\n",
    "(*Click Figure to Enlarge Image in Current Browser Tab*)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Another visualization tool is [**rdf-visualizer**](https://issemantic.net/rdf-visualizer), which allows us to dynamically inspect the graph (in the figure below you can find a screenshot of the output).\n",
    "\n",
    "<a target=\"_blank\" href=https://github.com/emanuelecavalleri/testRNA-KG/assets/33032169/f9a1feaa-7381-4fb4-8e72-78d097faaddc> <img src=https://github.com/emanuelecavalleri/testRNA-KG/assets/33032169/f9a1feaa-7381-4fb4-8e72-78d097faaddc></a> \n",
    "\n",
    "(*Click Figure to Enlarge Image in Current Browser Tab*)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Semantifying RNA-KG...\n",
      "TM: file:///home/ecavalleri/testRNA-KG/notebooks/gene-miRNA\n",
      "TM: file:///home/ecavalleri/testRNA-KG/notebooks/gene-disease\n",
      "TM: file:///home/ecavalleri/testRNA-KG/notebooks/miRNA-disease\n",
      "TM: file:///home/ecavalleri/testRNA-KG/notebooks/disease-mondo-map\n",
      "TM: file:///home/ecavalleri/testRNA-KG/notebooks/mirnaLabel-mirbase-map\n",
      "TM: file:///home/ecavalleri/testRNA-KG/notebooks/doid-mondo-map\n",
      "Successfully semantified RNA-KG.\n",
      "\n",
      "\n",
      "Successfully semantified all datasets in 25.313 seconds.\n"
     ]
    }
   ],
   "source": [
    "!python3 -m rdfizer -c config.ini"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us consider for instance *MEGF8* gene (entrez ID [1954](http://www.ncbi.nlm.nih.gov/gene/1954)). Biologically speaking, the protein encoded by this gene is a membrane protein that contains several [EGF-like](https://www.uniprot.org/keywords/KW-0245) and [PSI](https://www.ebi.ac.uk/interpro/entry/InterPro/IPR016201/) domains. Defects in this gene are a cause of [Carpenter syndrome 2](https://www.ncbi.nlm.nih.gov/medgen/905199) ([MONDO_0013998](http://purl.obolibrary.org/obo/MONDO_0013998))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Graph identifier=Nee8f1de8e98e4eb4a46c0e735b5641bf (<class 'rdflib.graph.Graph'>)>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from rdflib import Graph\n",
    "\n",
    "# Read in the RNA-KG generated using RDFizer\n",
    "g = Graph()\n",
    "g.parse(\"output/RNA-KG.ttl\", format='ttl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Utility function to add edges to a graph\n",
    "\n",
    "from typing import List, Set, Union\n",
    "\n",
    "def adds_edges_to_graph(graph: Graph, edge_list: Union[List, Set], progress_bar: bool = True) -> Graph:\n",
    "    \n",
    "    edge_list_updated = set(edge_list) if isinstance(edge_list, List) else edge_list\n",
    "    edge_set = edge_list_updated if progress_bar else edge_list_updated\n",
    "    for edge in edge_set: graph.add(edge)\n",
    "\n",
    "    return graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 33 edges\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Graph identifier=N65feae6a362347b9b301833770f73c66 (<class 'rdflib.graph.Graph'>)>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Obtain a Knowledge subGraph having only MEGF8 as subject\n",
    "\n",
    "from rdflib.extras.external_graph_libs import *\n",
    "from rdflib import Graph, URIRef\n",
    "\n",
    "filtered_triples = []\n",
    "for s, p, o in g:\n",
    "    if str(s).startswith(str(URIRef('http://www.ncbi.nlm.nih.gov/gene/1954'))):\n",
    "        filtered_triples += [(s, p, o)]\n",
    "\n",
    "print('There are {} edges'.format(len(filtered_triples)))\n",
    "sub_graph = adds_edges_to_graph(Graph(), filtered_triples)\n",
    "\n",
    "sub_graph.serialize(destination='output/subRNA-KG.nt', format='nt', encoding='utf-8')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below I depict 12 of them for the sake of visualization."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a target=\"_blank\" href=https://github.com/emanuelecavalleri/testRNA-KG/assets/33032169/ed6c1ead-03fa-4bd4-addf-010592f36a6b> <img src=https://github.com/emanuelecavalleri/testRNA-KG/assets/33032169/ed6c1ead-03fa-4bd4-addf-010592f36a6b></a> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "MEGF8 (entrez ID [1954](http://www.ncbi.nlm.nih.gov/gene/1954)):\n",
    "- *is_a* (http://www.w3.org/1999/02/22-rdf-syntax-ns#type) gene ([SO_0000704](http://purl.obolibrary.org/obo/SO_0000704))\n",
    "- which *causes or contributes to condition* ([RO_0003302](http://purl.obolibrary.org/obo/RO_0003302)) Carpenter syndrome ([MONDO_0013998](http://purl.obolibrary.org/obo/MONDO_0013998)),\n",
    "- and *interacts with* ([RO_0002434](http://purl.obolibrary.org/obo/RO_0002434)) hsa-mir-1271 ([MI0003814](https://mirbase.org/hairpin/MI0003814)).\n",
    "\n",
    "hsa-mir-1271 is a microRNA (miRNA) known to [target](https://en.wikipedia.org/wiki/MicroRNA#Targets) MEGF8 transcripts. If you are interested in this topic, please watch the seminar [*An introduction to the RNA world*](https://youtu.be/QjNeOWaKKEE) for further biologial information on how this targeting occurs."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
